<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Manipulating genbank records with Plasmid &mdash; Plasmid: A python library for gene editing 2023.06.29 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Designing primers with Plasmid" href="primer_design.html" />
    <link rel="prev" title="Installation" href="../installation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            Plasmid: A python library for gene editing
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">How-to guides &amp; Examples</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Manipulating genbank records with Plasmid</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#table-of-contents">Table of Contents</a></li>
<li class="toctree-l2"><a class="reference internal" href="#reading-genbank-or-fasta-files-a-class-anchor-id-reading-genbank-a">Reading genbank or fasta files <a class="anchor" id="reading_genbank"></a></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#visualizing-the-genbank-record-a-class-anchor-id-visualizing-genbank-a">Visualizing the genbank record <a class="anchor" id="visualizing_genbank"></a></a></li>
<li class="toctree-l3"><a class="reference internal" href="#selecting-and-filtering-for-genomic-features-a-class-anchor-id-filtering-genbank-a">Selecting and filtering for genomic features  <a class="anchor" id="filtering_genbank"></a></a></li>
<li class="toctree-l3"><a class="reference internal" href="#filtering-by-keyword-a-class-anchor-id-filtering-keyword-a">Filtering by keyword <a class="anchor" id="filtering_keyword"></a></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sequence-translation-a-class-anchor-id-gene-translation-a">Sequence translation  <a class="anchor" id="gene_translation"></a></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#creating-new-genbank-records-a-id-gene-concat-a">Creating new genbank records <a id='gene_concat'></a></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#concatenating-genomic-features">Concatenating genomic features</a></li>
<li class="toctree-l3"><a class="reference internal" href="#writing-genbank-records-a-class-anchor-id-genbank-write-a">Writing genbank records  <a class="anchor" id="genbank_write"></a></a></li>
<li class="toctree-l3"><a class="reference internal" href="#adding-new-annotations-to-genbank-records-a-class-anchor-id-gene-annotation-a">Adding new annotations to genbank records  <a class="anchor" id="gene_annotation"></a></a></li>
<li class="toctree-l3"><a class="reference internal" href="#annotation-via-amino-acid-search">Annotation via amino acid search</a></li>
<li class="toctree-l3"><a class="reference internal" href="#searching-for-open-reading-frames-a-class-anchor-id-gene-orf-search-a">Searching for open reading frames  <a class="anchor" id="gene_ORF_search"></a></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="primer_design.html">Designing primers with Plasmid</a></li>
<li class="toctree-l1"><a class="reference internal" href="clustering.html">Aligning and clustering sequences with Plasmid</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Plasmid: A python library for gene editing</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Manipulating genbank records with Plasmid</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/notebooks/genbank_editing.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="manipulating-genbank-records-with-plasmid">
<h1>Manipulating genbank records with Plasmid<a class="headerlink" href="#manipulating-genbank-records-with-plasmid" title="Permalink to this heading"></a></h1>
<p>This notebook illustrates the various functions contained in the <code class="docutils literal notranslate"><span class="pre">plasmid</span></code> class, which can be used to read and manipulate genbank files.</p>
<section id="table-of-contents">
<h2>Table of Contents<a class="headerlink" href="#table-of-contents" title="Permalink to this heading"></a></h2>
<ol class="arabic simple">
<li><p><span class="xref myst">Reading and viewing genbank files</span></p>
<ol class="arabic simple">
<li><p><span class="xref myst">Visualizing genbank records</span></p></li>
<li><p><span class="xref myst">Selecting and filtering for genomic features</span></p></li>
<li><p><span class="xref myst">Gene translation to amino acids</span></p></li>
</ol>
</li>
<li><p><span class="xref myst">Creating new genbank records</span></p>
<ol class="arabic simple">
<li><p><span class="xref myst">Writing genbank files</span></p></li>
<li><p><span class="xref myst">Sequence annotation</span></p></li>
<li><p><span class="xref myst">Searching for open reading frames</span></p></li>
</ol>
</li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">plasmid</span> <span class="k">as</span> <span class="nn">pge</span>
<span class="kn">import</span> <span class="nn">importlib</span>
<span class="n">importlib</span><span class="o">.</span><span class="n">reload</span><span class="p">(</span><span class="n">pge</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;module &#39;plasmid&#39; from &#39;/home/zchen/Public/python/lib/python3.11/site-packages/plasmid/__init__.py&#39;&gt;
</pre></div>
</div>
</div>
</div>
</section>
<section id="reading-genbank-or-fasta-files-a-class-anchor-id-reading-genbank-a">
<h2>Reading genbank or fasta files <a class="anchor" id="reading_genbank"></a><a class="headerlink" href="#reading-genbank-or-fasta-files-a-class-anchor-id-reading-genbank-a" title="Permalink to this heading"></a></h2>
<p>Genbank, fasta, and text files can be read using <code class="docutils literal notranslate"><span class="pre">read_genbank</span></code>. This function loads genbank information via <code class="docutils literal notranslate"><span class="pre">biopython</span></code> into a pandas like dataframe called <code class="docutils literal notranslate"><span class="pre">Plasmid</span></code>. This is analogous to <code class="docutils literal notranslate"><span class="pre">read_csv</span></code> in pandas. The underlying data in Biopython can be sliced, sorted, filtered, and concatenated like a string.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">pge</span><span class="o">.</span><span class="n">read_genbank</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Help on function read_genbank in module plasmid.plasmid:

read_genbank(fname)
    Reads a genbank file into plasmid class object
    fname = genbank file to read
    returns Plasmid object with data loaded
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df</span> <span class="o">=</span> <span class="n">pge</span><span class="o">.</span><span class="n">read_genbank</span><span class="p">(</span><span class="s1">&#39;../data/xRFP.gb&#39;</span><span class="p">)</span>
<span class="n">df</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>reading  ../data/xRFP.gb  as genbank file
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;class &#39;plasmid.plasmid.Plasmid&#39;&gt; at 0x7f443caa33d0
molecule_type:DNA
topology:circular
data_file_division:   
date:05-DEC-2022
accessions:[&#39;&lt;unknown&#39;, &#39;id&gt;&#39;]
keywords:[&#39;&#39;]
source:
organism:. .
taxonomy:[]
comment:
ApEinfo:methylated:1
                              locus_tag          type        location  length   
0                         AmpR promoter      promoter       [0:29](+)      29  \
1                              AmpR RBS           RBS      [29:70](+)      41   
2                          SpectomycinR           CDS     [70:868](+)     798   
3                             BBa_B0053    terminator    [878:948](+)      70   
4   J23108, 0.51, Constitutive Promoter      promoter  [1032:1067](+)      35   
5          RBS 1.00 strength, BBa_B0034           RBS  [1086:1098](+)      12   
6                       LacI, BBa_C0012           CDS  [1104:2196](+)    1092   
7                             Esp3I_fix  misc_feature  [2072:2078](+)       6   
8         Forward Terminator, BBa_B1002    terminator  [2196:2230](+)      34   
9   pLacO2, single operon pLac ZC082818      promoter  [2230:2268](+)      38   
10                                LacO2  protein_bind  [2238:2255](+)      17   
11                 RFP targeting region  misc_binding  [2268:2288](+)      20   
12         spec/1_xRFPg1_0_xRFPg1_0_.gb       unknown  [2268:2371](+)     103   
13      dCas9 handle conserved region A  protein_bind  [2288:2300](+)      12   
14     dCas9 handle conserved region A&#39;  protein_bind  [2304:2318](+)      14   
15      dCas9 handle conserved region B  protein_bind  [2318:2330](+)      12   
16                   S. pyog terminator    terminator  [2330:2371](+)      41   
17        Forward Terminator, BBa_B1010    terminator  [2371:2411](+)      40   
18                               pSC101    rep_origin  [2411:4448](+)    2037   
19                        T0 terminator    terminator  [2435:2541](+)     106   
20                      repA101 protein           CDS  [3296:4247](+)     951   

      color  
0   #0f7ffe  
1      cyan  
2   #d40037  
3   #9d1b1c  
4   #408000  
5      cyan  
6   #fc6665  
7      cyan  
8   #804000  
9   #008080  
10  #073f80  
11  #ff0000  
12     cyan  
13  #0080ff  
14  #0080ff  
15  #004080  
16  #804000  
17  #804000  
18  #b4abac  
19  #804000  
20  #d40037  
total length:4465
</pre></div>
</div>
</div>
</div>
<section id="visualizing-the-genbank-record-a-class-anchor-id-visualizing-genbank-a">
<h3>Visualizing the genbank record <a class="anchor" id="visualizing_genbank"></a><a class="headerlink" href="#visualizing-the-genbank-record-a-class-anchor-id-visualizing-genbank-a" title="Permalink to this heading"></a></h3>
<p>The raw sequence with color annotations can be displayed with <code class="docutils literal notranslate"><span class="pre">.get_colored()</span></code>, which generates a string containing colorization flags that can be interpreted by a bash terminal. The dataframe can also be sliced or cast as a string to obtain the raw sequence.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">get_colored</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Help on method get_colored in module plasmid.plasmid:

get_colored(width=0) method of plasmid.plasmid.Plasmid instance
    Print out colored representation of Plasmid
    width = number of characters per line
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># displays the first 1000bp of the plasmid</span>
<span class="n">x</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">1000</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="fm">__repr__</span><span class="p">())</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">get_colored</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">out</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;class &#39;plasmid.plasmid.Plasmid&#39;&gt; at 0x7f4455a3f950
molecule_type:DNA
topology:circular
data_file_division:   
date:05-DEC-2022
accessions:[&#39;&lt;unknown&#39;, &#39;id&gt;&#39;]
keywords:[&#39;&#39;]
source:
organism:. .
taxonomy:[]
comment:
ApEinfo:methylated:1
                   locus_tag        type      location  length    color
0  AmpR promoter subsequence    promoter     [0:29](+)      29  #0f7ffe
1       AmpR RBS subsequence         RBS    [29:70](+)      41     cyan
2   SpectomycinR subsequence         CDS   [70:868](+)     798  #d40037
3      BBa_B0053 subsequence  terminator  [878:948](+)      70  #9d1b1c
total length:1000

TTCAAATATCTATCCGCTCATGAGACAATAACCCTGATAAATGCTTCAATAATATTGAAAAAGGAAGAATATGAGTGAAAAAGTGCCCGCCGAGATTTCGGTGCAACTATCACAAGCACTCAACGTCATCGGGCGCCACTTGGAGTCGACGTTGCTGGCCGTGCATTTGTACGGCTCCGCACTGGATGGCGGATTGAAACCGTACAGTGATATTGATTTGCTGGTGACTGTAGCTGCACCGCTCAATGATGCCGTGCGGCAAGCCCTGCTCGTCGATCTCTTGGAGGTTTCAGCTTCCCCTGGCCAAAACAAGGCACTCCGCGCCTTGGAAGTGACCATCGTCGTGCACAGTGACATCGTACCTTGGCGTTATCCGGCCAGGCGGGAACTGCAGTTCGGAGAGTGGCAGCGCAAAGACATCCTTGCGGGCATCTTCGAGCCCGCCACAACCGATTCTGACTTGGCGATTCTGCTAACAAAGGCAAAGCAACATAGCGTCGTCTTGGCAGGTTCAGCAGCGAAGGATCTCTTCAGCTCAGTCCCAGAAAGCGATCTATTCAAGGCACTGGCCGATACTCTGAAGCTATGGAACTCGCCGCCAGATTGGGCGGGCGATGAGCGGAATGTAGTGCTTACTTTGTCTCGTATCTGGTACACCGCAGCAACCGGCAAGATCGCGCCAAAGGATGTTGCTGCCACTTGGGCAATGGCACGCTTGCCAGCTCAACATCAGCCCATCCTGTTGAATGCCAAGCGGGCTTATCTTGGGCAAGAAGAAGATTATTTGCCCGCTCGTGCGGATCAGGTGGCGGCGCTCATTAAATTCGTGAAGTATGAAGCAGTTAAACTGCTTGGTGCCAGCCAATGATAATACTAGCTCCGGCAAAAAAACGGGCAAGGTGTCACCACCCTGCCCTTTTTCTTTAAAACCGAAAAGATTACTTCGCGAAATAGGCGTATCACGAGGCAGAATTTCAGATAAAAAAAATCCTTAGCTTTC
</pre></div>
</div>
</div>
</div>
</section>
<section id="selecting-and-filtering-for-genomic-features-a-class-anchor-id-filtering-genbank-a">
<h3>Selecting and filtering for genomic features  <a class="anchor" id="filtering_genbank"></a><a class="headerlink" href="#selecting-and-filtering-for-genomic-features-a-class-anchor-id-filtering-genbank-a" title="Permalink to this heading"></a></h3>
<p>Genes can also be selected and sliced out of the dataframe via boolean selection like in pandas.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="fm">__getitem__</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Help on method __getitem__ in module plasmid.plasmid:

__getitem__(key) method of plasmid.plasmid.Plasmid instance
    Splice Plasmid sequence with range or filter through features on the Plasmid
    key = if key is a range or slice, then slice using base pair index from start to end
          if key is a str, then return the column values of pandas dataframe
          if key is an int, then return feature at the feature index
          if key is a list of integers, then return features only at the indexes in the list 
          if key is a boolean array, then return features only at indexes which evaluate True
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># select all genes which are terminator sequences</span>
<span class="n">x</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="mi">800</span><span class="p">:</span><span class="mi">2500</span><span class="p">]</span>
<span class="n">y</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;terminator&#39;</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">y</span><span class="o">.</span><span class="fm">__repr__</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="n">y</span><span class="o">.</span><span class="n">get_colored</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;class &#39;plasmid.plasmid.Plasmid&#39;&gt; at 0x7f4455b62ad0
molecule_type:DNA
topology:circular
data_file_division:   
date:05-DEC-2022
accessions:[&#39;&lt;unknown&#39;, &#39;id&gt;&#39;]
keywords:[&#39;&#39;]
source:
organism:. .
taxonomy:[]
comment:
ApEinfo:methylated:1
                                   locus_tag        type        location   
0                      BBa_B0053 subsequence  terminator     [78:148](+)  \
1  Forward Terminator, BBa_B1002 subsequence  terminator  [1396:1430](+)   
2             S. pyog terminator subsequence  terminator  [1530:1571](+)   
3  Forward Terminator, BBa_B1010 subsequence  terminator  [1571:1611](+)   
4                  T0 terminator subsequence  terminator  [1635:1700](+)   

   length    color  
0      70  #9d1b1c  
1      34  #804000  
2      41  #804000  
3      40  #804000  
4      65  #804000  
total length:1700

ATCAGGTGGCGGCGCTCATTAAATTCGTGAAGTATGAAGCAGTTAAACTGCTTGGTGCCAGCCAATGATAATACTAGCTCCGGCAAAAAAACGGGCAAGGTGTCACCACCCTGCCCTTTTTCTTTAAAACCGAAAAGATTACTTCGCGAAATAGGCGTATCACGAGGCAGAATTTCAGATAAAAAAAATCCTTAGCTTTCGCTAAGGATGATTTCTGGAATTCGCGGCCGCTCTGACAGCTAGCTCAGTCCTAGGTATAATGCTAGCAAAAGAATTCAAAAGATCTAAAGAGGAGAAAGGATCTATGGTGAATGTGAAACCAGTAACGTTATACGATGTCGCAGAGTATGCCGGTGTCTCTTATCAGACCGTTTCCCGCGTGGTGAACCAGGCCAGCCACGTTTCTGCGAAAACGCGGGAAAAAGTGGAAGCGGCGATGGCGGAGCTGAATTACATTCCCAACCGCGTGGCACAACAACTGGCGGGCAAACAGTCGTTGCTGATTGGCGTTGCCACCTCCAGTCTGGCCCTGCACGCGCCGTCGCAAATTGTCGCGGCGATTAAATCTCGCGCCGATCAACTGGGTGCCAGCGTGGTGGTGTCGATGGTAGAACGAAGCGGCGTCGAAGCCTGTAAAGCGGCGGTGCACAATCTTCTCGCGCAACGCGTCAGTGGGCTGATCATTAACTATCCGCTGGATGACCAGGATGCCATTGCTGTGGAAGCTGCCTGCACTAATGTTCCGGCGTTATTTCTTGATGTCTCTGACCAGACACCCATCAACAGTATTATTTTCTCCCATGAAGACGGTACGCGACTGGGCGTGGAGCATCTGGTCGCATTGGGTCACCAGCAAATCGCGCTGTTAGCGGGCCCATTAAGTTCTGTCTCGGCGCGTCTGCGTCTGGCTGGCTGGCATAAATATCTCACTCGCAATCAAATTCAGCCGATAGCGGAACGGGAAGGCGACTGGAGTGCCATGTCCGGTTTTCAACAAACCATGCAAATGCTGAATGAGGGCATCGTTCCCACTGCGATGCTGGTTGCCAACGATCAGATGGCGCTGGGCGCAATGCGCGCCATTACCGAGTCCGGGCTGCGCGTTGGTGCGGATATCTCGGTAGTGGGATACGACGATACCGAAGACAGCTCATGTTATATCCCGCCGTTAACCACCATCAAACAGGATTTTCGCCTGCTGGGGCAAACCAGCGTGGACCGCTTGCTGCAACTCTCTCAGGGCCAGGCGGTGAAGGGCAATCAGCTGTTGCCCGTTTCACTGGTGAAAAGAAAAACCACCCTGGCGCCCAATACGCGAACCGCCTCTCCCCGCGCGTTGGCCGATTCATTAATGCAGCTGGCACGACAGGTTTCCCGACTGGAAAGCGGGCAGTAACGCAAAAAACCCCGCTTCGGCGGGGTTTTTTCGCAATTGACAATGTGAGCGAGTAACAAGATACTGAGCACAAACTTTCAGTTTAGCGGTCTGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTTCGCCGCAAACCCCGCCCCTGACAGGGCGGGGTTTCGCCGCCTGTCAGACCAAGTTTACGAGCTCGCTTGGACTCCTGTTGATAGATCCAGTAATGACCTCAGAACTCCATCTGGATTTGTTCAGAACGC
</pre></div>
</div>
</div>
</div>
</section>
<section id="filtering-by-keyword-a-class-anchor-id-filtering-keyword-a">
<h3>Filtering by keyword <a class="anchor" id="filtering_keyword"></a><a class="headerlink" href="#filtering-by-keyword-a-class-anchor-id-filtering-keyword-a" title="Permalink to this heading"></a></h3>
<p>Keyword selection is also possible because of the compatibility with pandas. In the following, we create a boolean array by looking for all strings in the <code class="docutils literal notranslate"><span class="pre">locus_tag</span></code> column which contains <code class="docutils literal notranslate"><span class="pre">LacI</span></code>. We use the boolean array to filter for genomic features satisfying this keyword. <code class="docutils literal notranslate"><span class="pre">.splice()</span></code> is used to slice out the gene sequence from the parent plasmid.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">splice</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Help on method splice in module plasmid.plasmid:

splice(inplace=False) method of plasmid.plasmid.Plasmid instance
    Get a slice sequence from start of all features to end of all features
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Select all genes containing LacI in the locus_tag column</span>
<span class="n">filt</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;LacI&#39;</span><span class="p">)</span>
<span class="n">y</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">filt</span><span class="p">]</span>
<span class="n">y</span> <span class="o">=</span> <span class="n">y</span><span class="o">.</span><span class="n">splice</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">y</span><span class="o">.</span><span class="fm">__repr__</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="n">y</span><span class="o">.</span><span class="n">get_colored</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;class &#39;plasmid.plasmid.Plasmid&#39;&gt; at 0x7f443dba6590
molecule_type:DNA
topology:circular
data_file_division:   
date:05-DEC-2022
accessions:[&#39;&lt;unknown&#39;, &#39;id&gt;&#39;]
keywords:[&#39;&#39;]
source:
organism:. .
taxonomy:[]
comment:
ApEinfo:methylated:1
                     locus_tag type     location  length    color
0  LacI, BBa_C0012 subsequence  CDS  [0:1092](+)    1092  #fc6665
total length:1092

ATGGTGAATGTGAAACCAGTAACGTTATACGATGTCGCAGAGTATGCCGGTGTCTCTTATCAGACCGTTTCCCGCGTGGTGAACCAGGCCAGCCACGTTTCTGCGAAAACGCGGGAAAAAGTGGAAGCGGCGATGGCGGAGCTGAATTACATTCCCAACCGCGTGGCACAACAACTGGCGGGCAAACAGTCGTTGCTGATTGGCGTTGCCACCTCCAGTCTGGCCCTGCACGCGCCGTCGCAAATTGTCGCGGCGATTAAATCTCGCGCCGATCAACTGGGTGCCAGCGTGGTGGTGTCGATGGTAGAACGAAGCGGCGTCGAAGCCTGTAAAGCGGCGGTGCACAATCTTCTCGCGCAACGCGTCAGTGGGCTGATCATTAACTATCCGCTGGATGACCAGGATGCCATTGCTGTGGAAGCTGCCTGCACTAATGTTCCGGCGTTATTTCTTGATGTCTCTGACCAGACACCCATCAACAGTATTATTTTCTCCCATGAAGACGGTACGCGACTGGGCGTGGAGCATCTGGTCGCATTGGGTCACCAGCAAATCGCGCTGTTAGCGGGCCCATTAAGTTCTGTCTCGGCGCGTCTGCGTCTGGCTGGCTGGCATAAATATCTCACTCGCAATCAAATTCAGCCGATAGCGGAACGGGAAGGCGACTGGAGTGCCATGTCCGGTTTTCAACAAACCATGCAAATGCTGAATGAGGGCATCGTTCCCACTGCGATGCTGGTTGCCAACGATCAGATGGCGCTGGGCGCAATGCGCGCCATTACCGAGTCCGGGCTGCGCGTTGGTGCGGATATCTCGGTAGTGGGATACGACGATACCGAAGACAGCTCATGTTATATCCCGCCGTTAACCACCATCAAACAGGATTTTCGCCTGCTGGGGCAAACCAGCGTGGACCGCTTGCTGCAACTCTCTCAGGGCCAGGCGGTGAAGGGCAATCAGCTGTTGCCCGTTTCACTGGTGAAAAGAAAAACCACCCTGGCGCCCAATACGCGAACCGCCTCTCCCCGCGCGTTGGCCGATTCATTAATGCAGCTGGCACGACAGGTTTCCCGACTGGAAAGCGGGCAGTAA
</pre></div>
</div>
</div>
</div>
</section>
<section id="sequence-translation-a-class-anchor-id-gene-translation-a">
<h3>Sequence translation  <a class="anchor" id="gene_translation"></a><a class="headerlink" href="#sequence-translation-a-class-anchor-id-gene-translation-a" title="Permalink to this heading"></a></h3>
<p>DNA sequences can be translated to amino acid sequences via <code class="docutils literal notranslate"><span class="pre">.translate()</span></code></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">translate</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Help on method translate in module plasmid.plasmid:

translate(strand=None, frame=0, table=&#39;Standard&#39;) method of plasmid.plasmid.Plasmid instance
    Translate the DNA sequence to amino acids
    frame = codon reading frame such as 0, 1, or 2
    table = codon table to use. Standard codon table is used by default
    return a string
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">y</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;LacI&#39;</span><span class="p">)]</span><span class="o">.</span><span class="n">splice</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;This is the DNA sequence&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">y</span><span class="o">.</span><span class="n">get_colored</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;This is the amino acid sequence&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">y</span><span class="o">.</span><span class="n">translate</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>This is the DNA sequence
ATGGTGAATGTGAAACCAGTAACGTTATACGATGTCGCAGAGTATGCCGGTGTCTCTTATCAGACCGTTTCCCGCGTGGTGAACCAGGCCAGCCACGTTTCTGCGAAAACGCGGGAAAAAGTGGAAGCGGCGATGGCGGAGCTGAATTACATTCCCAACCGCGTGGCACAACAACTGGCGGGCAAACAGTCGTTGCTGATTGGCGTTGCCACCTCCAGTCTGGCCCTGCACGCGCCGTCGCAAATTGTCGCGGCGATTAAATCTCGCGCCGATCAACTGGGTGCCAGCGTGGTGGTGTCGATGGTAGAACGAAGCGGCGTCGAAGCCTGTAAAGCGGCGGTGCACAATCTTCTCGCGCAACGCGTCAGTGGGCTGATCATTAACTATCCGCTGGATGACCAGGATGCCATTGCTGTGGAAGCTGCCTGCACTAATGTTCCGGCGTTATTTCTTGATGTCTCTGACCAGACACCCATCAACAGTATTATTTTCTCCCATGAAGACGGTACGCGACTGGGCGTGGAGCATCTGGTCGCATTGGGTCACCAGCAAATCGCGCTGTTAGCGGGCCCATTAAGTTCTGTCTCGGCGCGTCTGCGTCTGGCTGGCTGGCATAAATATCTCACTCGCAATCAAATTCAGCCGATAGCGGAACGGGAAGGCGACTGGAGTGCCATGTCCGGTTTTCAACAAACCATGCAAATGCTGAATGAGGGCATCGTTCCCACTGCGATGCTGGTTGCCAACGATCAGATGGCGCTGGGCGCAATGCGCGCCATTACCGAGTCCGGGCTGCGCGTTGGTGCGGATATCTCGGTAGTGGGATACGACGATACCGAAGACAGCTCATGTTATATCCCGCCGTTAACCACCATCAAACAGGATTTTCGCCTGCTGGGGCAAACCAGCGTGGACCGCTTGCTGCAACTCTCTCAGGGCCAGGCGGTGAAGGGCAATCAGCTGTTGCCCGTTTCACTGGTGAAAAGAAAAACCACCCTGGCGCCCAATACGCGAACCGCCTCTCCCCGCGCGTTGGCCGATTCATTAATGCAGCTGGCACGACAGGTTTCCCGACTGGAAAGCGGGCAGTAA
This is the amino acid sequence
MVNVKPVTLYDVAEYAGVSYQTVSRVVNQASHVSAKTREKVEAAMAELNYIPNRVAQQLAGKQSLLIGVATSSLALHAPSQIVAAIKSRADQLGASVVVSMVERSGVEACKAAVHNLLAQRVSGLIINYPLDDQDAIAVEAACTNVPALFLDVSDQTPINSIIFSHEDGTRLGVEHLVALGHQQIALLAGPLSSVSARLRLAGWHKYLTRNQIQPIAEREGDWSAMSGFQQTMQMLNEGIVPTAMLVANDQMALGAMRAITESGLRVGADISVVGYDDTEDSSCYIPPLTTIKQDFRLLGQTSVDRLLQLSQGQAVKGNQLLPVSLVKRKTTLAPNTRTASPRALADSLMQLARQVSRLESGQ*
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="creating-new-genbank-records-a-id-gene-concat-a">
<h2>Creating new genbank records <a id='gene_concat'></a><a class="headerlink" href="#creating-new-genbank-records-a-id-gene-concat-a" title="Permalink to this heading"></a></h2>
<section id="concatenating-genomic-features">
<h3>Concatenating genomic features<a class="headerlink" href="#concatenating-genomic-features" title="Permalink to this heading"></a></h3>
<p>New genes and constructs can be generated by combining genetic parts via concatenation. The following extracts the RFP fluorescent reporter, ribosome binding site, and pLac promoter genes from existing genbank records and concatenates the components to form a new genbank record.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># slice out the RFP gene</span>
<span class="n">RFP</span> <span class="o">=</span> <span class="n">pge</span><span class="o">.</span><span class="n">read_genbank</span><span class="p">(</span><span class="s1">&#39;../data/dcas9_RFP.gb&#39;</span><span class="p">)</span>
<span class="n">RFP</span> <span class="o">=</span> <span class="n">RFP</span><span class="p">[</span><span class="n">RFP</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;mRFP&#39;</span><span class="p">)]</span><span class="o">.</span><span class="n">splice</span><span class="p">()</span>
<span class="c1"># slice out the ribosome binding site</span>
<span class="n">RBS</span> <span class="o">=</span> <span class="n">pge</span><span class="o">.</span><span class="n">read_genbank</span><span class="p">(</span><span class="s1">&#39;../data/xRFP.gb&#39;</span><span class="p">)</span>
<span class="n">RBS</span> <span class="o">=</span> <span class="n">RBS</span><span class="p">[</span><span class="n">RBS</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;BBa_B0034&#39;</span><span class="p">)]</span><span class="o">.</span><span class="n">splice</span><span class="p">()</span>
<span class="c1"># slice out the promoter</span>
<span class="n">pLac</span> <span class="o">=</span> <span class="n">pge</span><span class="o">.</span><span class="n">read_genbank</span><span class="p">(</span><span class="s1">&#39;../data/xRFP.gb&#39;</span><span class="p">)</span>
<span class="n">pLac</span> <span class="o">=</span> <span class="n">pLac</span><span class="p">[</span><span class="n">pLac</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;pLac&#39;</span><span class="p">)]</span><span class="o">.</span><span class="n">splice</span><span class="p">()</span>

<span class="c1"># assemble the promoter, rbs, and mRFP, adds spacer sequences between promoter, rbs, and rfp</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pLac</span> <span class="o">+</span> <span class="s1">&#39;gagacc&#39;</span> <span class="o">+</span> <span class="n">RBS</span> <span class="o">+</span> <span class="s1">&#39;ggtctc&#39;</span> <span class="o">+</span> <span class="n">RFP</span>
<span class="nb">print</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="fm">__repr__</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">get_colored</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>reading  ../data/dcas9_RFP.gb  as genbank file
reading  ../data/xRFP.gb  as genbank file
reading  ../data/xRFP.gb  as genbank file
&lt;class &#39;plasmid.plasmid.Plasmid&#39;&gt; at 0x7f44941fc350
molecule_type:DNA
topology:circular
                                         locus_tag      type     location   
0  pLacO2, single operon pLac ZC082818 subsequence  promoter    [0:38](+)  \
1         RBS 1.00 strength, BBa_B0034 subsequence       RBS   [44:56](+)   
2      mRFP, uniprot drFP583, pdb 2H5O subsequence       CDS  [62:740](+)   

   length    color  
0      38  #008080  
1      12     cyan  
2     678  #800040  
total length:740

AATTGACAATGTGAGCGAGTAACAAGATACTGAGCACAgagaccAAAGAGGAGAAAggtctcATGGCGAGTAGCGAAGACGTTATCAAAGAGTTCATGCGTTTCAAAGTTCGTATGGAAGGTTCCGTTAACGGTCACGAGTTCGAAATCGAAGGTGAAGGTGAAGGTCGTCCGTACGAAGGTACCCAGACCGCTAAACTGAAAGTTACCAAAGGTGGTCCGCTGCCGTTCGCTTGGGACATCCTGTCCCCGCAGTTCCAGTACGGTTCCAAAGCTTACGTTAAACACCCGGCTGACATCCCGGACTACCTGAAACTGTCCTTCCCGGAAGGTTTCAAATGGGAACGTGTTATGAACTTCGAAGACGGTGGTGTTGTTACCGTTACCCAGGACTCCTCCCTGCAAGACGGTGAGTTCATCTACAAAGTTAAACTGCGTGGTACCAACTTCCCGTCCGACGGTCCGGTTATGCAGAAAAAAACCATGGGTTGGGAAGCTTCCACCGAACGTATGTACCCGGAAGACGGTGCTCTGAAAGGTGAAATCAAAATGCGTCTGAAACTGAAAGACGGTGGTCACTACGACGCTGAAGTTAAAACCACCTACATGGCTAAAAAACCGGTTCAGCTGCCGGGTGCTTACAAAACCGACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTTGAACAGTACGAACGTGCTGAAGGTCGTCACTCCACCGGTGCTTAA
</pre></div>
</div>
</div>
</div>
</section>
<section id="writing-genbank-records-a-class-anchor-id-genbank-write-a">
<h3>Writing genbank records  <a class="anchor" id="genbank_write"></a><a class="headerlink" href="#writing-genbank-records-a-class-anchor-id-genbank-write-a" title="Permalink to this heading"></a></h3>
<p>We can write new gene products to genbank records with the <code class="docutils literal notranslate"><span class="pre">.to_genbank</span></code> function.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">to_genbank</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Help on method to_genbank in module plasmid.plasmid:

to_genbank(filename, timestamp=False) method of plasmid.plasmid.Plasmid instance
    Writes the Plasmid SeqRecord to a filename
    filename = filename of output file
    timestamp = adds a timestamp to the output file
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># create a new gene</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pLac</span> <span class="o">+</span> <span class="s1">&#39;gagacc&#39;</span> <span class="o">+</span> <span class="n">RBS</span> <span class="o">+</span> <span class="s1">&#39;ggtctc&#39;</span> <span class="o">+</span> <span class="n">RFP</span>
<span class="nb">print</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="fm">__repr__</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">get_colored</span><span class="p">())</span>
<span class="c1"># write the gene to genbank</span>
<span class="n">df</span><span class="o">.</span><span class="n">to_genbank</span><span class="p">(</span><span class="s1">&#39;test.gb&#39;</span><span class="p">)</span>
<span class="c1"># read the raw genbank file</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;test.gb&#39;</span><span class="p">,</span><span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">read</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;class &#39;plasmid.plasmid.Plasmid&#39;&gt; at 0x7f443caf8750
molecule_type:DNA
topology:circular
                                         locus_tag      type     location   
0  pLacO2, single operon pLac ZC082818 subsequence  promoter    [0:38](+)  \
1         RBS 1.00 strength, BBa_B0034 subsequence       RBS   [44:56](+)   
2      mRFP, uniprot drFP583, pdb 2H5O subsequence       CDS  [62:740](+)   

   length    color  
0      38  #008080  
1      12     cyan  
2     678  #800040  
total length:740

AATTGACAATGTGAGCGAGTAACAAGATACTGAGCACAgagaccAAAGAGGAGAAAggtctcATGGCGAGTAGCGAAGACGTTATCAAAGAGTTCATGCGTTTCAAAGTTCGTATGGAAGGTTCCGTTAACGGTCACGAGTTCGAAATCGAAGGTGAAGGTGAAGGTCGTCCGTACGAAGGTACCCAGACCGCTAAACTGAAAGTTACCAAAGGTGGTCCGCTGCCGTTCGCTTGGGACATCCTGTCCCCGCAGTTCCAGTACGGTTCCAAAGCTTACGTTAAACACCCGGCTGACATCCCGGACTACCTGAAACTGTCCTTCCCGGAAGGTTTCAAATGGGAACGTGTTATGAACTTCGAAGACGGTGGTGTTGTTACCGTTACCCAGGACTCCTCCCTGCAAGACGGTGAGTTCATCTACAAAGTTAAACTGCGTGGTACCAACTTCCCGTCCGACGGTCCGGTTATGCAGAAAAAAACCATGGGTTGGGAAGCTTCCACCGAACGTATGTACCCGGAAGACGGTGCTCTGAAAGGTGAAATCAAAATGCGTCTGAAACTGAAAGACGGTGGTCACTACGACGCTGAAGTTAAAACCACCTACATGGCTAAAAAACCGGTTCAGCTGCCGGGTGCTTACAAAACCGACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTTGAACAGTACGAACGTGCTGAAGGTCGTCACTCCACCGGTGCTTAA
LOCUS       .                        740 bp    DNA     circular UNK 01-JAN-1980
DEFINITION  .
ACCESSION   &lt;unknown id&gt;
VERSION     &lt;unknown id&gt;
KEYWORDS    .
SOURCE      .
  ORGANISM  .
            .
FEATURES             Location/Qualifiers
     promoter        1..38
                     /locus_tag=&quot;pLacO2, single operon pLac ZC082818
                     subsequence&quot;
                     /ApEinfo_label=&quot;pLacO2, single operon pLac ZC082818
                     subsequence&quot;
                     /ApEinfo_fwdcolor=&quot;#008080&quot;
                     /ApEinfo_revcolor=&quot;#008080&quot;
     RBS             45..56
                     /locus_tag=&quot;RBS 1.00 strength, BBa_B0034 subsequence&quot;
                     /ApEinfo_label=&quot;RBS 1.00 strength, BBa_B0034 subsequence&quot;
                     /ApEinfo_fwdcolor=&quot;cyan&quot;
                     /ApEinfo_revcolor=&quot;cyan&quot;
     CDS             63..740
                     /locus_tag=&quot;mRFP, uniprot drFP583, pdb 2H5O subsequence&quot;
                     /ApEinfo_label=&quot;mRFP, uniprot drFP583, pdb 2H5O
                     subsequence&quot;
                     /ApEinfo_fwdcolor=&quot;#800040&quot;
                     /ApEinfo_revcolor=&quot;#800040&quot;
ORIGIN
        1 aattgacaat gtgagcgagt aacaagatac tgagcacaga gaccaaagag gagaaaggtc
       61 tcatggcgag tagcgaagac gttatcaaag agttcatgcg tttcaaagtt cgtatggaag
      121 gttccgttaa cggtcacgag ttcgaaatcg aaggtgaagg tgaaggtcgt ccgtacgaag
      181 gtacccagac cgctaaactg aaagttacca aaggtggtcc gctgccgttc gcttgggaca
      241 tcctgtcccc gcagttccag tacggttcca aagcttacgt taaacacccg gctgacatcc
      301 cggactacct gaaactgtcc ttcccggaag gtttcaaatg ggaacgtgtt atgaacttcg
      361 aagacggtgg tgttgttacc gttacccagg actcctccct gcaagacggt gagttcatct
      421 acaaagttaa actgcgtggt accaacttcc cgtccgacgg tccggttatg cagaaaaaaa
      481 ccatgggttg ggaagcttcc accgaacgta tgtacccgga agacggtgct ctgaaaggtg
      541 aaatcaaaat gcgtctgaaa ctgaaagacg gtggtcacta cgacgctgaa gttaaaacca
      601 cctacatggc taaaaaaccg gttcagctgc cgggtgctta caaaaccgac atcaaactgg
      661 acatcacctc ccacaacgaa gactacacca tcgttgaaca gtacgaacgt gctgaaggtc
      721 gtcactccac cggtgcttaa
//
</pre></div>
</div>
</div>
</div>
</section>
<section id="adding-new-annotations-to-genbank-records-a-class-anchor-id-gene-annotation-a">
<h3>Adding new annotations to genbank records  <a class="anchor" id="gene_annotation"></a><a class="headerlink" href="#adding-new-annotations-to-genbank-records-a-class-anchor-id-gene-annotation-a" title="Permalink to this heading"></a></h3>
<p>Annotation of new genes or features can be done with the <code class="docutils literal notranslate"><span class="pre">.annotate</span></code> function. In the following, we annotate BsaI restriction enzyme sites in a new construct we just generated.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">annotate</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Help on method annotate in module plasmid.plasmid:

annotate(name, sequence, feature=&#39;unknown&#39;, color=None, circular=True, inplace=False) method of plasmid.plasmid.Plasmid instance
    Adds annotations to a plasmid using sequence information
    name = name of the gene
    sequence = DNA or amino acid sequence of the feature
    feature = type of genetic feature such as cds, mRNA, primer_bind
    color = [fwd_color, rev_color] to use
    circular = search as though sequence is a circular construct
    inplace = performs modifications inplace
    returns a modified plasmid dataframe
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># assemble the promoter, rbs, and mRFP, adds spacer sequences between promoter, rbs, and rfp</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pLac</span> <span class="o">+</span> <span class="s1">&#39;gagacc&#39;</span> <span class="o">+</span> <span class="n">RBS</span> <span class="o">+</span> <span class="s1">&#39;ggtctc&#39;</span> <span class="o">+</span> <span class="n">RFP</span>
<span class="c1"># annotate BsaI sites</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;BsaI&#39;</span><span class="p">,</span> <span class="n">sequence</span><span class="o">=</span><span class="s1">&#39;GGTCTC&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;red&#39;</span><span class="p">,</span><span class="s1">&#39;orange&#39;</span><span class="p">],</span> <span class="n">feature</span><span class="o">=</span><span class="s1">&#39;protein_bind&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="fm">__repr__</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">get_colored</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;class &#39;plasmid.plasmid.Plasmid&#39;&gt; at 0x7f443caba710
molecule_type:DNA
topology:circular
                                         locus_tag          type     location   
0  pLacO2, single operon pLac ZC082818 subsequence      promoter    [0:38](+)  \
1                                             BsaI  protein_bind   [38:44](-)   
2         RBS 1.00 strength, BBa_B0034 subsequence           RBS   [44:56](+)   
3                                             BsaI  protein_bind   [56:62](+)   
4      mRFP, uniprot drFP583, pdb 2H5O subsequence           CDS  [62:740](+)   

   length    color  
0      38  #008080  
1       6   orange  
2      12     cyan  
3       6      red  
4     678  #800040  
total length:740

AATTGACAATGTGAGCGAGTAACAAGATACTGAGCACAgagaccAAAGAGGAGAAAggtctcATGGCGAGTAGCGAAGACGTTATCAAAGAGTTCATGCGTTTCAAAGTTCGTATGGAAGGTTCCGTTAACGGTCACGAGTTCGAAATCGAAGGTGAAGGTGAAGGTCGTCCGTACGAAGGTACCCAGACCGCTAAACTGAAAGTTACCAAAGGTGGTCCGCTGCCGTTCGCTTGGGACATCCTGTCCCCGCAGTTCCAGTACGGTTCCAAAGCTTACGTTAAACACCCGGCTGACATCCCGGACTACCTGAAACTGTCCTTCCCGGAAGGTTTCAAATGGGAACGTGTTATGAACTTCGAAGACGGTGGTGTTGTTACCGTTACCCAGGACTCCTCCCTGCAAGACGGTGAGTTCATCTACAAAGTTAAACTGCGTGGTACCAACTTCCCGTCCGACGGTCCGGTTATGCAGAAAAAAACCATGGGTTGGGAAGCTTCCACCGAACGTATGTACCCGGAAGACGGTGCTCTGAAAGGTGAAATCAAAATGCGTCTGAAACTGAAAGACGGTGGTCACTACGACGCTGAAGTTAAAACCACCTACATGGCTAAAAAACCGGTTCAGCTGCCGGGTGCTTACAAAACCGACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTTGAACAGTACGAACGTGCTGAAGGTCGTCACTCCACCGGTGCTTAA
</pre></div>
</div>
</div>
</div>
</section>
<section id="annotation-via-amino-acid-search">
<h3>Annotation via amino acid search<a class="headerlink" href="#annotation-via-amino-acid-search" title="Permalink to this heading"></a></h3>
<p>Annotations can also be done via searching for amino acid sequences. In the following, we search for a short amino acid sequence in the LacI gene and annotate it. Exact amino acid search is performed in both forward and reverse orientations for the given gene.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">RFP</span><span class="o">.</span><span class="n">translate</span><span class="p">())</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pLac</span> <span class="o">+</span> <span class="s1">&#39;gagacc&#39;</span> <span class="o">+</span> <span class="n">RBS</span> <span class="o">+</span> <span class="s1">&#39;ggtctc&#39;</span> <span class="o">+</span> <span class="n">RFP</span>
<span class="n">aaseq</span> <span class="o">=</span> <span class="s1">&#39;DGALKGEIKMRLKLKDG&#39;</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;peptide&#39;</span><span class="p">,</span> <span class="n">sequence</span><span class="o">=</span><span class="n">aaseq</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;orange&#39;</span><span class="p">)</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">get_colored</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>MASSEDVIKEFMRFKVRMEGSVNGHEFEIEGEGEGRPYEGTQTAKLKVTKGGPLPFAWDILSPQFQYGSKAYVKHPADIPDYLKLSFPEGFKWERVMNFEDGGVVTVTQDSSLQDGEFIYKVKLRGTNFPSDGPVMQKKTMGWEASTERMYPEDGALKGEIKMRLKLKDGGHYDAEVKTTYMAKKPVQLPGAYKTDIKLDITSHNEDYTIVEQYERAEGRHSTGA*
AATTGACAATGTGAGCGAGTAACAAGATACTGAGCACAgagaccAAAGAGGAGAAAggtctcATGGCGAGTAGCGAAGACGTTATCAAAGAGTTCATGCGTTTCAAAGTTCGTATGGAAGGTTCCGTTAACGGTCACGAGTTCGAAATCGAAGGTGAAGGTGAAGGTCGTCCGTACGAAGGTACCCAGACCGCTAAACTGAAAGTTACCAAAGGTGGTCCGCTGCCGTTCGCTTGGGACATCCTGTCCCCGCAGTTCCAGTACGGTTCCAAAGCTTACGTTAAACACCCGGCTGACATCCCGGACTACCTGAAACTGTCCTTCCCGGAAGGTTTCAAATGGGAACGTGTTATGAACTTCGAAGACGGTGGTGTTGTTACCGTTACCCAGGACTCCTCCCTGCAAGACGGTGAGTTCATCTACAAAGTTAAACTGCGTGGTACCAACTTCCCGTCCGACGGTCCGGTTATGCAGAAAAAAACCATGGGTTGGGAAGCTTCCACCGAACGTATGTACCCGGAAGACGGTGCTCTGAAAGGTGAAATCAAAATGCGTCTGAAACTGAAAGACGGTGGTCACTACGACGCTGAAGTTAAAACCACCTACATGGCTAAAAAACCGGTTCAGCTGCCGGGTGCTTACAAAACCGACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTTGAACAGTACGAACGTGCTGAAGGTCGTCACTCCACCGGTGCTTAA
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/zchen/Public/python/lib/python3.11/site-packages/Bio/Seq.py:2804: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.
  warnings.warn(
/home/zchen/Public/python/lib/python3.11/site-packages/Bio/Seq.py:2804: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.
  warnings.warn(
/home/zchen/Public/python/lib/python3.11/site-packages/Bio/Seq.py:2804: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.
  warnings.warn(
/home/zchen/Public/python/lib/python3.11/site-packages/Bio/Seq.py:2804: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.
  warnings.warn(
</pre></div>
</div>
</div>
</div>
</section>
<section id="searching-for-open-reading-frames-a-class-anchor-id-gene-orf-search-a">
<h3>Searching for open reading frames  <a class="anchor" id="gene_ORF_search"></a><a class="headerlink" href="#searching-for-open-reading-frames-a-class-anchor-id-gene-orf-search-a" title="Permalink to this heading"></a></h3>
<p>Genbank records may sometimes contain unannotated genes. In these situations, we can perform a search of open reading frames in the sequence to help identify hidden genomic features of plasmid. In the following, we will load an annotated genbank record of a plasmid and delete the antibiotic resistance gene. We can search for the open reading frames with <code class="docutils literal notranslate"><span class="pre">Aligner.search_ORF</span></code> to recover some missing annotations.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">pge</span><span class="o">.</span><span class="n">Aligner</span><span class="o">.</span><span class="n">search_ORF</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Help on function search_ORF in module plasmid.aligner:

search_ORF(seq, table=&#39;Standard&#39;, fwd_only=False)
    Generates a dataframe of open reading frames for a given sequence
    seq = input sequence
    table = codon table to use
    return pandas Dataframe with columns
        [name, start, stop, orientation, amino acid sequence]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># load a genbank record and delete the antibiotic resistance gene</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pge</span><span class="o">.</span><span class="n">read_genbank</span><span class="p">(</span><span class="s1">&#39;../data/xRFP.gb&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="fm">__repr__</span><span class="p">())</span>
<span class="k">del</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">df</span><span class="p">[:</span><span class="mi">1000</span><span class="p">]</span><span class="o">.</span><span class="fm">__repr__</span><span class="p">())</span>
<span class="c1"># generate list of open reading frames</span>
<span class="n">seq</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">df</span><span class="p">[:</span><span class="mi">1000</span><span class="p">])</span>
<span class="n">x</span> <span class="o">=</span> <span class="n">pge</span><span class="o">.</span><span class="n">Aligner</span><span class="o">.</span><span class="n">search_ORF</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="c1"># compute length of the sequence</span>
<span class="n">x</span><span class="p">[</span><span class="s1">&#39;length&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;stop&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span>
<span class="c1"># get the top 5 longest sequences</span>
<span class="n">x</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;length&#39;</span><span class="p">,</span><span class="s1">&#39;orientation&#39;</span><span class="p">,</span><span class="s1">&#39;start&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">iloc</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="n">col</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">,</span><span class="s1">&#39;stop&#39;</span><span class="p">,</span><span class="s1">&#39;length&#39;</span><span class="p">,</span><span class="s1">&#39;sequence&#39;</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:</span><span class="mi">5</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>reading  ../data/xRFP.gb  as genbank file
&lt;class &#39;plasmid.plasmid.Plasmid&#39;&gt; at 0x7f445487f3d0
molecule_type:DNA
topology:circular
data_file_division:   
date:05-DEC-2022
accessions:[&#39;&lt;unknown&#39;, &#39;id&gt;&#39;]
keywords:[&#39;&#39;]
source:
organism:. .
taxonomy:[]
comment:
ApEinfo:methylated:1
      locus_tag type     location  length    color
0  SpectomycinR  CDS  [70:868](+)     798  #d40037
total length:4465

&lt;class &#39;plasmid.plasmid.Plasmid&#39;&gt; at 0x7f443d447690
molecule_type:DNA
topology:circular
data_file_division:   
date:05-DEC-2022
accessions:[&#39;&lt;unknown&#39;, &#39;id&gt;&#39;]
keywords:[&#39;&#39;]
source:
organism:. .
taxonomy:[]
comment:
ApEinfo:methylated:1
                   locus_tag        type      location  length    color
0  AmpR promoter subsequence    promoter     [0:29](+)      29  #0f7ffe
1       AmpR RBS subsequence         RBS    [29:70](+)      41     cyan
2      BBa_B0053 subsequence  terminator  [878:948](+)      70  #9d1b1c
total length:1000

  start stop length                                           sequence
2    70  868    798  MSEKVPAEISVQLSQALNVIGRHLESTLLAVHLYGSALDGGLKPYS...
4    45  357    312  MSLCTTMVTSKARSALFWPGEAETSKRSTSRACRTASLSGAATVTS...
5    45  339    294  MVTSKARSALFWPGEAETSKRSTSRACRTASLSGAATVTSKSISLY...
7   833  995    162  MKQLNCLVPANDNTSSGKKTGKVSPPCPFSLKPKRLLREIGVSRGR...
3   706  868    162  MARLPAQHQPILLNAKRAYLGQEEDYLPARADQVAALIKFVKYEAV...
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/zchen/Public/python/lib/python3.11/site-packages/Bio/Seq.py:2804: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.
  warnings.warn(
/home/zchen/Public/python/lib/python3.11/site-packages/Bio/Seq.py:2804: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.
  warnings.warn(
/home/zchen/Public/python/lib/python3.11/site-packages/Bio/Seq.py:2804: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.
  warnings.warn(
/home/zchen/Public/python/lib/python3.11/site-packages/Bio/Seq.py:2804: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.
  warnings.warn(
</pre></div>
</div>
</div>
</div>
<p>In the above, the longest sequence is likely the missing genes. We can search on NCBI blast to see if this matches any missing genes.</p>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../installation.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="primer_design.html" class="btn btn-neutral float-right" title="Designing primers with Plasmid" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Zhewei Chen.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>