<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Designing primers &mdash; Plasmid: A python library for gene editing 2023.06.29 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Aligning and clustering sequences" href="clustering.html" />
    <link rel="prev" title="Manipulating genbank records" href="genbank_editing.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            Plasmid: A python library for gene editing
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">How-to guides &amp; Examples</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="genbank_editing.html">Manipulating genbank records</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Designing primers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#extension-pcr">Extension PCR</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#running-extension-pcr">Running extension PCR</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#gibson-assembly">Gibson Assembly</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#running-gibson-assembly">Running gibson assembly</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#goldengate-assembly">Goldengate Assembly</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#running-golden-gate-assembly">Running golden gate assembly</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="clustering.html">Aligning and clustering sequences</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Plasmid: A python library for gene editing</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Designing primers</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/notebooks/primer_design.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="designing-primers">
<h1>Designing primers<a class="headerlink" href="#designing-primers" title="Permalink to this heading"></a></h1>
<p>This notebook illustrates the various functions contained in the <code class="docutils literal notranslate"><span class="pre">designer</span></code> class, which can be used to generate primers for extension PCR and DNA assembly.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">plasmid</span> <span class="k">as</span> <span class="nn">pge</span>
<span class="kn">import</span> <span class="nn">importlib</span>
<span class="n">importlib</span><span class="o">.</span><span class="n">reload</span><span class="p">(</span><span class="n">pge</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;module &#39;plasmid&#39; from &#39;/home/peguin/python/lib/python3.9/site-packages/plasmid/__init__.py&#39;&gt;
</pre></div>
</div>
</div>
</div>
<p><a class="anchor" id="xtPCR"></a></p>
<section id="extension-pcr">
<h2>Extension PCR<a class="headerlink" href="#extension-pcr" title="Permalink to this heading"></a></h2>
<p><a class="reference external" href="https://en.wikipedia.org/wiki/Overlap_extension_polymerase_chain_reaction">Extension PCR</a> uses primers with overlapping annealing sites to assemble a larger DNA fragment. This method is useful for adding promoters, ribosome binding sites, or cloning overhangs onto a larger gene fragment.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">Designer</span></code> class has the following parameters and functions associated with extension PCR.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pcr</span> <span class="o">=</span> <span class="n">pge</span><span class="o">.</span><span class="n">Designer</span><span class="p">()</span>
<span class="n">help</span><span class="p">(</span><span class="n">pcr</span><span class="o">.</span><span class="n">xtPCR</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;parameters for extension PCR&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">pcr</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;xtPCR&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Help on method xtPCR in module plasmid.designer:

xtPCR(fL, seq, fR=None, padding=[2, 2], niter=3, w=[10, 100, 1, 1, 2], get_cost=False) method of plasmid.designer.Designer instance
    Find primers which can seed and extend a PCR fragment
    fL = flanking sequence on 5&#39; end
    seq = sequence on 3&#39; end which gets amplified
    fR = flanking sequence on 3&#39; end
    padding = number of extra primers to try
    w = weights for cost function
    method = optimization method
    returns list of primers

parameters for extension PCR
{&#39;Tm&#39;: 50, &#39;len&#39;: [15, 60], &#39;nM&#39;: [20, 500], &#39;Na&#39;: 50}
</pre></div>
</div>
</div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">Tm</span></code> is the target melting temperature for all primers.</p>
<p><code class="docutils literal notranslate"><span class="pre">len</span></code> is min and max primer lengths allowed in the design.</p>
<p><code class="docutils literal notranslate"><span class="pre">nM</span></code> is the concentration of seed and finishing primers used in the reaction.</p>
<p><code class="docutils literal notranslate"><span class="pre">Na</span></code> is the salt concentration of the PCR reaction.</p>
<p>These parameters are used to compute the thermodynamic properties of the PCR reaction such as annealing temperature. In the backend, several optimization algorithms can be used to find the optimal primer sequences that minimize deviation from the target annealing temperature and off target binding.</p>
<p><a class="anchor" id="running_xtPCR"></a></p>
<section id="running-extension-pcr">
<h3>Running extension PCR<a class="headerlink" href="#running-extension-pcr" title="Permalink to this heading"></a></h3>
<p>The following code blocks generate a generic gene fragment from RFP, RBS, and pLac promoters. Extension PCR is used to add the pLac promoter and ribosome binding site sequences onto the RFP gene.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># slice out the RFP gene</span>
<span class="n">RFP</span> <span class="o">=</span> <span class="n">pge</span><span class="o">.</span><span class="n">read_genbank</span><span class="p">(</span><span class="s1">&#39;../data/dcas9_RFP.gb&#39;</span><span class="p">)</span>
<span class="n">RFP</span> <span class="o">=</span> <span class="n">RFP</span><span class="p">[</span><span class="n">RFP</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;mRFP&#39;</span><span class="p">)]</span><span class="o">.</span><span class="n">splice</span><span class="p">()</span>
<span class="c1"># slice out the ribosome binding site</span>
<span class="n">RBS</span> <span class="o">=</span> <span class="n">pge</span><span class="o">.</span><span class="n">read_genbank</span><span class="p">(</span><span class="s1">&#39;../data/xRFP.gb&#39;</span><span class="p">)</span>
<span class="n">RBS</span> <span class="o">=</span> <span class="n">RBS</span><span class="p">[</span><span class="n">RBS</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;BBa_B0034&#39;</span><span class="p">)]</span><span class="o">.</span><span class="n">splice</span><span class="p">()</span>
<span class="c1"># slice out the promoter</span>
<span class="n">pLac</span> <span class="o">=</span> <span class="n">pge</span><span class="o">.</span><span class="n">read_genbank</span><span class="p">(</span><span class="s1">&#39;../data/xRFP.gb&#39;</span><span class="p">)</span>
<span class="n">pLac</span> <span class="o">=</span> <span class="n">pLac</span><span class="p">[</span><span class="n">pLac</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;pLac&#39;</span><span class="p">)]</span><span class="o">.</span><span class="n">splice</span><span class="p">()</span>

<span class="c1"># assemble the promoter, rbs, and mRFP</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pLac</span> <span class="o">+</span> <span class="s1">&#39;gagacc&#39;</span> <span class="o">+</span> <span class="n">RBS</span> <span class="o">+</span> <span class="s1">&#39;ggtctc&#39;</span> <span class="o">+</span> <span class="n">RFP</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>reading  ../data/dcas9_RFP.gb  as genbank file
reading  ../data/xRFP.gb  as genbank file
reading  ../data/xRFP.gb  as genbank file
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pcr</span> <span class="o">=</span> <span class="n">pge</span><span class="o">.</span><span class="n">Designer</span><span class="p">()</span>
<span class="n">pcr</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;xtPCR&#39;</span><span class="p">][</span><span class="s1">&#39;Tm&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">55</span>         <span class="c1"># target annealing temperature for xtPCR</span>
<span class="n">pcr</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;xtPCR&#39;</span><span class="p">][</span><span class="s1">&#39;len&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">15</span><span class="p">,</span> <span class="mi">60</span><span class="p">]</span>  <span class="c1"># defines the [min, max] primer lengths</span>
<span class="n">pcr</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;xtPCR&#39;</span><span class="p">][</span><span class="s1">&#39;Nm&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">25</span><span class="p">,</span><span class="mi">500</span><span class="p">]</span>   <span class="c1"># defines the seed and finishing primer concentration in nM</span>
<span class="n">pcr</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;verbose&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">False</span>

<span class="n">insert</span> <span class="o">=</span> <span class="n">pLac</span> <span class="o">+</span> <span class="s1">&#39;gagacc&#39;</span> <span class="o">+</span> <span class="n">RBS</span> <span class="o">+</span> <span class="s1">&#39;ggtctc&#39;</span>
<span class="n">res</span> <span class="o">=</span> <span class="n">pcr</span><span class="o">.</span><span class="n">xtPCR</span><span class="p">(</span><span class="n">insert</span><span class="p">,</span> <span class="n">RFP</span><span class="p">,</span> <span class="s1">&#39; &#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>running fwd
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>running rev
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>  locus_tag         Tm                                           sequence  \
0       0_F  56.437456  ACAgagaccAAAGAGGAGAAAggtctc ATGGCGAGTAGCGAAGAC...   
1     fin_F  55.207985  AATTGACAATGTGAGCGAGTAACAAGATACTGAGC ACAgagaccA...   
0     fin_R  56.777386                                 TTAAGCACCGGTGGAGTG   

                   annealed  strand  
0  ATGGCGAGTAGCGAAGACGTTATC       1  
1     ACAgagaccAAAGAGGAGAAA       1  
0        TTAAGCACCGGTGGAGTG      -1  
</pre></div>
</div>
</div>
</div>
<p>In the above output dataframe, <code class="docutils literal notranslate"><span class="pre">fin_F</span></code> and <code class="docutils literal notranslate"><span class="pre">fin_R</span></code> are the finishing primers (used at 500nM), which are the last set of primers to amplify in the PCR reaction. <code class="docutils literal notranslate"><span class="pre">0_F</span></code> is a forward primer (used at 25nM) associated with seeding and extending the sequence from the 5’ end of the oligo.</p>
<p>In an extension PCR reaction, all these primers would be mixed together to generate the desired oligo at the target annealing tempertaure.</p>
<p><a class="anchor" id="gibson"></a></p>
</section>
</section>
<section id="gibson-assembly">
<h2>Gibson Assembly<a class="headerlink" href="#gibson-assembly" title="Permalink to this heading"></a></h2>
<p><a class="reference external" href="https://www.neb.com/applications/cloning-and-synthetic-biology/dna-assembly-and-cloning/nebuilder-hifi-dna-assembly">Gibson Assembly</a> uses exonucleases to reveal basepair overlaps between multiple DNA oligos. These overlaps can anneal and assemble into a longer DNA fragment. The following shows how to design primers for amplifying individual DNA fragments that can later be assembled into a longer construct.</p>
<p>In the design process, optimal overlaps are first found via optimization. Next <span class="xref myst">xtPCR</span> is used to design primers which can add the proper overlaps to the DNA oligos.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pcr</span> <span class="o">=</span> <span class="n">pge</span><span class="o">.</span><span class="n">Designer</span><span class="p">()</span>
<span class="n">help</span><span class="p">(</span><span class="n">pcr</span><span class="o">.</span><span class="n">Gibson</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">pcr</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;gibson&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Help on method Gibson in module plasmid.designer:

Gibson(seqlist, w=[10, 1], method=&#39;differential_evolution&#39;, circular=True) method of plasmid.designer.Designer instance
    Design primers for gibson assembly
    seqlist = list of sequences to assemble via gibson in order 
    circular = assemble fragments into a circular construct
    returns list of primers

{&#39;Tm&#39;: 50, &#39;nM&#39;: 500, &#39;len&#39;: 30, &#39;window&#39;: 40}
</pre></div>
</div>
</div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">Tm</span></code> is the target melting temperature of the overlaps.</p>
<p><code class="docutils literal notranslate"><span class="pre">nM</span></code> is the concentration of the oligos.</p>
<p><code class="docutils literal notranslate"><span class="pre">len</span></code> is the overlap length.</p>
<p><code class="docutils literal notranslate"><span class="pre">window</span></code> is the window of sequences from the fragment edges from which to check for optimal overlaps.</p>
<p>These parameters are used to compute the annealing temperature of the overlap such that the target annealing temperature can be reached and overlaps minimize off target binding.</p>
<section id="running-gibson-assembly">
<h3>Running gibson assembly<a class="headerlink" href="#running-gibson-assembly" title="Permalink to this heading"></a></h3>
<p>The following code blocks generate a new construct from LacI, RFP, and plasmid vector. Gibson is used to generate primers and oligos with optimal annealing overlaps.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">get_parts</span><span class="p">():</span>
    <span class="c1"># slice out the LacI gene</span>
    <span class="n">LacI</span> <span class="o">=</span> <span class="n">pge</span><span class="o">.</span><span class="n">read_genbank</span><span class="p">(</span><span class="s1">&#39;../data/xRFP.gb&#39;</span><span class="p">)</span>
    <span class="n">LacI</span> <span class="o">=</span> <span class="n">LacI</span><span class="p">[</span><span class="n">LacI</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;LacI&#39;</span><span class="p">)]</span><span class="o">.</span><span class="n">splice</span><span class="p">()</span>

    <span class="c1"># slice out the RFP gene</span>
    <span class="n">RFP</span> <span class="o">=</span> <span class="n">pge</span><span class="o">.</span><span class="n">read_genbank</span><span class="p">(</span><span class="s1">&#39;../data/dcas9_RFP.gb&#39;</span><span class="p">)</span>
    <span class="n">RFP</span> <span class="o">=</span> <span class="n">RFP</span><span class="p">[</span><span class="n">RFP</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;mRFP&#39;</span><span class="p">)]</span><span class="o">.</span><span class="n">splice</span><span class="p">()</span>

    <span class="c1"># slice out the origin of replication</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pge</span><span class="o">.</span><span class="n">read_genbank</span><span class="p">(</span><span class="s1">&#39;../data/xRFP.gb&#39;</span><span class="p">)</span>
    <span class="n">vec</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;pSC101&#39;</span><span class="p">)]</span>
    <span class="n">start</span> <span class="o">=</span> <span class="n">vec</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">stop</span> <span class="o">=</span> <span class="n">vec</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">vec</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">stop</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">LacI</span><span class="p">,</span> <span class="n">RFP</span><span class="p">,</span> <span class="n">vec</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Generate the parts which need to be combined</span>
<span class="n">LacI</span><span class="p">,</span> <span class="n">RFP</span><span class="p">,</span> <span class="n">vec</span> <span class="o">=</span> <span class="n">get_parts</span><span class="p">()</span>
<span class="n">seq</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">seq</span><span class="o">+=</span> <span class="p">[[</span><span class="s1">&#39; &#39;</span><span class="p">,</span><span class="n">LacI</span><span class="p">,</span><span class="s1">&#39;AAAActttt&#39;</span><span class="p">]]</span> <span class="c1"># add the LacI gene with 3&#39; flanking sequences</span>
<span class="n">seq</span><span class="o">+=</span> <span class="p">[[</span><span class="s1">&#39; &#39;</span><span class="p">,</span><span class="n">RFP</span><span class="p">,</span><span class="s1">&#39;CGCCctttt&#39;</span><span class="p">]]</span>  <span class="c1"># add the RFP gene with 3&#39; flanking sequences </span>
<span class="n">seq</span><span class="o">+=</span> <span class="p">[[</span><span class="s1">&#39;&#39;</span><span class="p">,</span><span class="n">vec</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">]]</span>  <span class="c1"># add the vector</span>

<span class="n">pcr</span> <span class="o">=</span> <span class="n">pge</span><span class="o">.</span><span class="n">Designer</span><span class="p">()</span>
<span class="n">pcr</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;gibson&#39;</span><span class="p">][</span><span class="s1">&#39;Tm&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">50</span>     <span class="c1"># target annealing temperature of gibson fragments    </span>
<span class="n">pcr</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;gibson&#39;</span><span class="p">][</span><span class="s1">&#39;window&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">30</span> <span class="c1"># +/i window in bp around frag edges to look for gibson overlap</span>
<span class="n">pcr</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;gibson&#39;</span><span class="p">][</span><span class="s1">&#39;len&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">20</span>    <span class="c1"># length of gibson overlap</span>

<span class="n">pcr</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;xtPCR&#39;</span><span class="p">][</span><span class="s1">&#39;Tm&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">55</span>         <span class="c1"># target annealing temperature for xtPCR</span>
<span class="n">pcr</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;xtPCR&#39;</span><span class="p">][</span><span class="s1">&#39;len&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">15</span><span class="p">,</span> <span class="mi">60</span><span class="p">]</span>  <span class="c1"># defines the [min, max] primer lengths</span>
<span class="n">pcr</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;xtPCR&#39;</span><span class="p">][</span><span class="s1">&#39;nM&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">20</span><span class="p">,</span> <span class="mi">500</span><span class="p">]</span>  <span class="c1"># defines the [seed, finisher] primer conc in nM</span>
<span class="n">pcr</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;verbose&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">False</span>

<span class="n">res</span> <span class="o">=</span> <span class="n">pcr</span><span class="o">.</span><span class="n">Gibson</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>reading  ../data/xRFP.gb  as genbank file
reading  ../data/dcas9_RFP.gb  as genbank file
reading  ../data/xRFP.gb  as genbank file
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>res.x [11.51496424  1.67750846 15.96703809]
res.fun -54.0
exclude: []
overlaps: [&#39;CGGGCAGTAAAAAActttt &#39;, &#39;GTCACTCCACCGGTGCTTAA&#39;, &#39;GATCGACAATGTAAC ATGG&#39;]
Tm overlap: [45.828042650749126, 53.47392347753578, 45.13818370718934]
processing primers for frag 0
running fwd
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>running rev
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>processing primers for frag 1
running fwd
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>running rev
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>parasail_sg_flags_scan_avx2_256_16: s2Len must be &gt; 0
</pre></div>
</div>
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">ValueError</span><span class="g g-Whitespace">                                </span>Traceback (most recent call last)
<span class="nn">~/python/lib/python3.9/site-packages/scipy/optimize/_differentialevolution.py</span> in <span class="ni">?</span><span class="nt">(self, population)</span>
<span class="g g-Whitespace">   </span><span class="mi">1151</span>             <span class="c1"># wrong number of arguments for _mapwrapper</span>
<span class="g g-Whitespace">   </span><span class="mi">1152</span>             <span class="c1"># or wrong length returned from the mapper</span>
<span class="ne">-&gt; </span><span class="mi">1153</span>             <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1154</span>                 <span class="s2">&quot;The map-like callable must be of the form f(func, iterable), &quot;</span>

<span class="nn">~/python/lib/python3.9/site-packages/scipy/_lib/_util.py</span> in <span class="ni">?</span><span class="nt">(self, x)</span>
<span class="g g-Whitespace">    </span><span class="mi">359</span>     <span class="k">def</span> <span class="fm">__call__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">x</span><span class="p">):</span>
<span class="ne">--&gt; </span><span class="mi">360</span>         <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">f</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="o">*</span><span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="p">)</span>

<span class="nn">~/python/lib/python3.9/site-packages/plasmid/designer.py</span> in <span class="ni">?</span><span class="nt">(self, pos, fL, seq, w, verbose)</span>
<span class="g g-Whitespace">    </span><span class="mi">233</span>         <span class="c1"># cross reactivity</span>
<span class="g g-Whitespace">    </span><span class="mi">234</span>         <span class="k">if</span> <span class="n">w</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">!=</span><span class="mi">0</span><span class="p">:</span>
<span class="ne">--&gt; </span><span class="mi">235</span>             <span class="n">cost</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">xtPCR_crosscheck</span><span class="p">(</span><span class="n">pos</span><span class="p">,</span> <span class="n">fL</span><span class="p">,</span> <span class="n">seq</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">236</span>         <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>

<span class="nn">~/python/lib/python3.9/site-packages/plasmid/designer.py</span> in <span class="ni">?</span><span class="nt">(self, pos, fL, seq)</span>
<span class="g g-Whitespace">    </span><span class="mi">296</span>             <span class="c1"># score cross reactivity</span>
<span class="g g-Whitespace">    </span><span class="mi">297</span>             <span class="n">res1</span> <span class="o">=</span> <span class="n">parasail</span><span class="o">.</span><span class="n">sg_scan_16</span><span class="p">(</span><span class="n">annealed</span><span class="p">,</span> <span class="n">annealed</span><span class="p">,</span> <span class="n">g1</span><span class="p">,</span> <span class="n">g2</span><span class="p">,</span> <span class="n">parasail</span><span class="o">.</span><span class="n">nuc44</span><span class="p">)</span>
<span class="ne">--&gt; </span><span class="mi">298</span>             <span class="n">score</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">res1</span><span class="o">.</span><span class="n">score</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_reactivity</span><span class="p">(</span><span class="n">annealed</span><span class="p">,</span> <span class="n">exclude</span><span class="p">,</span> <span class="n">res1</span><span class="o">.</span><span class="n">score</span><span class="p">))</span>
<span class="g g-Whitespace">    </span><span class="mi">299</span> 

<span class="nn">~/python/lib/python3.9/site-packages/plasmid/designer.py</span> in <span class="ni">?</span><span class="nt">(self, seq, seqlist, best)</span>
<span class="g g-Whitespace">    </span><span class="mi">548</span>         <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seqlist</span><span class="p">)):</span>
<span class="g g-Whitespace">    </span><span class="mi">549</span>             <span class="n">res2</span> <span class="o">=</span> <span class="n">parasail</span><span class="o">.</span><span class="n">sg_scan_16</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">seqlist</span><span class="p">[</span><span class="n">j</span><span class="p">],</span> <span class="n">g1</span><span class="p">,</span> <span class="n">g2</span><span class="p">,</span> <span class="n">parasail</span><span class="o">.</span><span class="n">nuc44</span><span class="p">)</span>
<span class="ne">--&gt; </span><span class="mi">550</span>             <span class="n">score</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">([</span><span class="n">score</span><span class="p">,</span> <span class="n">res2</span><span class="o">.</span><span class="n">score</span><span class="p">])</span>
<span class="g g-Whitespace">    </span><span class="mi">551</span>             <span class="k">if</span> <span class="n">best</span><span class="o">!=</span><span class="kc">None</span> <span class="ow">and</span> <span class="n">best</span><span class="o">==</span><span class="n">res2</span><span class="o">.</span><span class="n">score</span><span class="p">:</span>

<span class="nn">~/python/lib/python3.9/site-packages/parasail/bindings_v2.py</span> in <span class="ni">?</span><span class="nt">(self)</span>
<span class="g g-Whitespace">    </span><span class="mi">279</span>     <span class="nd">@property</span>
<span class="g g-Whitespace">    </span><span class="mi">280</span>     <span class="k">def</span> <span class="nf">score</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="ne">--&gt; </span><span class="mi">281</span>         <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">pointer</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">score</span>

<span class="ne">ValueError</span>: NULL pointer access

<span class="n">The</span> <span class="n">above</span> <span class="n">exception</span> <span class="n">was</span> <span class="n">the</span> <span class="n">direct</span> <span class="n">cause</span> <span class="n">of</span> <span class="n">the</span> <span class="n">following</span> <span class="n">exception</span><span class="p">:</span>

<span class="ne">RuntimeError</span><span class="g g-Whitespace">                              </span>Traceback (most recent call last)
<span class="n">Cell</span> <span class="n">In</span><span class="p">[</span><span class="mi">7</span><span class="p">],</span> <span class="n">line</span> <span class="mi">18</span>
<span class="g g-Whitespace">     </span><span class="mi">15</span> <span class="n">pcr</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;xtPCR&#39;</span><span class="p">][</span><span class="s1">&#39;nM&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">20</span><span class="p">,</span> <span class="mi">500</span><span class="p">]</span>  <span class="c1"># defines the [seed, finisher] primer conc in nM</span>
<span class="g g-Whitespace">     </span><span class="mi">16</span> <span class="n">pcr</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;verbose&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">False</span>
<span class="ne">---&gt; </span><span class="mi">18</span> <span class="n">res</span> <span class="o">=</span> <span class="n">pcr</span><span class="o">.</span><span class="n">Gibson</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="g g-Whitespace">     </span><span class="mi">19</span> <span class="nb">print</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>

<span class="nn">File ~/python/lib/python3.9/site-packages/plasmid/designer.py:568,</span> in <span class="ni">Designer.Gibson</span><span class="nt">(self, seqlist, w, method, circular)</span>
<span class="g g-Whitespace">    </span><span class="mi">566</span> <span class="bp">self</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;goldengate&#39;</span><span class="p">][</span><span class="s1">&#39;ggsite&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<span class="g g-Whitespace">    </span><span class="mi">567</span> <span class="bp">self</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;goldengate&#39;</span><span class="p">][</span><span class="s1">&#39;padding&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<span class="ne">--&gt; </span><span class="mi">568</span> <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">GoldenGate</span><span class="p">(</span><span class="n">seqlist</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="n">circular</span><span class="p">)</span>

<span class="nn">File ~/python/lib/python3.9/site-packages/plasmid/designer.py:489,</span> in <span class="ni">Designer.GoldenGate</span><span class="nt">(self, seqlist, exclude, w, circular)</span>
<span class="g g-Whitespace">    </span><span class="mi">487</span> <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;processing primers for frag&#39;</span><span class="p">,</span><span class="n">i</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">488</span> <span class="p">[</span><span class="n">fL</span><span class="p">,</span> <span class="n">seq</span><span class="p">,</span> <span class="n">fR</span><span class="p">]</span> <span class="o">=</span> <span class="n">frags</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
<span class="ne">--&gt; </span><span class="mi">489</span> <span class="n">p</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">xtPCR</span><span class="p">(</span><span class="n">fL</span><span class="p">,</span> <span class="n">seq</span><span class="p">,</span> <span class="n">fR</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">490</span> <span class="n">p</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;frag&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;_&#39;</span><span class="o">+</span><span class="n">j</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">p</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">]]</span>
<span class="g g-Whitespace">    </span><span class="mi">491</span> <span class="n">out</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>

<span class="nn">File ~/python/lib/python3.9/site-packages/plasmid/designer.py:190,</span> in <span class="ni">Designer.xtPCR</span><span class="nt">(self, fL, seq, fR, padding, niter, w, get_cost)</span>
<span class="g g-Whitespace">    </span><span class="mi">188</span> <span class="n">fwd</span> <span class="o">=</span> <span class="n">df_F</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="g g-Whitespace">    </span><span class="mi">189</span> <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;running rev&#39;</span><span class="p">)</span>
<span class="ne">--&gt; </span><span class="mi">190</span> <span class="n">df_R</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">xtPCR</span><span class="p">(</span><span class="n">fR</span><span class="p">,</span> <span class="n">seqR</span><span class="p">,</span> <span class="n">padding</span><span class="o">=</span><span class="n">padding</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">w</span><span class="o">=</span><span class="n">w</span><span class="p">,</span> <span class="n">get_cost</span><span class="o">=</span><span class="n">get_cost</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">191</span> <span class="n">rev</span> <span class="o">=</span> <span class="n">df_R</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="g g-Whitespace">    </span><span class="mi">192</span> <span class="n">fwd</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;_F&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">fwd</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">)]</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;fin_F&#39;</span><span class="p">]</span>

<span class="nn">File ~/python/lib/python3.9/site-packages/plasmid/designer.py:158,</span> in <span class="ni">Designer.xtPCR</span><span class="nt">(self, fL, seq, fR, padding, niter, w, get_cost)</span>
<span class="g g-Whitespace">    </span><span class="mi">156</span> <span class="n">bounds</span> <span class="o">=</span> <span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;xtPCR&#39;</span><span class="p">][</span><span class="s1">&#39;len&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;xtPCR&#39;</span><span class="p">][</span><span class="s1">&#39;len&#39;</span><span class="p">][</span><span class="mi">1</span><span class="p">])]</span><span class="o">*</span><span class="n">N_left</span> <span class="o">+</span> <span class="p">[(</span><span class="mi">0</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;xtPCR&#39;</span><span class="p">][</span><span class="s1">&#39;len&#39;</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;xtPCR&#39;</span><span class="p">][</span><span class="s1">&#39;len&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">])]</span><span class="o">*</span><span class="n">N_left</span>
<span class="g g-Whitespace">    </span><span class="mi">157</span> <span class="n">x0</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;xtPCR&#39;</span><span class="p">][</span><span class="s1">&#39;len&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span><span class="o">*</span><span class="n">N_left</span> <span class="o">+</span> <span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">*</span><span class="n">N_left</span>
<span class="ne">--&gt; </span><span class="mi">158</span> <span class="n">res</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">optimizer</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">xtPCR_cost</span><span class="p">,</span> <span class="n">args</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">x0</span><span class="p">)</span> 
<span class="g g-Whitespace">    </span><span class="mi">159</span> <span class="c1"># save the best result</span>
<span class="g g-Whitespace">    </span><span class="mi">160</span> <span class="k">if</span> <span class="n">best_res</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">res</span><span class="o">.</span><span class="n">fun</span><span class="p">:</span>

<span class="nn">File ~/python/lib/python3.9/site-packages/plasmid/designer.py:93,</span> in <span class="ni">Designer.optimizer</span><span class="nt">(self, cost_func, args, bounds, x0, discrete)</span>
<span class="g g-Whitespace">     </span><span class="mi">91</span>     <span class="k">return</span> <span class="n">sp</span><span class="o">.</span><span class="n">optimize</span><span class="o">.</span><span class="n">basinhopping</span><span class="p">(</span><span class="n">cost_func</span><span class="p">,</span> <span class="n">x0</span><span class="p">,</span> <span class="n">minimizer_kwargs</span><span class="o">=</span><span class="n">kwargs</span><span class="p">)</span>
<span class="g g-Whitespace">     </span><span class="mi">92</span> <span class="k">elif</span> <span class="n">method</span><span class="o">==</span><span class="s1">&#39;differential_evolution&#39;</span><span class="p">:</span>
<span class="ne">---&gt; </span><span class="mi">93</span>     <span class="k">return</span> <span class="n">sp</span><span class="o">.</span><span class="n">optimize</span><span class="o">.</span><span class="n">differential_evolution</span><span class="p">(</span><span class="n">cost_func</span><span class="p">,</span> <span class="n">bounds</span><span class="o">=</span><span class="n">bounds</span><span class="p">,</span> <span class="n">args</span><span class="o">=</span><span class="n">args</span><span class="p">,</span> <span class="n">constraints</span><span class="o">=</span><span class="n">cons</span><span class="p">)</span>

<span class="nn">File ~/python/lib/python3.9/site-packages/scipy/optimize/_differentialevolution.py:402,</span> in <span class="ni">differential_evolution</span><span class="nt">(func, bounds, args, strategy, maxiter, popsize, tol, mutation, recombination, seed, callback, disp, polish, init, atol, updating, workers, constraints, x0, integrality, vectorized)</span>
<span class="g g-Whitespace">    </span><span class="mi">385</span> <span class="c1"># using a context manager means that any created Pool objects are</span>
<span class="g g-Whitespace">    </span><span class="mi">386</span> <span class="c1"># cleared up.</span>
<span class="g g-Whitespace">    </span><span class="mi">387</span> <span class="k">with</span> <span class="n">DifferentialEvolutionSolver</span><span class="p">(</span><span class="n">func</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">args</span><span class="o">=</span><span class="n">args</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">388</span>                                  <span class="n">strategy</span><span class="o">=</span><span class="n">strategy</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">389</span>                                  <span class="n">maxiter</span><span class="o">=</span><span class="n">maxiter</span><span class="p">,</span>
   <span class="p">(</span><span class="o">...</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">400</span>                                  <span class="n">integrality</span><span class="o">=</span><span class="n">integrality</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">401</span>                                  <span class="n">vectorized</span><span class="o">=</span><span class="n">vectorized</span><span class="p">)</span> <span class="k">as</span> <span class="n">solver</span><span class="p">:</span>
<span class="ne">--&gt; </span><span class="mi">402</span>     <span class="n">ret</span> <span class="o">=</span> <span class="n">solver</span><span class="o">.</span><span class="n">solve</span><span class="p">()</span>
<span class="g g-Whitespace">    </span><span class="mi">404</span> <span class="k">return</span> <span class="n">ret</span>

<span class="nn">File ~/python/lib/python3.9/site-packages/scipy/optimize/_differentialevolution.py:1013,</span> in <span class="ni">DifferentialEvolutionSolver.solve</span><span class="nt">(self)</span>
<span class="g g-Whitespace">   </span><span class="mi">1008</span>     <span class="bp">self</span><span class="o">.</span><span class="n">feasible</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">constraint_violation</span> <span class="o">=</span> <span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1009</span>         <span class="bp">self</span><span class="o">.</span><span class="n">_calculate_population_feasibilities</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">population</span><span class="p">))</span>
<span class="g g-Whitespace">   </span><span class="mi">1011</span>     <span class="c1"># only work out population energies for feasible solutions</span>
<span class="g g-Whitespace">   </span><span class="mi">1012</span>     <span class="bp">self</span><span class="o">.</span><span class="n">population_energies</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">feasible</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
<span class="ne">-&gt; </span><span class="mi">1013</span>         <span class="bp">self</span><span class="o">.</span><span class="n">_calculate_population_energies</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1014</span>             <span class="bp">self</span><span class="o">.</span><span class="n">population</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">feasible</span><span class="p">]))</span>
<span class="g g-Whitespace">   </span><span class="mi">1016</span>     <span class="bp">self</span><span class="o">.</span><span class="n">_promote_lowest_energy</span><span class="p">()</span>
<span class="g g-Whitespace">   </span><span class="mi">1018</span> <span class="c1"># do the optimization.</span>

<span class="nn">File ~/python/lib/python3.9/site-packages/scipy/optimize/_differentialevolution.py:1153,</span> in <span class="ni">DifferentialEvolutionSolver._calculate_population_energies</span><span class="nt">(self, population)</span>
<span class="g g-Whitespace">   </span><span class="mi">1149</span>     <span class="n">calc_energies</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">squeeze</span><span class="p">(</span><span class="n">calc_energies</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1150</span> <span class="k">except</span> <span class="p">(</span><span class="ne">TypeError</span><span class="p">,</span> <span class="ne">ValueError</span><span class="p">)</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
<span class="g g-Whitespace">   </span><span class="mi">1151</span>     <span class="c1"># wrong number of arguments for _mapwrapper</span>
<span class="g g-Whitespace">   </span><span class="mi">1152</span>     <span class="c1"># or wrong length returned from the mapper</span>
<span class="ne">-&gt; </span><span class="mi">1153</span>     <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1154</span>         <span class="s2">&quot;The map-like callable must be of the form f(func, iterable), &quot;</span>
<span class="g g-Whitespace">   </span><span class="mi">1155</span>         <span class="s2">&quot;returning a sequence of numbers the same length as &#39;iterable&#39;&quot;</span>
<span class="g g-Whitespace">   </span><span class="mi">1156</span>     <span class="p">)</span> <span class="kn">from</span> <span class="nn">e</span>
<span class="g g-Whitespace">   </span><span class="mi">1158</span> <span class="k">if</span> <span class="n">calc_energies</span><span class="o">.</span><span class="n">size</span> <span class="o">!=</span> <span class="n">S</span><span class="p">:</span>
<span class="g g-Whitespace">   </span><span class="mi">1159</span>     <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">vectorized</span><span class="p">:</span>

<span class="ne">RuntimeError</span>: The map-like callable must be of the form f(func, iterable), returning a sequence of numbers the same length as &#39;iterable&#39;
</pre></div>
</div>
</div>
</div>
<p>In the above output, <code class="docutils literal notranslate"><span class="pre">fin_F</span></code> and <code class="docutils literal notranslate"><span class="pre">fin_R</span></code> are the finishing primers (used at 500nM) for each gibson fragment. These are used independently in their own extension PCR reaction. After amplification, the PCR products are combined with gibson reaction mix to join the DNA oligos.</p>
<p>seq0, seq1, and seq2 are full length fragments with the overlaps which can be order as gene fragments from vendors like Twist or IDT.</p>
<p><a class="anchor" id="ggate"></a></p>
</section>
</section>
<section id="goldengate-assembly">
<h2>Goldengate Assembly<a class="headerlink" href="#goldengate-assembly" title="Permalink to this heading"></a></h2>
<p><a class="reference external" href="https://www.neb.com/applications/cloning-and-synthetic-biology/dna-assembly-and-cloning/golden-gate-assembly">Goldengate</a> assembly is another method of DNA assembly which relies on restriction enzymes to generate overlaps that can be annealed and ligated. This assembly process is more efficient than gibson assembly, but requires that the DNA oligos do not contain the restriction enzyme sites. Only the edges of the DNA fragments can contain the restriction sites.</p>
<p>The goal of primer design is to find a set of primers which can artificially add restriction enzyme sites and overlaps to a set of DNA oligos that need to be assembled.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pcr</span> <span class="o">=</span> <span class="n">pge</span><span class="o">.</span><span class="n">Designer</span><span class="p">()</span>
<span class="n">help</span><span class="p">(</span><span class="n">pcr</span><span class="o">.</span><span class="n">GoldenGate</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">pcr</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;goldengate&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">Tm</span></code> is the target melting temperature of the overlaps.</p>
<p><code class="docutils literal notranslate"><span class="pre">nM</span></code> is the concentration of the oligos or DNA fragments.</p>
<p><code class="docutils literal notranslate"><span class="pre">padding</span></code> the extra nucleotide sequence that needs to be added to the 5’ and 3’ ends of the oligo after the restriction site so the enzymes can cut DNA properly.</p>
<p><code class="docutils literal notranslate"><span class="pre">ggsite</span></code> is the restriction enzyme to use.</p>
<p><code class="docutils literal notranslate"><span class="pre">ggN</span></code> is the overlap length generated by the restriction enzyme.</p>
<p><code class="docutils literal notranslate"><span class="pre">window</span></code> is the window of sequences from the fragment edges from which to check for optimal overlaps.</p>
<p>These parameters are used to compute the annealing temperature of the overlap and primers such that oligos with the proper goldengate sites and overlaps can be generated for cloning.</p>
<section id="running-golden-gate-assembly">
<h3>Running golden gate assembly<a class="headerlink" href="#running-golden-gate-assembly" title="Permalink to this heading"></a></h3>
<p>The following code blocks generate a new construct from LacI, RFP, and plasmid vector. Goldengate is used to generate primers that add goldengate sites to the flanks of the oligos. Optimization works by first finding optimal overlaps for annealing the oligos. xtPCR is then used to add the proper goldengate sites to the oligos.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">LacI</span><span class="p">,</span> <span class="n">RFP</span><span class="p">,</span> <span class="n">vec</span> <span class="o">=</span> <span class="n">get_parts</span><span class="p">()</span>
<span class="n">seq</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">seq</span><span class="o">+=</span> <span class="p">[[</span><span class="s1">&#39;&#39;</span><span class="p">,</span><span class="n">LacI</span><span class="p">,</span><span class="s1">&#39;AAAActttt&#39;</span><span class="p">]]</span>
<span class="n">seq</span><span class="o">+=</span> <span class="p">[[</span><span class="s1">&#39;&#39;</span><span class="p">,</span><span class="n">RFP</span><span class="p">,</span><span class="s1">&#39;CGCCctttt&#39;</span><span class="p">]]</span>
<span class="n">seq</span><span class="o">+=</span> <span class="p">[[</span><span class="s1">&#39;&#39;</span><span class="p">,</span><span class="n">vec</span><span class="p">,</span><span class="s1">&#39;GGGGctttt&#39;</span><span class="p">]]</span>

<span class="n">pcr</span> <span class="o">=</span> <span class="n">pge</span><span class="o">.</span><span class="n">Designer</span><span class="p">()</span>
<span class="n">pcr</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;goldengate&#39;</span><span class="p">][</span><span class="s1">&#39;window&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">20</span> <span class="c1"># +/i window in bp around frag edges to look for overlap</span>
<span class="n">pcr</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;goldengate&#39;</span><span class="p">][</span><span class="s1">&#39;ggN&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">4</span>     <span class="c1"># length of golden gate overlap</span>
<span class="n">pcr</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;goldengate&#39;</span><span class="p">][</span><span class="s1">&#39;ggsite&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;GGTCTCc&#39;</span>     <span class="c1"># golden gate enzyme site</span>
<span class="n">pcr</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;goldengate&#39;</span><span class="p">][</span><span class="s1">&#39;padding&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;atatatatgg&#39;</span> <span class="c1"># padding around the golden gate site</span>
<span class="n">pcr</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;xtPCR&#39;</span><span class="p">][</span><span class="s1">&#39;len&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">15</span><span class="p">,</span> <span class="mi">60</span><span class="p">]</span>  <span class="c1"># defines the [min, max] primer lengths</span>
<span class="n">pcr</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;xtPCR&#39;</span><span class="p">][</span><span class="s1">&#39;nM&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">20</span><span class="p">,</span> <span class="mi">500</span><span class="p">]</span> <span class="c1"># defines the [seed, finisher] primer conc in nM</span>
<span class="n">pcr</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;xtPCR&#39;</span><span class="p">][</span><span class="s1">&#39;Tm&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">55</span> <span class="c1"># defines the [seed, finisher] primer conc in nM</span>

<span class="n">res</span> <span class="o">=</span> <span class="n">pcr</span><span class="o">.</span><span class="n">GoldenGate</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>reading  ../data/xRFP.gb  as genbank file
reading  ../data/dcas9_RFP.gb  as genbank file
reading  ../data/xRFP.gb  as genbank file
res.x [10.34275508  4.10592271  9.27857462]
res.fun -12.0
exclude: []
overlaps: [&#39;AAAA&#39;, &#39;TGCT&#39;, &#39;ACGG&#39;]
Tm overlap: [-79.98814282127228, -58.96562554863294, -46.79908254371483]
processing primers for frag 0
running fwd
running rev
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/zchen/Public/python/lib/python3.11/site-packages/scipy/optimize/_numdiff.py:576: RuntimeWarning: invalid value encountered in subtract
  df = fun(x) - f0
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>processing primers for frag 1
running fwd
running rev
processing primers for frag 2
running fwd
running rev
     locus_tag         Tm                                           sequence   
0  frag0_fin_F  55.851352  atatatatggGGTCTCcACGGGGctttt ATGGTGAATGTGAAACC...  \
1  frag0_fin_R  56.106442             atatatatggGGTCTCcTTT TTACTGCCCGCTTTCCA   
2  frag1_fin_F  55.335316      atatatatggGGTCTCcAAAAActttt ATGGCGAGTAGCGAAGA   
3  frag1_fin_R  56.119961                  atatatatggGGTCTCc AGCACCGGTGGAGTG   
4  frag2_fin_F  55.363272  atatatatggGGTCTCcTGCTTAACGCCctttt CTGTCAGACCAA...   
5  frag2_fin_R  54.627626        atatatatggGGTCTCcCC GTTACATTGTCGATCTGTTCATG   
6         seq0        NaN  atatatatggGGTCTCcACGGGGcttttATGGTGAATGTGAAACCA...   
7         seq1        NaN  atatatatggGGTCTCcAAAAActtttATGGCGAGTAGCGAAGACG...   
8         seq2        NaN  atatatatggGGTCTCcTGCTTAACGCCcttttCTGTCAGACCAAG...   

                  annealed  strand  
0  ATGGTGAATGTGAAACCAGTAAC     1.0  
1        TTACTGCCCGCTTTCCA    -1.0  
2        ATGGCGAGTAGCGAAGA     1.0  
3          AGCACCGGTGGAGTG    -1.0  
4    CTGTCAGACCAAGTTTACGAG     1.0  
5  GTTACATTGTCGATCTGTTCATG    -1.0  
6                      NaN     NaN  
7                      NaN     NaN  
8                      NaN     NaN  
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/zchen/Public/python/lib/python3.11/site-packages/scipy/optimize/_numdiff.py:576: RuntimeWarning: invalid value encountered in subtract
  df = fun(x) - f0
</pre></div>
</div>
</div>
</div>
<p>In the above output, <code class="docutils literal notranslate"><span class="pre">fin_F</span></code> and <code class="docutils literal notranslate"><span class="pre">fin_R</span></code> are the finishing primers (used at 500nM) for each goldengate fragment. These primers add BsaI sites to each gene fragment. These primers are used independently in their own extension PCR reaction. After amplification, the PCR products are combined with the golden gate enzyme and ligase to join the DNA oligos.</p>
<p>seq0, seq1, and seq2 are full length fragments with the overlaps which can be order as gene fragments from vendors like Twist or IDT.</p>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="genbank_editing.html" class="btn btn-neutral float-left" title="Manipulating genbank records" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="clustering.html" class="btn btn-neutral float-right" title="Aligning and clustering sequences" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Zhewei Chen.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>