<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>plasmid.misc &mdash; Plasmid: A python library for gene editing 2023.06.29 documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/sphinx_highlight.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            Plasmid: A python library for gene editing
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <!-- Local TOC -->
              <div class="local-toc"></div>
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">Plasmid: A python library for gene editing</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">Module code</a></li>
      <li class="breadcrumb-item active">plasmid.misc</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for plasmid.misc</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python</span>
<span class="c1"># This is a python library for editing and plotting genbank files</span>

<span class="c1"># computing libraries</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="c1"># biopython imports</span>
<span class="kn">import</span> <span class="nn">Bio</span>
<span class="kn">import</span> <span class="nn">Bio.Seq</span>
<span class="kn">import</span> <span class="nn">Bio.SeqIO</span>
<span class="kn">import</span> <span class="nn">json</span>

<span class="c1"># system imports</span>
<span class="kn">import</span> <span class="nn">datetime</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">os</span>

<div class="viewcode-block" id="get_codons"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.get_codons">[docs]</a><span class="k">def</span> <span class="nf">get_codons</span><span class="p">():</span>
    <span class="c1"># generate codon table</span>
    <span class="n">codons</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;I&#39;</span><span class="p">:[</span><span class="s1">&#39;ATT&#39;</span><span class="p">,</span> <span class="s1">&#39;ATC&#39;</span><span class="p">,</span> <span class="s1">&#39;ATA&#39;</span><span class="p">],</span>
              <span class="s1">&#39;L&#39;</span><span class="p">:[</span><span class="s1">&#39;CTT&#39;</span><span class="p">,</span> <span class="s1">&#39;CTC&#39;</span><span class="p">,</span> <span class="s1">&#39;CTA&#39;</span><span class="p">,</span> <span class="s1">&#39;CTG&#39;</span><span class="p">,</span> <span class="s1">&#39;TTA&#39;</span><span class="p">,</span> <span class="s1">&#39;TTG&#39;</span><span class="p">],</span>
              <span class="s1">&#39;V&#39;</span><span class="p">:[</span><span class="s1">&#39;GTT&#39;</span><span class="p">,</span> <span class="s1">&#39;GTC&#39;</span><span class="p">,</span> <span class="s1">&#39;GTA&#39;</span><span class="p">,</span> <span class="s1">&#39;GTG&#39;</span><span class="p">],</span>
              <span class="s1">&#39;F&#39;</span><span class="p">:[</span><span class="s1">&#39;TTT&#39;</span><span class="p">,</span> <span class="s1">&#39;TTC&#39;</span><span class="p">],</span>
              <span class="s1">&#39;M&#39;</span><span class="p">:[</span><span class="s1">&#39;ATG&#39;</span><span class="p">],</span>
              <span class="s1">&#39;C&#39;</span><span class="p">:[</span><span class="s1">&#39;TGT&#39;</span><span class="p">,</span> <span class="s1">&#39;TGC&#39;</span><span class="p">],</span>
              <span class="s1">&#39;A&#39;</span><span class="p">:[</span><span class="s1">&#39;GCT&#39;</span><span class="p">,</span> <span class="s1">&#39;GCC&#39;</span><span class="p">,</span> <span class="s1">&#39;GCA&#39;</span><span class="p">,</span> <span class="s1">&#39;GCG&#39;</span><span class="p">],</span>
              <span class="s1">&#39;G&#39;</span><span class="p">:[</span><span class="s1">&#39;GGT&#39;</span><span class="p">,</span> <span class="s1">&#39;GGC&#39;</span><span class="p">,</span> <span class="s1">&#39;GGA&#39;</span><span class="p">,</span> <span class="s1">&#39;GGG&#39;</span><span class="p">],</span>
              <span class="s1">&#39;P&#39;</span><span class="p">:[</span><span class="s1">&#39;CCT&#39;</span><span class="p">,</span> <span class="s1">&#39;CCC&#39;</span><span class="p">,</span> <span class="s1">&#39;CCA&#39;</span><span class="p">,</span> <span class="s1">&#39;CCG&#39;</span><span class="p">],</span>
              <span class="s1">&#39;T&#39;</span><span class="p">:[</span><span class="s1">&#39;ACT&#39;</span><span class="p">,</span> <span class="s1">&#39;ACC&#39;</span><span class="p">,</span> <span class="s1">&#39;ACA&#39;</span><span class="p">,</span> <span class="s1">&#39;ACG&#39;</span><span class="p">],</span>
              <span class="s1">&#39;S&#39;</span><span class="p">:[</span><span class="s1">&#39;TCT&#39;</span><span class="p">,</span> <span class="s1">&#39;TCC&#39;</span><span class="p">,</span> <span class="s1">&#39;TCA&#39;</span><span class="p">,</span> <span class="s1">&#39;TCG&#39;</span><span class="p">,</span> <span class="s1">&#39;AGT&#39;</span><span class="p">,</span> <span class="s1">&#39;AGC&#39;</span><span class="p">],</span>
              <span class="s1">&#39;Y&#39;</span><span class="p">:[</span><span class="s1">&#39;TAT&#39;</span><span class="p">,</span> <span class="s1">&#39;TAC&#39;</span><span class="p">],</span>
              <span class="s1">&#39;W&#39;</span><span class="p">:[</span><span class="s1">&#39;TGG&#39;</span><span class="p">],</span>
              <span class="s1">&#39;Q&#39;</span><span class="p">:[</span><span class="s1">&#39;CAA&#39;</span><span class="p">,</span> <span class="s1">&#39;CAG&#39;</span><span class="p">],</span>
              <span class="s1">&#39;N&#39;</span><span class="p">:[</span><span class="s1">&#39;AAT&#39;</span><span class="p">,</span> <span class="s1">&#39;AAC&#39;</span><span class="p">],</span>
              <span class="s1">&#39;H&#39;</span><span class="p">:[</span><span class="s1">&#39;CAT&#39;</span><span class="p">,</span> <span class="s1">&#39;CAC&#39;</span><span class="p">],</span>
              <span class="s1">&#39;E&#39;</span><span class="p">:[</span><span class="s1">&#39;GAA&#39;</span><span class="p">,</span> <span class="s1">&#39;GAG&#39;</span><span class="p">],</span>
              <span class="s1">&#39;D&#39;</span><span class="p">:[</span><span class="s1">&#39;GAT&#39;</span><span class="p">,</span> <span class="s1">&#39;GAC&#39;</span><span class="p">],</span>
              <span class="s1">&#39;K&#39;</span><span class="p">:[</span><span class="s1">&#39;AAA&#39;</span><span class="p">,</span> <span class="s1">&#39;AAG&#39;</span><span class="p">],</span>
              <span class="s1">&#39;R&#39;</span><span class="p">:[</span><span class="s1">&#39;CGT&#39;</span><span class="p">,</span> <span class="s1">&#39;CGC&#39;</span><span class="p">,</span> <span class="s1">&#39;CGA&#39;</span><span class="p">,</span> <span class="s1">&#39;CGG&#39;</span><span class="p">,</span> <span class="s1">&#39;AGA&#39;</span><span class="p">,</span> <span class="s1">&#39;AGG&#39;</span><span class="p">],</span>
              <span class="s1">&#39;*&#39;</span><span class="p">:[</span><span class="s1">&#39;TAA&#39;</span><span class="p">,</span> <span class="s1">&#39;TAG&#39;</span><span class="p">,</span> <span class="s1">&#39;TGA&#39;</span><span class="p">]}</span>
    <span class="n">xcodon</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">codons</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">k</span><span class="o">!=</span><span class="s1">&#39;*&#39;</span><span class="p">:</span>
            <span class="n">xcodon</span><span class="o">+=</span> <span class="n">codons</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>
    <span class="n">codons</span><span class="p">[</span><span class="s1">&#39;X&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">xcodon</span>
    <span class="k">return</span> <span class="n">codons</span></div>

<span class="n">codon_table</span> <span class="o">=</span> <span class="n">get_codons</span><span class="p">()</span>

<div class="viewcode-block" id="get_codon_synonymous"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.get_codon_synonymous">[docs]</a><span class="k">def</span> <span class="nf">get_codon_synonymous</span><span class="p">():</span>
    <span class="c1"># generate table of synonymous codons</span>
    <span class="n">codons</span> <span class="o">=</span> <span class="n">get_codons</span><span class="p">()</span>
    <span class="n">out</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">codons</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">k</span><span class="o">!=</span><span class="s1">&#39;X&#39;</span><span class="p">:</span>
            <span class="n">x</span> <span class="o">=</span> <span class="n">codons</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">x</span><span class="p">:</span>
                    <span class="n">out</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">j</span><span class="p">)</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">x</span><span class="p">[</span><span class="n">x</span><span class="o">!=</span><span class="n">i</span><span class="p">]]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">out</span><span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
    <span class="k">return</span> <span class="n">out</span></div>

<span class="n">synonymous_codon_table</span> <span class="o">=</span> <span class="n">get_codon_synonymous</span><span class="p">()</span>

<div class="viewcode-block" id="get_IUPAC_nt"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.get_IUPAC_nt">[docs]</a><span class="k">def</span> <span class="nf">get_IUPAC_nt</span><span class="p">():</span>
    <span class="k">return</span> <span class="s1">&#39;ATGCRYSWKMBDHVN&#39;</span></div>

<div class="viewcode-block" id="get_AAlib"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.get_AAlib">[docs]</a><span class="k">def</span> <span class="nf">get_AAlib</span><span class="p">(</span><span class="n">AAseq</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Define a library of codons for each amino acid</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="n">catalog</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">codons</span> <span class="o">=</span> <span class="n">get_codons</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">AAseq</span><span class="o">.</span><span class="n">upper</span><span class="p">():</span>
        <span class="n">catalog</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">codons</span><span class="p">[</span><span class="n">j</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">catalog</span></div>

<div class="viewcode-block" id="load_args"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.load_args">[docs]</a><span class="k">def</span> <span class="nf">load_args</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="n">params</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Writes values from args into params dictionary</span>
<span class="sd">    argnew = new parameters</span>
<span class="sd">    params = old parameter set</span>
<span class="sd">    return updated parameter set as a dictionary</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="k">if</span> <span class="n">params</span><span class="o">==</span><span class="kc">None</span><span class="p">:</span>
        <span class="n">params</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">!=</span><span class="kc">None</span><span class="p">:</span>
        <span class="n">argvars</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="vm">__dict__</span>
        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">argvars</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">argvars</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">!=</span><span class="kc">None</span><span class="p">:</span>
                <span class="n">params</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">argvars</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">params</span></div>

<div class="viewcode-block" id="read_to_df"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.read_to_df">[docs]</a><span class="k">def</span> <span class="nf">read_to_df</span><span class="p">(</span><span class="n">fname</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Function used to read csv, genbank, fasta, or text file of sequences</span>
<span class="sd">    fname = filename to read</span>
<span class="sd">    return pandas dataframe of sequences in the file</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="n">out</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">fname</span><span class="p">)</span> <span class="o">==</span> <span class="nb">str</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Try reading&#39;</span><span class="p">,</span><span class="n">fname</span><span class="p">,</span><span class="s1">&#39;as genbank&#39;</span><span class="p">)</span>
            <span class="n">rec</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fname</span><span class="p">,</span> <span class="s1">&#39;genbank&#39;</span><span class="p">)</span>
            <span class="n">out</span> <span class="o">=</span> <span class="p">[[</span><span class="n">fname</span><span class="p">,</span> <span class="n">fname</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">rec</span><span class="o">.</span><span class="n">seq</span><span class="p">)]]</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Failed to read&#39;</span><span class="p">,</span><span class="n">fname</span><span class="p">,</span><span class="s1">&#39;as genbank&#39;</span><span class="p">)</span>
            
            <span class="k">try</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Try reading&#39;</span><span class="p">,</span><span class="n">fname</span><span class="p">,</span><span class="s1">&#39;as csv&#39;</span><span class="p">)</span>
                <span class="n">out</span> <span class="o">=</span> <span class="n">read_csv</span><span class="p">(</span><span class="n">fname</span><span class="p">)</span>
            <span class="k">except</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Failed to read&#39;</span><span class="p">,</span><span class="n">fname</span><span class="p">,</span><span class="s1">&#39;as csv&#39;</span><span class="p">)</span>

                <span class="k">try</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Try reading&#39;</span><span class="p">,</span><span class="n">fname</span><span class="p">,</span><span class="s1">&#39;as fasta&#39;</span><span class="p">)</span>
                    <span class="n">out</span> <span class="o">=</span> <span class="n">read_fasta</span><span class="p">(</span><span class="n">fname</span><span class="p">)</span>
                <span class="k">except</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Failed to read&#39;</span><span class="p">,</span><span class="n">fname</span><span class="p">,</span><span class="s1">&#39;as fasta&#39;</span><span class="p">)</span>

    <span class="c1"># format into a pandas dataframe</span>
    <span class="n">out</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">out</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;filename&#39;</span><span class="p">,</span><span class="s1">&#39;name&#39;</span><span class="p">,</span><span class="s1">&#39;sequence&#39;</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">out</span></div>

<div class="viewcode-block" id="read_csv"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.read_csv">[docs]</a><span class="k">def</span> <span class="nf">read_csv</span><span class="p">(</span><span class="n">fname</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Read sequences from a csv file</span>
<span class="sd">    fname = file to open</span>
<span class="sd">    returns a list with [[filename, name, sequence]]</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">fname</span><span class="p">)</span>
    <span class="c1"># figure out which has name</span>
    <span class="n">name</span> <span class="o">=</span> <span class="p">[</span><span class="n">fname</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
    <span class="n">col</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">,</span><span class="s1">&#39;label&#39;</span><span class="p">,</span><span class="s1">&#39;Strand&#39;</span><span class="p">,</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">col</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="n">name</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
            <span class="k">break</span>
    <span class="c1"># figure out which has sequence</span>
    <span class="n">col</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Sequence&#39;</span><span class="p">,</span><span class="s1">&#39;sequence&#39;</span><span class="p">,</span><span class="s1">&#39;seq&#39;</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">col</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="n">seq</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
            <span class="k">break</span>
    <span class="c1"># assign filename </span>
    <span class="k">if</span> <span class="p">(</span><span class="s1">&#39;filename&#39;</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">==</span> <span class="kc">False</span><span class="p">:</span>
        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;filename&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">fname</span>
    <span class="n">filename</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">filename</span>
    <span class="k">return</span> <span class="p">[[</span><span class="n">filename</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">name</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">seq</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">))]</span></div>

<div class="viewcode-block" id="read_fasta"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.read_fasta">[docs]</a><span class="k">def</span> <span class="nf">read_fasta</span><span class="p">(</span><span class="n">fname</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Read sequences from a fasta file</span>
<span class="sd">    fname = file to open</span>
<span class="sd">    returns a list with [[filename, name, sequence]]</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">fname</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span> 
        <span class="n">rec</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">Bio</span><span class="o">.</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="s1">&#39;fasta&#39;</span><span class="p">))</span>
    <span class="n">df</span> <span class="o">=</span> <span class="p">[[</span><span class="n">fname</span><span class="p">,</span> <span class="n">i</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">seq</span><span class="p">)]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">rec</span><span class="p">]</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">df</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">error</span><span class="p">()</span></div>

<div class="viewcode-block" id="translate"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.translate">[docs]</a><span class="k">def</span> <span class="nf">translate</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">frame</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">table</span><span class="o">=</span><span class="s1">&#39;Standard&#39;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Translates a nucleotide sequence into amino acids</span>
<span class="sd">    seq = input nucleotide sequence</span>
<span class="sd">    frame = reading frame</span>
<span class="sd">    table = codon table to use</span>
<span class="sd">    return amino acid sequence</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">Seq</span><span class="o">.</span><span class="n">Seq</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
    <span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="n">seq</span><span class="p">[</span><span class="n">frame</span><span class="p">:]</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">table</span><span class="o">=</span><span class="n">table</span><span class="p">))</span></div>

<div class="viewcode-block" id="revcomp"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.revcomp">[docs]</a><span class="k">def</span> <span class="nf">revcomp</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Returns reverse complement of a nucleic acid string</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="n">Bio</span><span class="o">.</span><span class="n">Seq</span><span class="o">.</span><span class="n">Seq</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">())</span></div>

<div class="viewcode-block" id="complement"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.complement">[docs]</a><span class="k">def</span> <span class="nf">complement</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Return complement of a nucleic acid string</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="n">Bio</span><span class="o">.</span><span class="n">Seq</span><span class="o">.</span><span class="n">Seq</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span><span class="o">.</span><span class="n">complement</span><span class="p">())</span></div>

<div class="viewcode-block" id="format_to_dna"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.format_to_dna">[docs]</a><span class="k">def</span> <span class="nf">format_to_dna</span><span class="p">(</span><span class="n">sequence</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Convert a string to have only IUPAC dna codes. Other erroneous letters are removed.</span>
<span class="sd">    sequence = string to change to IUPAC only codes</span>
<span class="sd">    return the cleaned string</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="n">allowed</span> <span class="o">=</span> <span class="n">get_IUPAC_nt</span><span class="p">()</span>
    <span class="n">out</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
    <span class="c1"># convert U to T</span>
    <span class="n">out</span> <span class="o">=</span> <span class="n">out</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;U&#39;</span><span class="p">,</span><span class="s1">&#39;T&#39;</span><span class="p">)</span>
    <span class="c1"># convert gap characters to N</span>
    <span class="n">out</span> <span class="o">=</span> <span class="n">out</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span><span class="s1">&#39;N&#39;</span><span class="p">)</span>
    <span class="n">out</span> <span class="o">=</span> <span class="n">out</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">,</span><span class="s1">&#39;N&#39;</span><span class="p">)</span>
    <span class="c1"># remove spaces</span>
    <span class="n">out</span> <span class="o">=</span> <span class="n">out</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="c1"># remove erroneous characters</span>
    <span class="n">remove</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">out</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">allowed</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">remove</span><span class="p">:</span>
        <span class="n">out</span> <span class="o">=</span> <span class="n">out</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">i</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">out</span></div>

<div class="viewcode-block" id="isDNA"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.isDNA">[docs]</a><span class="k">def</span> <span class="nf">isDNA</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
    <span class="n">dna</span> <span class="o">=</span> <span class="n">get_IUPAC_nt</span><span class="p">()</span>
    <span class="k">return</span> <span class="nb">set</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">upper</span><span class="p">())</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">dna</span><span class="p">))</span></div>

<div class="viewcode-block" id="isAA"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.isAA">[docs]</a><span class="k">def</span> <span class="nf">isAA</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
    <span class="k">return</span> <span class="nb">set</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">codon_table</span><span class="p">))</span></div>

<div class="viewcode-block" id="isBool"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.isBool">[docs]</a><span class="k">def</span> <span class="nf">isBool</span><span class="p">(</span><span class="n">value</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Check if list of a values are only booleans</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="n">values</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
    <span class="c1"># splice based on boolean array</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">values</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">3</span> <span class="ow">and</span> <span class="nb">type</span><span class="p">(</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">==</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span> <span class="ow">and</span> <span class="nb">type</span><span class="p">(</span><span class="n">values</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span><span class="o">==</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">True</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">False</span></div>

<div class="viewcode-block" id="isStringArray"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.isStringArray">[docs]</a><span class="k">def</span> <span class="nf">isStringArray</span><span class="p">(</span><span class="n">value</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Check if list contains only strings</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="n">x</span> <span class="o">=</span> <span class="p">[</span><span class="nb">type</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="o">==</span><span class="nb">str</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">value</span><span class="p">]</span>
    <span class="k">return</span> <span class="nb">sum</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">value</span><span class="p">)</span></div>

<div class="viewcode-block" id="isIntArray"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.isIntArray">[docs]</a><span class="k">def</span> <span class="nf">isIntArray</span><span class="p">(</span><span class="n">value</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Check if list contains only integers</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="n">x</span> <span class="o">=</span> <span class="p">[</span><span class="nb">type</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="o">==</span><span class="nb">int</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">value</span><span class="p">]</span>
    <span class="k">return</span> <span class="nb">sum</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">value</span><span class="p">)</span></div>

<div class="viewcode-block" id="substring_search"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.substring_search">[docs]</a><span class="k">def</span> <span class="nf">substring_search</span><span class="p">(</span><span class="n">ref</span><span class="p">,</span> <span class="n">qry</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Searches for all occurrences of the given substrings in a string</span>
<span class="sd">    ref = string to search</span>
<span class="sd">    qry = substring to locate</span>
<span class="sd">    return a list with start locations</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">out</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">x</span> <span class="o">=</span> <span class="n">ref</span>
    <span class="n">offset</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">while</span> <span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">idx</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="n">qry</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">idx</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
            <span class="k">break</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">out</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="o">+</span><span class="n">offset</span><span class="p">)</span>
            <span class="n">x</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="n">idx</span><span class="o">+</span><span class="mi">1</span><span class="p">:]</span>
            <span class="n">offset</span><span class="o">+=</span> <span class="n">idx</span><span class="o">+</span><span class="mi">1</span>
    <span class="k">return</span> <span class="n">out</span></div>

<div class="viewcode-block" id="search_key"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.search_key">[docs]</a><span class="k">def</span> <span class="nf">search_key</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">keyword</span><span class="p">,</span> <span class="n">rev</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Searches dictionary for first instance of keyword</span>
<span class="sd">    data = dictionary to search for the keyword</span>
<span class="sd">    keyword = keyword value to find in keys</span>
<span class="sd">    rev = search key in keyword instead of keyword in key</span>
<span class="sd">    return value of the keyword if found. Otherwise return None</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">rev</span><span class="p">:</span>
            <span class="n">value</span> <span class="o">=</span> <span class="p">(</span><span class="n">key</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="ow">in</span> <span class="n">keyword</span><span class="o">.</span><span class="n">lower</span><span class="p">())</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">value</span> <span class="o">=</span> <span class="p">(</span><span class="n">keyword</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="ow">in</span> <span class="n">key</span><span class="o">.</span><span class="n">lower</span><span class="p">())</span>
        <span class="k">if</span> <span class="n">value</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">data</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>
    <span class="k">return</span> <span class="kc">None</span></div>

<div class="viewcode-block" id="get_rand_letters"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.get_rand_letters">[docs]</a><span class="k">def</span> <span class="nf">get_rand_letters</span><span class="p">(</span><span class="n">N</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Generate nucleotide domain of length N</span>
<span class="sd">    N = length of nucleotides to generate</span>
<span class="sd">    method = method to use</span>
<span class="sd">    returns string of nucleotides</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="c1"># no repeating bases</span>
    <span class="n">m0</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;A&#39;</span><span class="p">:</span><span class="s1">&#39;TGC&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">:</span><span class="s1">&#39;GCA&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">:</span><span class="s1">&#39;ATC&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">:</span><span class="s1">&#39;ATG&#39;</span><span class="p">}</span>
    <span class="c1"># opposite day</span>
    <span class="n">m1</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;A&#39;</span><span class="p">:</span><span class="s1">&#39;GC&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">:</span><span class="s1">&#39;GC&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">:</span><span class="s1">&#39;AT&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">:</span><span class="s1">&#39;AT&#39;</span><span class="p">}</span>
    <span class="c1"># loop back on predecessor bases</span>
    <span class="n">m2</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;AT&#39;</span><span class="p">:</span><span class="s1">&#39;CA&#39;</span><span class="p">,</span> <span class="s1">&#39;AG&#39;</span><span class="p">:</span><span class="s1">&#39;CA&#39;</span><span class="p">,</span> <span class="s1">&#39;AC&#39;</span><span class="p">:</span><span class="s1">&#39;GA&#39;</span><span class="p">,</span>
          <span class="s1">&#39;TA&#39;</span><span class="p">:</span><span class="s1">&#39;GT&#39;</span><span class="p">,</span> <span class="s1">&#39;TG&#39;</span><span class="p">:</span><span class="s1">&#39;CT&#39;</span><span class="p">,</span> <span class="s1">&#39;TC&#39;</span><span class="p">:</span><span class="s1">&#39;GT&#39;</span><span class="p">,</span>
          <span class="s1">&#39;GA&#39;</span><span class="p">:</span><span class="s1">&#39;TG&#39;</span><span class="p">,</span> <span class="s1">&#39;GT&#39;</span><span class="p">:</span><span class="s1">&#39;AG&#39;</span><span class="p">,</span> <span class="s1">&#39;GC&#39;</span><span class="p">:</span><span class="s1">&#39;AG&#39;</span><span class="p">,</span>
          <span class="s1">&#39;CA&#39;</span><span class="p">:</span><span class="s1">&#39;TC&#39;</span><span class="p">,</span> <span class="s1">&#39;CT&#39;</span><span class="p">:</span><span class="s1">&#39;AC&#39;</span><span class="p">,</span> <span class="s1">&#39;CG&#39;</span><span class="p">:</span><span class="s1">&#39;TC&#39;</span><span class="p">}</span>
    <span class="c1"># loop back on predecessor bases</span>
    <span class="n">m3</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;AT&#39;</span><span class="p">:</span><span class="s1">&#39;CA&#39;</span><span class="p">,</span> <span class="s1">&#39;AG&#39;</span><span class="p">:</span><span class="s1">&#39;CA&#39;</span><span class="p">,</span> <span class="s1">&#39;AC&#39;</span><span class="p">:</span><span class="s1">&#39;GA&#39;</span><span class="p">,</span>
          <span class="s1">&#39;TA&#39;</span><span class="p">:</span><span class="s1">&#39;GT&#39;</span><span class="p">,</span> <span class="s1">&#39;TG&#39;</span><span class="p">:</span><span class="s1">&#39;CT&#39;</span><span class="p">,</span> <span class="s1">&#39;TC&#39;</span><span class="p">:</span><span class="s1">&#39;GT&#39;</span><span class="p">,</span>
          <span class="s1">&#39;GA&#39;</span><span class="p">:</span><span class="s1">&#39;TG&#39;</span><span class="p">,</span> <span class="s1">&#39;GT&#39;</span><span class="p">:</span><span class="s1">&#39;AG&#39;</span><span class="p">,</span> <span class="s1">&#39;GC&#39;</span><span class="p">:</span><span class="s1">&#39;TG&#39;</span><span class="p">,</span>
          <span class="s1">&#39;CA&#39;</span><span class="p">:</span><span class="s1">&#39;TC&#39;</span><span class="p">,</span> <span class="s1">&#39;CT&#39;</span><span class="p">:</span><span class="s1">&#39;AC&#39;</span><span class="p">,</span> <span class="s1">&#39;CG&#39;</span><span class="p">:</span><span class="s1">&#39;AC&#39;</span><span class="p">}</span>
    <span class="c1"># loop back on predecessor bases</span>
    <span class="n">m4</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;AT&#39;</span><span class="p">:</span><span class="s1">&#39;GA&#39;</span><span class="p">,</span> <span class="s1">&#39;AG&#39;</span><span class="p">:</span><span class="s1">&#39;CA&#39;</span><span class="p">,</span> <span class="s1">&#39;AC&#39;</span><span class="p">:</span><span class="s1">&#39;GA&#39;</span><span class="p">,</span>
          <span class="s1">&#39;TA&#39;</span><span class="p">:</span><span class="s1">&#39;CT&#39;</span><span class="p">,</span> <span class="s1">&#39;TG&#39;</span><span class="p">:</span><span class="s1">&#39;CT&#39;</span><span class="p">,</span> <span class="s1">&#39;TC&#39;</span><span class="p">:</span><span class="s1">&#39;GT&#39;</span><span class="p">,</span>
          <span class="s1">&#39;GA&#39;</span><span class="p">:</span><span class="s1">&#39;TG&#39;</span><span class="p">,</span> <span class="s1">&#39;GT&#39;</span><span class="p">:</span><span class="s1">&#39;AG&#39;</span><span class="p">,</span> <span class="s1">&#39;GC&#39;</span><span class="p">:</span><span class="s1">&#39;AG&#39;</span><span class="p">,</span>
          <span class="s1">&#39;CA&#39;</span><span class="p">:</span><span class="s1">&#39;TC&#39;</span><span class="p">,</span> <span class="s1">&#39;CT&#39;</span><span class="p">:</span><span class="s1">&#39;AC&#39;</span><span class="p">,</span> <span class="s1">&#39;CG&#39;</span><span class="p">:</span><span class="s1">&#39;TC&#39;</span><span class="p">}</span>
    <span class="c1"># loop back on predecessor bases</span>
    <span class="n">m5</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;AT&#39;</span><span class="p">:</span><span class="s1">&#39;GA&#39;</span><span class="p">,</span> <span class="s1">&#39;AG&#39;</span><span class="p">:</span><span class="s1">&#39;CA&#39;</span><span class="p">,</span> <span class="s1">&#39;AC&#39;</span><span class="p">:</span><span class="s1">&#39;GA&#39;</span><span class="p">,</span>
          <span class="s1">&#39;TA&#39;</span><span class="p">:</span><span class="s1">&#39;CT&#39;</span><span class="p">,</span> <span class="s1">&#39;TG&#39;</span><span class="p">:</span><span class="s1">&#39;CT&#39;</span><span class="p">,</span> <span class="s1">&#39;TC&#39;</span><span class="p">:</span><span class="s1">&#39;GT&#39;</span><span class="p">,</span>
          <span class="s1">&#39;GA&#39;</span><span class="p">:</span><span class="s1">&#39;TG&#39;</span><span class="p">,</span> <span class="s1">&#39;GT&#39;</span><span class="p">:</span><span class="s1">&#39;AG&#39;</span><span class="p">,</span> <span class="s1">&#39;GC&#39;</span><span class="p">:</span><span class="s1">&#39;TG&#39;</span><span class="p">,</span>
          <span class="s1">&#39;CA&#39;</span><span class="p">:</span><span class="s1">&#39;TC&#39;</span><span class="p">,</span> <span class="s1">&#39;CT&#39;</span><span class="p">:</span><span class="s1">&#39;AC&#39;</span><span class="p">,</span> <span class="s1">&#39;CG&#39;</span><span class="p">:</span><span class="s1">&#39;AC&#39;</span><span class="p">}</span>
    <span class="n">letters</span> <span class="o">=</span> <span class="p">[</span><span class="n">m0</span><span class="p">,</span> <span class="n">m1</span><span class="p">,</span> <span class="n">m2</span><span class="p">,</span> <span class="n">m3</span><span class="p">,</span> <span class="n">m4</span><span class="p">,</span> <span class="n">m5</span><span class="p">]</span>
    <span class="n">letters</span> <span class="o">=</span> <span class="n">letters</span><span class="p">[</span><span class="n">method</span><span class="p">]</span>
    <span class="n">out</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">letters</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
    <span class="n">out</span> <span class="o">=</span> <span class="n">out</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">out</span><span class="p">))]</span>
    <span class="n">N_in</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">out</span><span class="p">)</span>
    <span class="n">N_rand</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">letters</span><span class="p">[</span><span class="n">out</span><span class="p">])</span>
    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">N_rand</span><span class="p">,</span> <span class="n">N</span><span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">out</span><span class="p">)):</span>
        <span class="n">key</span> <span class="o">=</span> <span class="n">out</span><span class="p">[</span><span class="o">-</span><span class="n">N_in</span><span class="p">:]</span>
        <span class="n">out</span><span class="o">+=</span> <span class="n">letters</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="n">j</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">out</span></div>

<div class="viewcode-block" id="json_pretty"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.json_pretty">[docs]</a><span class="k">def</span> <span class="nf">json_pretty</span><span class="p">(</span><span class="n">data</span><span class="p">):</span>
    <span class="k">return</span> <span class="n">json</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">indent</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span></div>

<div class="viewcode-block" id="get_timestamp"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.get_timestamp">[docs]</a><span class="k">def</span> <span class="nf">get_timestamp</span><span class="p">():</span>
    <span class="k">return</span> <span class="s1">&#39;._&#39;</span><span class="o">+</span><span class="n">datetime</span><span class="o">.</span><span class="n">datetime</span><span class="o">.</span><span class="n">fromtimestamp</span><span class="p">(</span><span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">())</span><span class="o">.</span><span class="n">strftime</span><span class="p">(</span><span class="s1">&#39;%Y_%m_</span><span class="si">%d</span><span class="s1">_%H_%M_%S&#39;</span><span class="p">)</span></div>

<div class="viewcode-block" id="set_log_level"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.set_log_level">[docs]</a><span class="k">def</span> <span class="nf">set_log_level</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s1">&#39;run.log&#39;</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="s1">&#39;INFO&#39;</span><span class="p">):</span>
    <span class="n">fmt</span> <span class="o">=</span> <span class="s1">&#39;pid[&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">getpid</span><span class="p">())</span><span class="o">+</span><span class="s1">&#39;] </span><span class="si">%(asctime)s</span><span class="s1">.</span><span class="si">%(msecs)03d</span><span class="s1"> </span><span class="si">%(levelname)s</span><span class="s1">: </span><span class="si">%(message)s</span><span class="s1">&#39;</span>
    <span class="n">dfmt</span> <span class="o">=</span> <span class="s1">&#39;%Y-%m-</span><span class="si">%d</span><span class="s1"> %H:%M:%S&#39;</span>
    <span class="n">formatter</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">Formatter</span><span class="p">(</span><span class="n">fmt</span><span class="p">,</span> <span class="n">datefmt</span><span class="o">=</span><span class="n">dfmt</span><span class="p">)</span>
    <span class="c1"># print to stdout</span>
    <span class="n">stdout_handler</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">StreamHandler</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="p">)</span>
    <span class="n">stdout_handler</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">level</span><span class="p">)</span>
    <span class="n">stdout_handler</span><span class="o">.</span><span class="n">setFormatter</span><span class="p">(</span><span class="n">formatter</span><span class="p">)</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">()</span><span class="o">.</span><span class="n">addHandler</span><span class="p">(</span><span class="n">stdout_handler</span><span class="p">)</span>
    <span class="c1"># print to log file</span>
    <span class="n">handle</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">FileHandler</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="n">filename</span><span class="p">)</span>
    <span class="n">handle</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">level</span><span class="p">)</span>
    <span class="n">handle</span><span class="o">.</span><span class="n">setFormatter</span><span class="p">(</span><span class="n">formatter</span><span class="p">)</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">()</span><span class="o">.</span><span class="n">addHandler</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">()</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">level</span><span class="p">)</span> </div>

<div class="viewcode-block" id="get_memory_usage"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.get_memory_usage">[docs]</a><span class="k">def</span> <span class="nf">get_memory_usage</span><span class="p">():</span>
    <span class="n">pid</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getpid</span><span class="p">()</span>
    <span class="n">process</span> <span class="o">=</span> <span class="n">psutil</span><span class="o">.</span><span class="n">Process</span><span class="p">(</span><span class="n">pid</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">process</span><span class="o">.</span><span class="n">memory_full_info</span><span class="p">()</span><span class="o">.</span><span class="n">rss</span><span class="o">/</span><span class="mi">1024</span><span class="o">/</span><span class="mi">1024</span></div>

<div class="viewcode-block" id="generate_biobricks"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.generate_biobricks">[docs]</a><span class="k">def</span> <span class="nf">generate_biobricks</span><span class="p">(</span><span class="n">database</span><span class="p">,</span> <span class="n">keyword</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Searches an annotation database for sequences matching the keyword</span>
<span class="sd">    database = list or andas dataframe with columns [label, feature_type, color, strand, sequence]</span>
<span class="sd">    keyword = keyword matching the label label in the annotation database</span>
<span class="sd">    verbose = print list of parts that were generated</span>
<span class="sd">    return a list of plasmid objects which can be inserted into another plasmid or replace existing features</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">database</span><span class="p">[[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">,</span><span class="s1">&#39;feature_type&#39;</span><span class="p">,</span><span class="s1">&#39;color&#39;</span><span class="p">,</span><span class="s1">&#39;strand&#39;</span><span class="p">,</span><span class="s1">&#39;sequence&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span>
    <span class="k">for</span> <span class="p">[</span><span class="n">label</span><span class="p">,</span> <span class="n">feature_type</span><span class="p">,</span> <span class="n">color</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">sequence</span><span class="p">]</span> <span class="ow">in</span> <span class="n">data</span><span class="p">:</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">keyword</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="ow">in</span> <span class="n">label</span><span class="o">.</span><span class="n">lower</span><span class="p">()):</span>
            <span class="n">loc</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqFeature</span><span class="o">.</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">strand</span><span class="p">))</span>
            <span class="n">qualifiers</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">:[</span><span class="n">label</span><span class="p">],</span> <span class="s1">&#39;color&#39;</span><span class="p">:</span><span class="n">color</span><span class="p">}</span>
            <span class="n">gene</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqFeature</span><span class="o">.</span><span class="n">SeqFeature</span><span class="p">(</span><span class="nb">type</span><span class="o">=</span><span class="n">feature_type</span><span class="p">,</span> <span class="n">location</span><span class="o">=</span><span class="n">loc</span><span class="p">,</span> <span class="n">qualifiers</span><span class="o">=</span><span class="n">qualifiers</span><span class="p">)</span>
            <span class="n">SeqRecord</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">SeqRecord</span><span class="p">(</span><span class="n">seq</span><span class="o">=</span><span class="n">sequence</span><span class="p">,</span> <span class="n">features</span><span class="o">=</span><span class="p">[</span><span class="n">gene</span><span class="p">])</span>
            <span class="n">output</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">plasmid</span><span class="p">(</span><span class="n">SeqRecord</span><span class="p">))</span>
    <span class="c1"># print results</span>
    <span class="k">if</span> <span class="n">verbose</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">output</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">:</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">i</span><span class="o">.</span><span class="n">to_dataframe</span><span class="p">()</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">output</span><span class="p">])</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">,</span><span class="s1">&#39;location&#39;</span><span class="p">,</span><span class="s1">&#39;start&#39;</span><span class="p">,</span><span class="s1">&#39;end&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">output</span></div>

<div class="viewcode-block" id="load_annotation_database"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.load_annotation_database">[docs]</a><span class="k">def</span> <span class="nf">load_annotation_database</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">lib_format</span><span class="o">=</span><span class="s1">&#39;csv&#39;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Loads an annotation into a pandas dataframe, so it can be used to annotate sequences</span>
<span class="sd">    lib_format = type of formatting the annotation database is stored as.</span>
<span class="sd">                 valid options are csv or ApE</span>
<span class="sd">    returns a pandas dataframe with columns [label, feature, color, start, end, strand, sequence]</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="k">if</span> <span class="n">lib_format</span><span class="o">==</span><span class="s1">&#39;csv&#39;</span><span class="p">:</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
        <span class="k">if</span> <span class="p">(</span><span class="s1">&#39;strand&#39;</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span><span class="o">==</span><span class="kc">False</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Sequence orientation is not defined. Please add strand column to csv file&#39;</span><span class="p">)</span>
            <span class="k">return</span> <span class="kc">None</span>
        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;strand&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;strand&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">lib_format</span><span class="o">==</span><span class="s1">&#39;ApE&#39;</span><span class="p">:</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;Default_Features.txt&#39;</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">4</span><span class="p">:])</span>
        <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">,</span><span class="s1">&#39;sequence&#39;</span><span class="p">,</span><span class="s1">&#39;feature_type&#39;</span><span class="p">,</span><span class="s1">&#39;color&#39;</span><span class="p">]</span>
        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;strand&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="c1"># do some safety checks on the dna sequences</span>
    <span class="n">data</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">)):</span>
        <span class="n">seq</span> <span class="o">=</span> <span class="n">format_to_dna</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s1">&#39;sequence&#39;</span><span class="p">])</span>
        <span class="c1"># correct strand orientation to 5&#39;-&gt;3&#39;</span>
        <span class="k">if</span> <span class="n">df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s1">&#39;strand&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">seq</span> <span class="o">=</span> <span class="n">revcomp</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
        <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
    <span class="n">df</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span>    
    <span class="n">df</span><span class="p">[</span><span class="s1">&#39;length&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">]]</span>
    <span class="n">df</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">df</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">df</span><span class="p">[</span><span class="s1">&#39;strand&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="k">return</span> <span class="n">df</span></div>

<div class="viewcode-block" id="unpack_FeatureLocation"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.unpack_FeatureLocation">[docs]</a><span class="k">def</span> <span class="nf">unpack_FeatureLocation</span><span class="p">(</span><span class="n">loc</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Unpack locations nested in locations object</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="k">if</span> <span class="s1">&#39;parts&#39;</span> <span class="ow">in</span> <span class="n">loc</span><span class="o">.</span><span class="vm">__dict__</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="n">out</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">loc</span><span class="o">.</span><span class="n">parts</span><span class="p">:</span>
            <span class="n">out</span><span class="o">+=</span> <span class="n">unpack_FeatureLocation</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">out</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">[</span><span class="n">loc</span><span class="p">]</span></div>

<div class="viewcode-block" id="unwrap_FeatureLocation"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.unwrap_FeatureLocation">[docs]</a><span class="k">def</span> <span class="nf">unwrap_FeatureLocation</span><span class="p">(</span><span class="n">loc</span><span class="p">,</span> <span class="n">length</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Get proper start and end locations of a genetic feature that wraps around the origin</span>
<span class="sd">    loc = Bio.SeqFeature location</span>
<span class="sd">    length = length of circular Plasmid</span>
<span class="sd">    returns [start, end, strand]</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="n">strand</span> <span class="o">=</span> <span class="n">loc</span><span class="o">.</span><span class="n">strand</span>
    <span class="n">start</span> <span class="o">=</span> <span class="n">loc</span><span class="o">.</span><span class="n">start</span><span class="o">%</span><span class="n">length</span>
    <span class="n">end</span> <span class="o">=</span> <span class="n">loc</span><span class="o">.</span><span class="n">end</span><span class="o">%</span><span class="n">length</span>
    <span class="k">if</span> <span class="n">end</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span>
        <span class="n">end</span><span class="o">=</span><span class="n">length</span>
    <span class="k">return</span> <span class="p">[</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">strand</span><span class="p">]</span></div>

<div class="viewcode-block" id="wrap_FeatureLocation"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.wrap_FeatureLocation">[docs]</a><span class="k">def</span> <span class="nf">wrap_FeatureLocation</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">length</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Create a CompoundLocation or FeatureLocation object that wrap the annotation around the origin of a circular Plasmid</span>
<span class="sd">    start = start base position of the gene</span>
<span class="sd">    end = end base position of the gene</span>
<span class="sd">    strand = reading frame of the gene 5&#39;-&gt;3&#39; is +1</span>
<span class="sd">    length = base pair length of the Plasmid</span>
<span class="sd">    returns Bio.SeqFeature.CompoundLocation or Bio.SeqFeature.FeatureLocation object</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="n">start</span> <span class="o">=</span> <span class="n">start</span><span class="o">%</span><span class="n">length</span>
    <span class="n">end</span> <span class="o">=</span> <span class="n">end</span><span class="o">%</span><span class="n">length</span>
    <span class="k">if</span> <span class="n">end</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span>
        <span class="n">end</span> <span class="o">=</span> <span class="n">length</span>

    <span class="k">if</span> <span class="n">end</span> <span class="o">&lt;=</span> <span class="n">start</span><span class="p">:</span>
        <span class="n">loc1</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqFeature</span><span class="o">.</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">start</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">length</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">strand</span><span class="p">))</span>
        <span class="n">loc2</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqFeature</span><span class="o">.</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">end</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">strand</span><span class="p">))</span>
        <span class="n">loc</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqFeature</span><span class="o">.</span><span class="n">CompoundLocation</span><span class="p">([</span><span class="n">loc1</span><span class="p">,</span><span class="n">loc2</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">loc</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqFeature</span><span class="o">.</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">start</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">end</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">strand</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">loc</span></div>

<div class="viewcode-block" id="shift_CompoundLocation"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.shift_CompoundLocation">[docs]</a><span class="k">def</span> <span class="nf">shift_CompoundLocation</span><span class="p">(</span><span class="n">loc</span><span class="p">,</span> <span class="n">N</span><span class="p">,</span> <span class="n">L</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Applies nucleotide shift to compound locs and merges adjacent FeatureLocations that arise after the shift</span>
<span class="sd">    loc = Bio.SeqFeature.FeatureLocation or Bio.SeqFeature.CompoundLocation object</span>
<span class="sd">    N = number of nucleotides to shift</span>
<span class="sd">    L = length of the Plasmid sequence</span>
<span class="sd">    circular = apply relevant corrections if the genome is circular</span>
<span class="sd">    returns a Bio.SeqFeature.CompoundLocation or Bio.SeqFeature.FeatureLocation</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="n">loc</span><span class="o">+=</span><span class="nb">int</span><span class="p">(</span><span class="n">N</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">circular</span><span class="p">:</span>
        <span class="k">if</span> <span class="s1">&#39;parts&#39;</span> <span class="ow">in</span> <span class="n">loc</span><span class="o">.</span><span class="vm">__dict__</span><span class="p">:</span>
            <span class="c1"># correct the shift in each sub loc</span>
            <span class="n">out</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">loc</span><span class="o">.</span><span class="vm">__dict__</span><span class="p">[</span><span class="s1">&#39;parts&#39;</span><span class="p">]:</span>
                <span class="n">x</span> <span class="o">=</span> <span class="n">shift_CompoundLocation</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">L</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                <span class="c1"># unwrap joint locations</span>
                <span class="k">if</span> <span class="s1">&#39;parts&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="o">.</span><span class="vm">__dict__</span><span class="p">:</span>
                    <span class="n">out</span><span class="o">+=</span> <span class="n">x</span><span class="o">.</span><span class="vm">__dict__</span><span class="p">[</span><span class="s1">&#39;parts&#39;</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">out</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
            <span class="c1"># merge adjacent locations</span>
            <span class="n">keep</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">start</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">out</span><span class="p">)):</span>
                <span class="c1"># find the start</span>
                <span class="k">if</span> <span class="n">start</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
                    <span class="n">start</span> <span class="o">=</span> <span class="n">out</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">start</span>
                <span class="c1"># record the new position when a discontinuity is found</span>
                <span class="k">if</span> <span class="n">i</span><span class="o">+</span><span class="mi">1</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">out</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">out</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">end</span> <span class="o">==</span> <span class="n">out</span><span class="p">[</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">start</span> <span class="ow">and</span> <span class="n">out</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="n">out</span><span class="p">[</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">strand</span><span class="p">)</span> <span class="o">==</span> <span class="kc">False</span><span class="p">:</span>
                    <span class="n">keep</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Bio</span><span class="o">.</span><span class="n">SeqFeature</span><span class="o">.</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">out</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">end</span><span class="p">,</span> <span class="n">out</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">strand</span><span class="p">))</span>
                    <span class="n">start</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">keep</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">loc</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqFeature</span><span class="o">.</span><span class="n">CompoundLocation</span><span class="p">(</span><span class="n">keep</span><span class="p">)</span>
            <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">keep</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">loc</span> <span class="o">=</span> <span class="n">keep</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">loc</span> <span class="o">=</span> <span class="n">wrap_FeatureLocation</span><span class="p">(</span><span class="n">loc</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">loc</span><span class="o">.</span><span class="n">end</span><span class="p">,</span> <span class="n">loc</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span> <span class="n">L</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">loc</span></div>

<div class="viewcode-block" id="simulate_reads"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.simulate_reads">[docs]</a><span class="k">def</span> <span class="nf">simulate_reads</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">N</span><span class="p">,</span> <span class="n">error</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Generate library of errored reads</span>
<span class="sd">    df = dataframe with columns [name, sequence]</span>
<span class="sd">    N = number of sequences to generate</span>
<span class="sd">    error = mean error for poisson distribution of errors</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="n">out</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">df</span><span class="p">[</span><span class="s1">&#39;mutations&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">df</span><span class="p">[[</span><span class="s1">&#39;name&#39;</span><span class="p">,</span><span class="s1">&#39;sequence&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
        <span class="n">mut</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">error</span><span class="p">,</span> <span class="n">N</span><span class="p">)</span>
        <span class="n">x</span> <span class="o">=</span> <span class="p">[[</span><span class="n">name</span><span class="p">,</span> <span class="n">m</span><span class="p">,</span> <span class="n">get_mutations</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">m</span><span class="p">)]</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">mut</span><span class="p">]</span>
        <span class="n">col</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">,</span><span class="s1">&#39;mutations&#39;</span><span class="p">,</span><span class="s1">&#39;sequence&#39;</span><span class="p">]</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">col</span><span class="p">)</span>
        <span class="n">out</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">out</span><span class="p">)</span></div>

<div class="viewcode-block" id="get_mutations"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.get_mutations">[docs]</a><span class="k">def</span> <span class="nf">get_mutations</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">N</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Adds mutations to the sequence</span>
<span class="sd">    seq = sequence to mutate</span>
<span class="sd">    N = number of base pairs to mutate</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="c1"># choose positions to mutate</span>
    <span class="n">L</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
    <span class="n">idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">permutation</span><span class="p">(</span><span class="n">L</span><span class="p">)[:</span><span class="n">N</span><span class="p">]</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="s1">&#39;-&#39;</span> <span class="o">+</span> <span class="n">seq</span> <span class="o">+</span> <span class="s1">&#39;-&#39;</span>
    <span class="n">idx</span><span class="o">+=</span><span class="mi">1</span>
    <span class="n">offset</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">idx</span><span class="p">:</span>
        <span class="n">i</span><span class="o">+=</span> <span class="n">offset</span>
        <span class="n">let</span> <span class="o">=</span> <span class="n">seq</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">mut</span> <span class="o">=</span> <span class="n">choose_mutation</span><span class="p">(</span><span class="n">let</span><span class="p">)</span>
        <span class="n">seq</span> <span class="o">=</span> <span class="n">seq</span><span class="p">[:</span><span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="n">mut</span> <span class="o">+</span> <span class="n">seq</span><span class="p">[</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">:]</span>
        <span class="n">offset</span><span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">mut</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">seq</span></div>

<div class="viewcode-block" id="choose_mutation"><a class="viewcode-back" href="../../plasmid.html#plasmid.misc.choose_mutation">[docs]</a><span class="k">def</span> <span class="nf">choose_mutation</span><span class="p">(</span><span class="n">let</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Choose type of mutation</span>
<span class="sd">    let = letters</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="n">mut</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;A&#39;</span><span class="p">:</span><span class="s1">&#39;TGC&#39;</span><span class="p">,</span><span class="s1">&#39;T&#39;</span><span class="p">:</span><span class="s1">&#39;AGC&#39;</span><span class="p">,</span><span class="s1">&#39;G&#39;</span><span class="p">:</span><span class="s1">&#39;ATC&#39;</span><span class="p">,</span><span class="s1">&#39;C&#39;</span><span class="p">:</span><span class="s1">&#39;ATG&#39;</span><span class="p">,</span><span class="s1">&#39;-&#39;</span><span class="p">:</span><span class="s1">&#39;ATGC&#39;</span><span class="p">}</span>
    <span class="n">nuc</span> <span class="o">=</span> <span class="s1">&#39;ATGC&#39;</span>
    <span class="c1"># perform deletion</span>
    <span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">3</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">x</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span> <span class="s1">&#39;-&#39;</span>
    <span class="c1"># perform insertion</span>
    <span class="k">elif</span> <span class="n">x</span><span class="o">==</span><span class="mi">1</span><span class="p">:</span>
        <span class="n">idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">nuc</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">let</span><span class="o">+</span><span class="n">nuc</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>
    <span class="c1"># perform replacement</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">nuc</span> <span class="o">=</span> <span class="n">mut</span><span class="p">[</span><span class="n">let</span><span class="p">]</span> 
        <span class="n">idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">nuc</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">nuc</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span></div>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Zhewei Chen.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>