<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>plasmid.plasmid &mdash; Plasmid: A python library for gene editing 2023.06.29 documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/sphinx_highlight.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            Plasmid: A python library for gene editing
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">How-to guides &amp; Examples</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../notebooks/genbank_editing.html">Manipulating genbank records with Plasmid</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../notebooks/primer_design.html">Designing primers with Plasmid</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../notebooks/clustering.html">Aligning and clustering sequences with Plasmid</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">Plasmid: A python library for gene editing</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">Module code</a></li>
      <li class="breadcrumb-item active">plasmid.plasmid</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for plasmid.plasmid</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python</span>
<span class="c1"># Class functions to examine genbank files and sequences</span>

<span class="c1"># computing libraries</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">copy</span>

<span class="c1"># reading and writing genbank files</span>
<span class="kn">import</span> <span class="nn">Bio</span>
<span class="kn">import</span> <span class="nn">Bio.SeqIO</span>

<span class="c1"># local libaries</span>
<span class="kn">from</span> <span class="nn">.misc</span> <span class="kn">import</span> <span class="n">search_key</span>
<span class="kn">from</span> <span class="nn">.misc</span> <span class="kn">import</span> <span class="n">format_to_dna</span>
<span class="kn">from</span> <span class="nn">.misc</span> <span class="kn">import</span> <span class="n">revcomp</span>

<span class="kn">from</span> <span class="nn">.misc</span> <span class="kn">import</span> <span class="n">isDNA</span>
<span class="kn">from</span> <span class="nn">.misc</span> <span class="kn">import</span> <span class="n">isAA</span>
<span class="kn">from</span> <span class="nn">.misc</span> <span class="kn">import</span> <span class="n">isBool</span>
<span class="kn">from</span> <span class="nn">.misc</span> <span class="kn">import</span> <span class="n">isStringArray</span>
<span class="kn">from</span> <span class="nn">.misc</span> <span class="kn">import</span> <span class="n">isIntArray</span>

<span class="kn">from</span> <span class="nn">.misc</span> <span class="kn">import</span> <span class="n">read_to_df</span>
<span class="kn">from</span> <span class="nn">.misc</span> <span class="kn">import</span> <span class="n">get_timestamp</span>
<span class="kn">from</span> <span class="nn">.misc</span> <span class="kn">import</span> <span class="n">unwrap_FeatureLocation</span>
<span class="kn">from</span> <span class="nn">.misc</span> <span class="kn">import</span> <span class="n">wrap_FeatureLocation</span>
<span class="kn">from</span> <span class="nn">.misc</span> <span class="kn">import</span> <span class="n">shift_CompoundLocation</span>

<span class="kn">from</span> <span class="nn">.graphic</span> <span class="kn">import</span> <span class="n">Graphic</span>
<span class="kn">from</span> <span class="nn">.aligner</span> <span class="kn">import</span> <span class="n">Aligner</span>

<div class="viewcode-block" id="read_genbank"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.read_genbank">[docs]</a><span class="k">def</span> <span class="nf">read_genbank</span><span class="p">(</span><span class="n">fname</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Reads a genbank file into plasmid class object</span>
<span class="sd">    fname = genbank file to read</span>
<span class="sd">    returns Plasmid object with data loaded</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="c1"># load data with SeqIO</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;reading &#39;</span><span class="p">,</span><span class="n">fname</span><span class="p">,</span><span class="s1">&#39; as genbank file&#39;</span><span class="p">)</span>
        <span class="n">SeqRecord</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fname</span><span class="p">,</span> <span class="s1">&#39;genbank&#39;</span><span class="p">)</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;failed to read&#39;</span><span class="p">,</span><span class="n">fname</span><span class="p">,</span><span class="s1">&#39; as genbank file&#39;</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;reading &#39;</span><span class="p">,</span><span class="n">fname</span><span class="p">,</span><span class="s1">&#39; as snapgene file&#39;</span><span class="p">)</span>
            <span class="n">SeqRecord</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fname</span><span class="p">,</span> <span class="s1">&#39;snapgene&#39;</span><span class="p">)</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;failed to read&#39;</span><span class="p">,</span><span class="n">fname</span><span class="p">,</span><span class="s1">&#39; as snapgene file&#39;</span><span class="p">)</span>
        
    <span class="c1"># add some info to the genbank files</span>
    <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
        <span class="c1"># find the label ordering is the label that takes precedance</span>
        <span class="n">name</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ApEinfo_label&#39;</span><span class="p">,</span><span class="s1">&#39;label&#39;</span><span class="p">,</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">,</span><span class="s1">&#39;gene&#39;</span><span class="p">]</span>
        <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;unknown&#39;</span>
        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">name</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="n">k</span><span class="p">])</span> <span class="o">==</span> <span class="nb">list</span><span class="p">:</span>
                    <span class="n">label</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="n">k</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">elif</span> <span class="nb">type</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="n">k</span><span class="p">])</span> <span class="o">==</span> <span class="nb">str</span><span class="p">:</span>
                    <span class="n">label</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">label</span><span class="p">]</span>

    <span class="c1"># correct genbank read issue in parsing features with CompoundLocations which wrap around the origin</span>
    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">)):</span>
        <span class="n">d</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="vm">__dict__</span> 
        <span class="k">if</span> <span class="s1">&#39;parts&#39;</span> <span class="ow">in</span> <span class="n">d</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">d</span><span class="p">[</span><span class="s1">&#39;parts&#39;</span><span class="p">])</span><span class="o">==</span><span class="mi">2</span><span class="p">:</span>
            <span class="c1"># if ordering is wrong, swap it</span>
            <span class="k">if</span> <span class="n">d</span><span class="p">[</span><span class="s1">&#39;parts&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">start</span><span class="o">==</span><span class="mi">0</span> <span class="ow">and</span> <span class="n">d</span><span class="p">[</span><span class="s1">&#39;parts&#39;</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">end</span><span class="o">==</span><span class="nb">len</span><span class="p">(</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span> <span class="ow">and</span> <span class="n">d</span><span class="p">[</span><span class="s1">&#39;parts&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="n">d</span><span class="p">[</span><span class="s1">&#39;parts&#39;</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">strand</span><span class="p">:</span>
                <span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="vm">__dict__</span><span class="p">[</span><span class="s1">&#39;parts&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">d</span><span class="p">[</span><span class="s1">&#39;parts&#39;</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span> <span class="n">d</span><span class="p">[</span><span class="s1">&#39;parts&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span>

    <span class="c1"># create the Plasmid dataframe</span>
    <span class="k">return</span> <span class="n">Plasmid</span><span class="p">(</span><span class="n">SeqRecord</span><span class="p">)</span></div>

<div class="viewcode-block" id="read_fasta"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.read_fasta">[docs]</a><span class="k">def</span> <span class="nf">read_fasta</span><span class="p">(</span><span class="n">fname</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Read sequences from csv or fasta files</span>
<span class="sd">    fname = file to read</span>
<span class="sd">    returns list of Plasmid objects with data loaded</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">read_to_df</span><span class="p">(</span><span class="n">fname</span><span class="p">)</span>
    <span class="n">out</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1"># generate list of plasmids</span>
    <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">df</span><span class="p">[[</span><span class="s1">&#39;name&#39;</span><span class="p">,</span><span class="s1">&#39;sequence&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span><span class="p">:</span> <span class="c1"># todo test debug</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">Plasmid</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">name</span><span class="p">,</span> <span class="n">sequence</span><span class="o">=</span><span class="n">seq</span><span class="p">)</span>
        <span class="n">out</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">out</span></div>

<div class="viewcode-block" id="Plasmid"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid">[docs]</a><span class="k">class</span> <span class="nc">Plasmid</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    This class holds data and functions for manipulating sequence records read from a genbank file</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="c1"># default color palette</span>
    <span class="n">params</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="c1"># default colors for fluorescent proteins</span>
    <span class="n">default_annotations</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;molecule_type&#39;</span><span class="p">:</span><span class="s1">&#39;DNA&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;topology&#39;</span><span class="p">:</span><span class="s1">&#39;circular&#39;</span><span class="p">}</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">seq</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Initialize some data into the Plasmid dataframe</span>
<span class="sd">        seq = Bio.SeqRecord or string</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="c1"># initialize the Bio.SeqRecord object</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="o">==</span> <span class="nb">str</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">SeqRecord</span><span class="p">(</span><span class="n">seq</span><span class="o">=</span><span class="n">Bio</span><span class="o">.</span><span class="n">Seq</span><span class="o">.</span><span class="n">Seq</span><span class="p">(</span><span class="n">seq</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span> <span class="o">=</span> <span class="n">seq</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;error object is not a Bio.SeqRecord or String&#39;</span><span class="p">)</span>
        
        <span class="c1"># set default cmap</span>
        <span class="n">features</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;rbs&#39;</span><span class="p">,</span><span class="s1">&#39;cds&#39;</span><span class="p">,</span><span class="s1">&#39;promoter&#39;</span><span class="p">,</span><span class="s1">&#39;misc_binding&#39;</span><span class="p">,</span><span class="s1">&#39;ncRNA&#39;</span><span class="p">,</span><span class="s1">&#39;terminator&#39;</span><span class="p">,</span><span class="s1">&#39;primer_bind&#39;</span><span class="p">,</span><span class="s1">&#39;unknown&#39;</span><span class="p">,</span><span class="s1">&#39;rep_origin&#39;</span><span class="p">,</span><span class="s1">&#39;protein_bind&#39;</span><span class="p">]</span>
        <span class="n">palette</span> <span class="o">=</span> <span class="s1">&#39;Category20_20&#39;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;colors&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;colors&#39;</span><span class="p">][</span><span class="s1">&#39;features&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">Graphic</span><span class="o">.</span><span class="n">get_colormap</span><span class="p">(</span><span class="n">features</span><span class="p">,</span> <span class="n">palette</span><span class="p">)</span>
        <span class="c1"># set default color for known or unknown feature or label</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;colors&#39;</span><span class="p">][</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;bfp&#39;</span><span class="p">:</span><span class="s1">&#39;blue&#39;</span><span class="p">,</span><span class="s1">&#39;cfp&#39;</span><span class="p">:</span><span class="s1">&#39;dodgerblue&#39;</span><span class="p">,</span><span class="s1">&#39;gfp&#39;</span><span class="p">:</span><span class="s1">&#39;green&#39;</span><span class="p">,</span><span class="s1">&#39;yfp&#39;</span><span class="p">:</span><span class="s1">&#39;gold&#39;</span><span class="p">,</span><span class="s1">&#39;rfp&#39;</span><span class="p">:</span><span class="s1">&#39;red&#39;</span><span class="p">,</span><span class="s1">&#39;mirfp&#39;</span><span class="p">:</span><span class="s1">&#39;purple&#39;</span><span class="p">}</span>

        <span class="c1"># update the feature list ordering</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reset_colors</span><span class="p">(</span><span class="n">ow_type</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ow_locus</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<div class="viewcode-block" id="Plasmid.reset_colors"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.reset_colors">[docs]</a>    <span class="k">def</span> <span class="nf">reset_colors</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ow_type</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ow_locus</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Apply default or custom color map to features or labels</span>
<span class="sd">        ow_type = overwrite existing colors for features</span>
<span class="sd">        ow_locus = overwrite existing colors for locus_tags</span>
<span class="sd">        inplace = apply inplace</span>
<span class="sd">        return Plasmid with new colors</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="n">label_cmap</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;colors&#39;</span><span class="p">][</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">]</span>
        <span class="n">ftype_cmap</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;colors&#39;</span><span class="p">][</span><span class="s1">&#39;features&#39;</span><span class="p">]</span>
        <span class="n">default_color</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;colors&#39;</span><span class="p">][</span><span class="s1">&#39;features&#39;</span><span class="p">][</span><span class="s1">&#39;unknown&#39;</span><span class="p">]</span>
        <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">inplace</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="n">inplace</span><span class="p">)</span>
        <span class="c1"># update colors in SeqRecord</span>
        <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
            <span class="c1"># check for label color</span>
            <span class="n">label_color</span> <span class="o">=</span> <span class="n">search_key</span><span class="p">(</span><span class="n">label_cmap</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">rev</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="c1"># check for feature color</span>
            <span class="n">feature_color</span> <span class="o">=</span> <span class="n">search_key</span><span class="p">(</span><span class="n">ftype_cmap</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">type</span><span class="p">)</span>
            <span class="c1"># check for saved colors</span>
            <span class="n">fwd_color</span> <span class="o">=</span> <span class="n">search_key</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">,</span> <span class="s1">&#39;ApEinfo_fwdcolor&#39;</span><span class="p">)</span>
            <span class="n">rev_color</span> <span class="o">=</span> <span class="n">search_key</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">,</span> <span class="s1">&#39;ApEinfo_revcolor&#39;</span><span class="p">)</span>
            <span class="c1"># use label_color if available</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">ow_locus</span> <span class="ow">or</span> <span class="p">(</span><span class="n">fwd_color</span><span class="o">==</span><span class="kc">None</span> <span class="ow">and</span> <span class="n">rev_color</span><span class="o">==</span><span class="kc">None</span><span class="p">))</span> <span class="ow">and</span> <span class="n">label_color</span><span class="o">!=</span><span class="kc">None</span><span class="p">:</span>
                <span class="n">Plasmid</span><span class="o">.</span><span class="n">set_color</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">label_color</span><span class="p">)</span>
            <span class="c1"># use feature color if available</span>
            <span class="k">elif</span> <span class="p">(</span><span class="n">ow_type</span> <span class="ow">or</span> <span class="p">(</span><span class="n">fwd_color</span><span class="o">==</span><span class="kc">None</span> <span class="ow">and</span> <span class="n">rev_color</span><span class="o">==</span><span class="kc">None</span><span class="p">))</span> <span class="ow">and</span> <span class="n">feature_color</span><span class="o">!=</span><span class="kc">None</span><span class="p">:</span>
                <span class="n">Plasmid</span><span class="o">.</span><span class="n">set_color</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">feature_color</span><span class="p">)</span>
            <span class="c1"># apply default colors if nothing found</span>
            <span class="k">elif</span> <span class="n">fwd_color</span><span class="o">==</span><span class="kc">None</span> <span class="ow">or</span> <span class="n">rev_color</span><span class="o">==</span><span class="kc">None</span><span class="p">:</span>
                <span class="n">Plasmid</span><span class="o">.</span><span class="n">set_color</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">default_color</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">out</span></div>

<div class="viewcode-block" id="Plasmid.set_color"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.set_color">[docs]</a>    <span class="k">def</span> <span class="nf">set_color</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">color</span><span class="p">):</span>
        <span class="c1"># sets the color of the gene</span>
        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">color</span><span class="p">)</span> <span class="o">==</span> <span class="nb">list</span><span class="p">:</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s1">&#39;ApEinfo_fwdcolor&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">color</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s1">&#39;ApEinfo_revcolor&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">color</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]]</span>
        <span class="k">elif</span> <span class="nb">type</span><span class="p">(</span><span class="n">color</span><span class="p">)</span> <span class="o">==</span> <span class="nb">str</span><span class="p">:</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s1">&#39;ApEinfo_fwdcolor&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">color</span><span class="p">]</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s1">&#39;ApEinfo_revcolor&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">color</span><span class="p">]</span></div>

<div class="viewcode-block" id="Plasmid.set_column"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.set_column">[docs]</a>    <span class="k">def</span> <span class="nf">set_column</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">key</span><span class="p">,</span> <span class="n">col</span><span class="p">,</span> <span class="n">value</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Sets columns of Plasmid dataframe</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="n">key</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">unpack_key</span><span class="p">(</span><span class="n">key</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">value</span><span class="p">)</span><span class="o">!=</span> <span class="nb">list</span><span class="p">:</span>
            <span class="n">value</span> <span class="o">=</span> <span class="p">[</span><span class="n">value</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">key</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">key</span><span class="p">)):</span>
            <span class="n">k</span> <span class="o">=</span> <span class="n">key</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">val</span> <span class="o">=</span> <span class="n">value</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="c1"># set color values of each feature</span>
            <span class="k">if</span> <span class="n">col</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;color&#39;</span><span class="p">,</span><span class="s1">&#39;fwd_color&#39;</span><span class="p">,</span><span class="s1">&#39;rev_color&#39;</span><span class="p">]:</span>
                <span class="k">if</span> <span class="n">col</span><span class="o">==</span><span class="s1">&#39;color&#39;</span><span class="p">:</span>
                    <span class="n">strand</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">strand</span>
                <span class="k">elif</span> <span class="n">col</span><span class="o">==</span><span class="s1">&#39;fwd_color&#39;</span><span class="p">:</span>
                    <span class="n">strand</span> <span class="o">=</span> <span class="mi">1</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">strand</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
                <span class="k">if</span> <span class="n">strand</span><span class="o">==</span><span class="mi">1</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s1">&#39;ApEinfo_fwdcolor&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">val</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s1">&#39;ApEinfo_revcolor&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">val</span><span class="p">]</span>
            <span class="k">elif</span> <span class="n">col</span><span class="o">==</span><span class="s1">&#39;location&#39;</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">.</span><span class="n">location</span> <span class="o">=</span> <span class="n">val</span>
            <span class="k">elif</span> <span class="n">col</span><span class="o">==</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">val</span><span class="p">]</span>
            <span class="k">elif</span> <span class="n">col</span><span class="o">==</span><span class="s1">&#39;type&#39;</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="n">val</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">k</span><span class="p">,</span><span class="s1">&#39; is not in column&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></div>

<div class="viewcode-block" id="Plasmid.unpack_key"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.unpack_key">[docs]</a>    <span class="k">def</span> <span class="nf">unpack_key</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">key</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Formats key into a list of values</span>
<span class="sd">        key = slice, range, bool array, int array, or list</span>
<span class="sd">        Returns a list of values</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">key</span><span class="p">)</span> <span class="o">==</span> <span class="nb">int</span><span class="p">:</span>
            <span class="k">return</span> <span class="p">[</span><span class="n">key</span><span class="p">]</span>
        <span class="k">elif</span> <span class="nb">type</span><span class="p">(</span><span class="n">key</span><span class="p">)</span> <span class="o">==</span> <span class="nb">slice</span> <span class="ow">or</span> <span class="nb">type</span><span class="p">(</span><span class="n">key</span><span class="p">)</span><span class="o">==</span><span class="nb">range</span><span class="p">:</span>
            <span class="n">start</span> <span class="o">=</span> <span class="n">key</span><span class="o">.</span><span class="n">start</span>
            <span class="n">stop</span> <span class="o">=</span> <span class="n">key</span><span class="o">.</span><span class="n">stop</span>
            <span class="k">if</span> <span class="n">start</span><span class="o">==</span><span class="kc">None</span><span class="p">:</span>
                <span class="n">start</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">if</span> <span class="n">stop</span><span class="o">==</span><span class="kc">None</span><span class="p">:</span>
                <span class="n">stop</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">)</span> 
            <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">))</span>
        <span class="k">elif</span> <span class="n">isIntArray</span><span class="p">(</span><span class="n">key</span><span class="p">):</span>
            <span class="k">return</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">k</span><span class="p">)</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">key</span><span class="p">]</span>
        <span class="k">elif</span> <span class="n">isBool</span><span class="p">(</span><span class="n">key</span><span class="p">):</span>
            <span class="n">key</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">key</span><span class="p">))[</span><span class="n">key</span><span class="p">]</span>
            <span class="k">return</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">k</span><span class="p">)</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">key</span><span class="p">]</span>
        <span class="k">elif</span> <span class="nb">type</span><span class="p">(</span><span class="n">key</span><span class="p">)</span> <span class="o">==</span> <span class="nb">list</span><span class="p">:</span>
            <span class="n">out</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">key</span><span class="p">:</span>
                <span class="n">out</span><span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">unpack_key</span><span class="p">(</span><span class="n">k</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">out</span></div>

<div class="viewcode-block" id="Plasmid.update"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.update">[docs]</a>    <span class="k">def</span> <span class="nf">update</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Synchronize entries from SeqRecord to Plasmid parts list</span>
<span class="sd">        Run sort_features and sort gene features by starting base index</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="k">if</span> <span class="n">inplace</span><span class="p">:</span>
            <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">out</span><span class="o">.</span><span class="n">sort_features</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="c1"># sync gene labels for backward compatibility with ApE</span>
        <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
            <span class="n">label</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s1">&#39;ApEinfo_label&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">label</span><span class="p">]</span>
            <span class="c1"># delete useless tags</span>
            <span class="n">tags</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;color&#39;</span><span class="p">,</span><span class="s1">&#39;label&#39;</span><span class="p">,</span><span class="s1">&#39;ApEinfo_graphicformat&#39;</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">tags</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">:</span>
                    <span class="k">del</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="n">t</span><span class="p">])</span>
        <span class="n">out</span><span class="o">.</span><span class="n">check_annotations</span><span class="p">()</span>
        <span class="k">return</span> <span class="n">out</span></div>
    
<div class="viewcode-block" id="Plasmid.sort_features"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.sort_features">[docs]</a>    <span class="k">def</span> <span class="nf">sort_features</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Sorts the features in the SeqRecord object</span>
<span class="sd">        returns a modified copy of the object</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="k">if</span> <span class="n">inplace</span><span class="p">:</span>
            <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="c1"># do sorting with pandas</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">out</span><span class="o">.</span><span class="n">to_dataframe</span><span class="p">()</span>
        <span class="n">idx</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">,</span><span class="s1">&#39;end&#39;</span><span class="p">,</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">,</span><span class="s1">&#39;type&#39;</span><span class="p">,</span><span class="s1">&#39;color&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">index</span>
        <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="p">[</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">idx</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">out</span></div>

<div class="viewcode-block" id="Plasmid.check_annotations"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.check_annotations">[docs]</a>    <span class="k">def</span> <span class="nf">check_annotations</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Check features for consistency against the sequence length</span>
<span class="sd">        returns false if there are issues with the gene annotation</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="c1"># check seq length consistency</span>
        <span class="n">seq_length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
        <span class="n">good</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
            <span class="n">start</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">start</span>
            <span class="n">end</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">end</span>
            <span class="k">if</span> <span class="n">start</span> <span class="o">&gt;</span> <span class="n">seq_length</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Warning: annotation issues with &#39;</span><span class="o">+</span><span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39; start &lt; 0&#39;</span><span class="p">)</span>
                <span class="n">good</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="k">if</span> <span class="n">end</span> <span class="o">&gt;</span> <span class="n">seq_length</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Warning: annotation issues with &#39;</span><span class="o">+</span><span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39; end &gt; seq_length&#39;</span><span class="p">)</span>
                <span class="n">good</span> <span class="o">=</span> <span class="kc">False</span>
        
        <span class="c1"># check header consistency</span>
        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">default_annotations</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="k">if</span> <span class="o">~</span><span class="p">(</span><span class="n">k</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">default_annotations</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>
        
        <span class="k">if</span> <span class="n">good</span><span class="o">==</span><span class="kc">False</span><span class="p">:</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">error</span><span class="p">()</span>
        <span class="k">return</span> <span class="n">good</span></div>

<div class="viewcode-block" id="Plasmid.to_genbank"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.to_genbank">[docs]</a>    <span class="k">def</span> <span class="nf">to_genbank</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename</span><span class="p">,</span> <span class="n">timestamp</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Writes the Plasmid SeqRecord to a filename</span>
<span class="sd">        filename = filename of output file</span>
<span class="sd">        timestamp = adds a timestamp to the output file</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="k">if</span> <span class="n">timestamp</span><span class="p">:</span>
            <span class="n">filename</span><span class="o">+=</span> <span class="n">get_timestamp</span><span class="p">()</span>
        <span class="c1"># sync info and update it</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span><span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">Bio</span><span class="o">.</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="p">,</span> <span class="n">f</span><span class="p">,</span> <span class="s1">&#39;genbank&#39;</span><span class="p">)</span></div>
    
<div class="viewcode-block" id="Plasmid.copy"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.copy">[docs]</a>    <span class="k">def</span> <span class="nf">copy</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Returns a copy of self</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="k">return</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></div>

<div class="viewcode-block" id="Plasmid.inplace"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.inplace">[docs]</a>    <span class="k">def</span> <span class="nf">inplace</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">inplace</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Decides if a copy of self should be returned</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="k">if</span> <span class="n">inplace</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span></div>

<div class="viewcode-block" id="Plasmid.get_best_origin"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.get_best_origin">[docs]</a>    <span class="k">def</span> <span class="nf">get_best_origin</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Find best index to set the origin of the Plasmid</span>
<span class="sd">        return index to set the origin</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="n">genes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">genes</span><span class="p">)):</span>
            <span class="p">[</span><span class="n">s1</span><span class="p">,</span> <span class="n">s2</span><span class="p">,</span> <span class="n">strand</span><span class="p">]</span> <span class="o">=</span> <span class="n">unwrap_FeatureLocation</span><span class="p">(</span><span class="n">genes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>
            <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="p">[</span><span class="n">s1</span><span class="p">,</span> <span class="n">s2</span><span class="p">]:</span>
                <span class="n">good</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">genes</span><span class="p">)):</span>
                    <span class="p">[</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">strand</span><span class="p">]</span> <span class="o">=</span> <span class="n">unwrap_FeatureLocation</span><span class="p">(</span><span class="n">genes</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>
                    <span class="c1"># break out of loop if origin is not in good location</span>
                    <span class="k">if</span> <span class="p">(</span><span class="n">end</span> <span class="o">&gt;</span> <span class="n">start</span> <span class="ow">and</span> <span class="n">k</span> <span class="o">&gt;</span> <span class="n">start</span> <span class="ow">and</span> <span class="n">k</span> <span class="o">&lt;</span> <span class="n">end</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">start</span> <span class="o">&gt;</span> <span class="n">end</span> <span class="ow">and</span> <span class="p">(</span><span class="n">k</span> <span class="o">&gt;</span> <span class="n">start</span> <span class="ow">or</span> <span class="n">k</span> <span class="o">&lt;</span> <span class="n">end</span><span class="p">)):</span>
                        <span class="n">good</span> <span class="o">=</span> <span class="kc">False</span>
                        <span class="k">break</span>
                <span class="k">if</span> <span class="n">good</span><span class="p">:</span>
                    <span class="k">return</span> <span class="n">k</span></div>

<div class="viewcode-block" id="Plasmid.set_origin"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.set_origin">[docs]</a>    <span class="k">def</span> <span class="nf">set_origin</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Sets a new position for the origin coordinates of a circular Plasmid</span>
<span class="sd">        index = base index to set new origin. If index==None, then origin is set at the first contiguous genetic feature</span>
<span class="sd">        inplace = performs operation inplace or returns a copy with the modified data</span>
<span class="sd">        returns the Plasmid dataframe</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">inplace</span><span class="p">(</span><span class="n">inplace</span><span class="p">)</span>
        <span class="c1"># find best index to set the origin</span>
        <span class="k">if</span> <span class="n">index</span><span class="o">==</span><span class="kc">None</span><span class="p">:</span>
            <span class="n">index</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_best_origin</span><span class="p">()</span>
        <span class="c1"># do nothing if better index is not found </span>
        <span class="k">if</span> <span class="n">index</span><span class="o">==</span><span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">out</span>
        <span class="k">elif</span> <span class="n">index</span> <span class="o">&gt;</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">):</span>
            <span class="n">index</span> <span class="o">=</span> <span class="n">index</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">index</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">index</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span> <span class="o">+</span> <span class="n">index</span>
        <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">Seq</span><span class="o">.</span><span class="n">Seq</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="n">index</span><span class="p">:])</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">index</span><span class="p">]))</span>
        <span class="n">out</span><span class="o">.</span><span class="n">shift_features</span><span class="p">(</span><span class="o">-</span><span class="n">index</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">out</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">out</span></div>

<div class="viewcode-block" id="Plasmid.to_graphic"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.to_graphic">[docs]</a>    <span class="k">def</span> <span class="nf">to_graphic</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">rec</span> <span class="o">=</span> <span class="n">Graphic</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="bp">self</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">rec</span></div>
 
<div class="viewcode-block" id="Plasmid.reverse_complement"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.reverse_complement">[docs]</a>    <span class="k">def</span> <span class="nf">reverse_complement</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Returns reverse complement of the Plasmid</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="k">if</span> <span class="n">inplace</span><span class="p">:</span>
            <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span> <span class="o">=</span> <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
        <span class="k">return</span> <span class="n">out</span> </div>

<div class="viewcode-block" id="Plasmid.insert_sequence"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.insert_sequence">[docs]</a>    <span class="k">def</span> <span class="nf">insert_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">,</span> <span class="n">seq</span><span class="p">,</span> <span class="n">disrupt</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Insert sequences into the Plasmid</span>
<span class="sd">        index = position to insert the sequence or Plasmid construct</span>
<span class="sd">        seq = sequence or linearized Plasmid construct to insert into the Plasmid</span>
<span class="sd">        disrupt = if sequence is inserted into an existing feature, break the feature into two subsequences</span>
<span class="sd">        inplace = modifications are done inplace</span>
<span class="sd">        return the modified Plasmid</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="c1"># format to integer</span>
        <span class="n">index</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">index</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">inplace</span><span class="p">:</span>
            <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        
        <span class="c1"># wrap location </span>
        <span class="k">if</span> <span class="n">index</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">index</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span> <span class="o">-</span> <span class="n">index</span>
        <span class="k">elif</span> <span class="n">index</span> <span class="o">&gt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">):</span>
            <span class="n">index</span> <span class="o">=</span> <span class="n">index</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
        
        <span class="c1"># format the values</span>
        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span><span class="o">==</span><span class="nb">str</span><span class="p">:</span>
            <span class="n">sequence</span> <span class="o">=</span> <span class="n">seq</span>
            <span class="n">insert_features</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">elif</span> <span class="nb">type</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span><span class="o">==</span><span class="s1">&#39;Plasmid&#39;</span><span class="p">:</span>
            <span class="n">sequence</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
            <span class="n">insert_features</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">shift_features</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;insert_sequence: no sequence provided&#39;</span><span class="p">)</span>
        
        <span class="c1"># modify the sequence</span>
        <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">Seq</span><span class="o">.</span><span class="n">Seq</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">[:</span><span class="n">index</span><span class="p">])</span> <span class="o">+</span> <span class="n">sequence</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="n">index</span><span class="p">:]))</span>
        <span class="c1"># shift features around</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">)):</span>
            <span class="c1"># handle compound locations</span>
            <span class="k">if</span> <span class="s1">&#39;parts&#39;</span> <span class="ow">in</span> <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="vm">__dict__</span><span class="p">:</span>
                <span class="n">loc</span> <span class="o">=</span> <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">location</span>
                <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">location</span> <span class="o">=</span> <span class="n">shift_CompoundLocation</span><span class="p">(</span><span class="n">loc</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">orig_len</span><span class="p">)</span>
                <span class="n">f</span> <span class="o">=</span> <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="vm">__dict__</span><span class="p">[</span><span class="s1">&#39;parts&#39;</span><span class="p">]</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">f</span><span class="p">)):</span>
                    <span class="n">start</span> <span class="o">=</span> <span class="n">f</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">start</span>
                    <span class="n">end</span> <span class="o">=</span> <span class="n">f</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">end</span>
                    <span class="n">strand</span> <span class="o">=</span> <span class="n">f</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">strand</span>
                    <span class="k">if</span> <span class="n">start</span> <span class="o">&gt;=</span> <span class="n">index</span><span class="p">:</span>
                        <span class="n">start</span><span class="o">+=</span><span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">end</span> <span class="o">&gt;</span> <span class="n">index</span><span class="p">:</span>
                        <span class="n">end</span><span class="o">+=</span><span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
                    <span class="c1"># special condition for genes which wrap around the origin debug</span>
                    <span class="k">elif</span> <span class="n">i</span><span class="o">+</span><span class="mi">1</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">f</span><span class="p">)</span> <span class="ow">and</span> <span class="n">end</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span><span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span> <span class="ow">and</span> <span class="n">f</span><span class="p">[</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">start</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                        <span class="n">end</span><span class="o">+=</span><span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
                    <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="vm">__dict__</span><span class="p">[</span><span class="s1">&#39;parts&#39;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqFeature</span><span class="o">.</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">strand</span><span class="p">)</span>
            <span class="c1"># handle single locations</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">start</span> <span class="o">=</span> <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">start</span>
                <span class="n">end</span> <span class="o">=</span> <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">end</span>
                <span class="n">strand</span> <span class="o">=</span> <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">strand</span>
                <span class="k">if</span> <span class="n">start</span> <span class="o">&gt;=</span> <span class="n">index</span><span class="p">:</span>
                    <span class="n">start</span><span class="o">+=</span><span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">end</span> <span class="o">&gt;</span> <span class="n">index</span><span class="p">:</span>
                    <span class="n">end</span><span class="o">+=</span><span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
                <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">location</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqFeature</span><span class="o">.</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">strand</span><span class="p">)</span>

        <span class="c1"># record disruptions to existing features</span>
        <span class="n">keep</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="n">disrupt</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">)):</span>
                <span class="c1"># handle compound locations</span>
                <span class="k">if</span> <span class="s1">&#39;parts&#39;</span> <span class="ow">in</span> <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="vm">__dict__</span><span class="p">:</span>
                    <span class="n">f</span> <span class="o">=</span> <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="vm">__dict__</span><span class="p">[</span><span class="s1">&#39;parts&#39;</span><span class="p">]</span>
                    <span class="c1"># partition sublocation to before or after the insert</span>
                    <span class="n">new</span> <span class="o">=</span> <span class="p">[]</span>
                    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">f</span><span class="p">)):</span>
                        <span class="n">start</span> <span class="o">=</span> <span class="n">f</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">start</span>
                        <span class="n">end</span> <span class="o">=</span> <span class="n">f</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">end</span>
                        <span class="n">strand</span> <span class="o">=</span> <span class="n">f</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">strand</span>
                        <span class="c1"># split the subsequences, detect that genes wrapping around origin were disrupted</span>
                        <span class="c1">#c1 = (start &lt;= index) and (end &gt; index + len(sequence))</span>
                        <span class="c1">#c2 = (start &lt; index) and (end &gt;= index + len(sequence))</span>
                        <span class="n">c3</span> <span class="o">=</span> <span class="p">(</span><span class="n">start</span> <span class="o">&gt;=</span> <span class="n">index</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">))</span> <span class="ow">or</span> <span class="p">(</span><span class="n">end</span> <span class="o">&lt;=</span> <span class="n">index</span><span class="p">)</span>
                        <span class="n">isDisrupted</span> <span class="o">=</span> <span class="p">(</span><span class="n">c3</span><span class="o">==</span><span class="kc">False</span><span class="p">)</span>
                        <span class="k">if</span> <span class="n">isDisrupted</span><span class="p">:</span> 
                            <span class="k">if</span> <span class="n">start</span> <span class="o">&lt;</span> <span class="n">index</span><span class="p">:</span> <span class="c1"># special logic </span>
                                <span class="n">new</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Bio</span><span class="o">.</span><span class="n">SeqFeature</span><span class="o">.</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">index</span><span class="p">,</span> <span class="n">strand</span><span class="p">))</span>
                            <span class="k">if</span> <span class="n">end</span> <span class="o">&gt;</span> <span class="n">index</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">):</span>
                                <span class="n">new</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Bio</span><span class="o">.</span><span class="n">SeqFeature</span><span class="o">.</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="n">index</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">),</span> <span class="n">end</span><span class="p">,</span> <span class="n">strand</span><span class="p">))</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">new</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">f</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
                    <span class="c1"># flag that the gene was split</span>
                    <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="vm">__dict__</span><span class="p">[</span><span class="s1">&#39;parts&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">new</span>
                    <span class="k">if</span> <span class="n">isDisrupted</span><span class="p">:</span>
                        <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">+=</span><span class="s1">&#39; disrupted&#39;</span>
                <span class="c1"># handle single locations</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">start</span> <span class="o">=</span> <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">start</span>
                    <span class="n">end</span> <span class="o">=</span> <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">end</span>
                    <span class="n">strand</span> <span class="o">=</span> <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">strand</span>
                    <span class="c1"># split the subsequences</span>
                    <span class="k">if</span> <span class="n">start</span> <span class="o">&lt;</span> <span class="n">index</span> <span class="ow">and</span> <span class="n">end</span> <span class="o">&gt;</span> <span class="n">index</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">):</span>
                        <span class="n">loc1</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqFeature</span><span class="o">.</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">index</span><span class="p">,</span> <span class="n">strand</span><span class="p">)</span>
                        <span class="n">loc2</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqFeature</span><span class="o">.</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="n">index</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">),</span> <span class="n">end</span><span class="p">,</span> <span class="n">strand</span><span class="p">)</span>
                        <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">location</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqFeature</span><span class="o">.</span><span class="n">CompoundLocation</span><span class="p">([</span><span class="n">loc1</span><span class="p">,</span> <span class="n">loc2</span><span class="p">])</span>
                        <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">+=</span><span class="s1">&#39; disrupted&#39;</span>
        <span class="c1"># record the disruption</span>
        <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="o">+=</span><span class="n">insert_features</span>
        <span class="n">out</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">out</span></div>

<div class="viewcode-block" id="Plasmid.delete_sequence"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.delete_sequence">[docs]</a>    <span class="k">def</span> <span class="nf">delete_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">key</span><span class="p">,</span> <span class="n">disrupt</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Delete sequences on the Plasmid</span>
<span class="sd">        key = range(start, end) base pair positions to delete.</span>
<span class="sd">        disrupt = if a sequence is deleted from a feature, flag the feature which had the deletion</span>
<span class="sd">        return copy of the modified Plasmid</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">inplace</span><span class="p">(</span><span class="n">inplace</span><span class="p">)</span>
        <span class="n">d1</span> <span class="o">=</span> <span class="n">key</span><span class="o">.</span><span class="n">start</span>
        <span class="n">d2</span> <span class="o">=</span> <span class="n">key</span><span class="o">.</span><span class="n">stop</span>
        <span class="n">dlen</span> <span class="o">=</span> <span class="n">d2</span><span class="o">-</span><span class="n">d1</span>
        <span class="n">orig_len</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">dlen</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">out</span><span class="o">.</span><span class="n">delete_sequence</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">d1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">)),</span> <span class="n">disrupt</span><span class="o">=</span><span class="n">disrupt</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">out</span><span class="o">.</span><span class="n">delete_sequence</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">d2</span><span class="p">),</span> <span class="n">disrupt</span><span class="o">=</span><span class="n">disrupt</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">dlen</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">Seq</span><span class="o">.</span><span class="n">Seq</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">[:</span><span class="n">d1</span><span class="p">])</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="n">d2</span><span class="p">:]))</span>
            <span class="c1"># shift locations around</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">)):</span>
                <span class="n">isDisrupted</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="c1"># handle compound location</span>
                <span class="k">if</span> <span class="s1">&#39;parts&#39;</span> <span class="ow">in</span> <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="vm">__dict__</span><span class="p">:</span>
                    <span class="c1"># unwrap parts</span>
                    <span class="n">loc</span> <span class="o">=</span> <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">location</span>
                    <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">location</span> <span class="o">=</span> <span class="n">shift_CompoundLocation</span><span class="p">(</span><span class="n">loc</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">orig_len</span><span class="p">)</span>
                    <span class="n">f</span> <span class="o">=</span> <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="vm">__dict__</span><span class="p">[</span><span class="s1">&#39;parts&#39;</span><span class="p">]</span>
                    <span class="c1"># process each sublocation</span>
                    <span class="n">keep</span> <span class="o">=</span> <span class="p">[]</span>
                    <span class="k">for</span> <span class="n">loc</span> <span class="ow">in</span> <span class="n">f</span><span class="p">:</span>
                        <span class="n">loc</span><span class="p">,</span> <span class="n">dis</span> <span class="o">=</span> <span class="n">Plasmid</span><span class="o">.</span><span class="n">delete_location</span><span class="p">(</span><span class="n">loc</span><span class="p">,</span> <span class="n">d1</span><span class="p">,</span> <span class="n">d2</span><span class="p">)</span>
                        <span class="c1"># genes caught in the deletion area are deleted</span>
                        <span class="k">if</span> <span class="n">loc</span><span class="o">==</span><span class="kc">None</span><span class="p">:</span>
                            <span class="n">isDisrupted</span> <span class="o">=</span> <span class="kc">True</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">isDisrupted</span> <span class="o">|=</span> <span class="n">dis</span>
                            <span class="n">keep</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">loc</span><span class="p">)</span>
                    <span class="c1"># generate new location object</span>
                    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">keep</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                        <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">location</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqFeature</span><span class="o">.</span><span class="n">CompoundLocation</span><span class="p">(</span><span class="n">keep</span><span class="p">)</span>
                    <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">keep</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                        <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">location</span> <span class="o">=</span> <span class="n">keep</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">location</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="c1"># handle single locations</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">loc</span> <span class="o">=</span> <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">location</span>
                    <span class="n">loc</span><span class="p">,</span> <span class="n">dis</span> <span class="o">=</span> <span class="n">Plasmid</span><span class="o">.</span><span class="n">delete_location</span><span class="p">(</span><span class="n">loc</span><span class="p">,</span> <span class="n">d1</span><span class="p">,</span> <span class="n">d2</span><span class="p">)</span>
                    <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">location</span> <span class="o">=</span> <span class="n">loc</span>
                    <span class="n">isDisrupted</span> <span class="o">|=</span> <span class="n">dis</span>
                <span class="k">if</span> <span class="n">isDisrupted</span> <span class="ow">and</span> <span class="n">disrupt</span><span class="p">:</span>
                    <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">+=</span><span class="s1">&#39; subsequence&#39;</span>
            <span class="c1"># purge genes without valid locations</span>
            <span class="n">j</span><span class="o">=</span><span class="mi">0</span>
            <span class="k">while</span> <span class="n">j</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">location</span><span class="o">==</span><span class="kc">None</span><span class="p">:</span>
                    <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">j</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">j</span><span class="o">+=</span><span class="mi">1</span>
            <span class="n">out</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">out</span></div>

<div class="viewcode-block" id="Plasmid.delete_location"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.delete_location">[docs]</a>    <span class="k">def</span> <span class="nf">delete_location</span><span class="p">(</span><span class="n">loc</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Checks if feature location should be deleted or shifted</span>
<span class="sd">        loc = Bio.SeqFeature.FeatureLocation object to be analyzed</span>
<span class="sd">        start = start position of deletion</span>
<span class="sd">        stop = end position of deletion</span>
<span class="sd">        </span>
<span class="sd">        returns loc, isDisrupted</span>
<span class="sd">        loc = updated FeatureLocation object</span>
<span class="sd">        isDisrupted = bool indicating features were disrupted or not</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="n">dlen</span> <span class="o">=</span> <span class="n">stop</span> <span class="o">-</span> <span class="n">start</span>
        <span class="n">s1</span> <span class="o">=</span> <span class="n">loc</span><span class="o">.</span><span class="n">start</span>
        <span class="n">s2</span> <span class="o">=</span> <span class="n">loc</span><span class="o">.</span><span class="n">end</span>
        <span class="n">strand</span> <span class="o">=</span> <span class="n">loc</span><span class="o">.</span><span class="n">strand</span>
        <span class="n">isDisrupted</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="c1"># genes above the area need to shift location</span>
        <span class="k">if</span> <span class="n">s1</span> <span class="o">&gt;=</span> <span class="n">stop</span> <span class="ow">and</span> <span class="n">s2</span> <span class="o">&gt;</span> <span class="n">stop</span><span class="p">:</span>
            <span class="n">loc</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqFeature</span><span class="o">.</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="n">s1</span> <span class="o">-</span> <span class="n">dlen</span><span class="p">,</span> <span class="n">s2</span> <span class="o">-</span> <span class="n">dlen</span><span class="p">,</span> <span class="n">strand</span><span class="p">)</span>
            <span class="n">isDisrupted</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="c1"># genes caught in the rear half are truncated</span>
        <span class="k">elif</span> <span class="n">s1</span> <span class="o">&lt;</span> <span class="n">start</span> <span class="ow">and</span> <span class="n">s2</span> <span class="o">&gt;</span> <span class="n">start</span>  <span class="ow">and</span> <span class="n">s2</span> <span class="o">&lt;=</span> <span class="n">stop</span><span class="p">:</span>
            <span class="n">loc</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqFeature</span><span class="o">.</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="n">s1</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">strand</span><span class="p">)</span>
        <span class="c1"># genes caught in front half are truncated</span>
        <span class="k">elif</span> <span class="n">s1</span> <span class="o">&lt;</span> <span class="n">stop</span> <span class="ow">and</span> <span class="n">s1</span> <span class="o">&gt;=</span> <span class="n">start</span>  <span class="ow">and</span> <span class="n">s2</span> <span class="o">&gt;</span> <span class="n">stop</span><span class="p">:</span>
            <span class="n">loc</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqFeature</span><span class="o">.</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">s2</span> <span class="o">-</span> <span class="n">dlen</span><span class="p">,</span> <span class="n">strand</span><span class="p">)</span>
        <span class="c1"># gene overlapping the deletion area truncated in middle</span>
        <span class="k">elif</span> <span class="n">s1</span> <span class="o">&lt;</span> <span class="n">start</span> <span class="ow">and</span> <span class="n">s2</span> <span class="o">&gt;</span> <span class="n">stop</span><span class="p">:</span>
            <span class="n">loc</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqFeature</span><span class="o">.</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="n">s1</span><span class="p">,</span> <span class="n">s2</span> <span class="o">-</span> <span class="n">dlen</span><span class="p">,</span> <span class="n">strand</span><span class="p">)</span>
        <span class="c1"># genes caught in the deletion area are deleted</span>
        <span class="k">elif</span> <span class="n">s1</span> <span class="o">&gt;=</span> <span class="n">start</span> <span class="ow">and</span> <span class="n">s2</span> <span class="o">&lt;=</span> <span class="n">stop</span><span class="p">:</span>
            <span class="n">loc</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">return</span> <span class="n">loc</span><span class="p">,</span> <span class="n">isDisrupted</span></div>

<div class="viewcode-block" id="Plasmid.replace_sequence"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.replace_sequence">[docs]</a>    <span class="k">def</span> <span class="nf">replace_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span><span class="p">,</span> <span class="n">disrupt</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Replace a sequence at a given location with a new sequence</span>
<span class="sd">        key = range(start,end) base pair positions on the Plasmid to replace the sequence</span>
<span class="sd">        value = a Plasmid dataframe, which contains one sequence and features associated with this sequence</span>
<span class="sd">        disrupt = disrupt existing genetic features. Meaning if a gene is caught in the deletion zone, it will be truncated.</span>
<span class="sd">        inplace = perform modifications inplace</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">inplace</span><span class="p">(</span><span class="n">inplace</span><span class="p">)</span>
        <span class="c1"># insert the new sequence</span>
        <span class="n">start</span> <span class="o">=</span> <span class="n">key</span><span class="o">.</span><span class="n">start</span>
        <span class="n">stop</span> <span class="o">=</span> <span class="n">key</span><span class="o">.</span><span class="n">stop</span>
        <span class="n">out</span><span class="o">.</span><span class="n">insert_sequence</span><span class="p">(</span><span class="n">key</span><span class="o">.</span><span class="n">stop</span><span class="p">,</span> <span class="n">value</span><span class="p">,</span> <span class="n">disrupt</span><span class="o">=</span><span class="n">disrupt</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="c1"># delete the sequence</span>
        <span class="n">out</span><span class="o">.</span><span class="n">delete_sequence</span><span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="n">disrupt</span><span class="o">=</span><span class="n">disrupt</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">out</span></div>

<div class="viewcode-block" id="Plasmid.splice_sequence"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.splice_sequence">[docs]</a>    <span class="k">def</span> <span class="nf">splice_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">key</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        splice a sequence from the Plasmid</span>
<span class="sd">        key = range(start, end) or slice(start, end) base pair position to cut Plasmid</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">inplace</span><span class="p">(</span><span class="n">inplace</span><span class="p">)</span>
        <span class="n">start</span> <span class="o">=</span> <span class="n">key</span><span class="o">.</span><span class="n">start</span>
        <span class="n">stop</span> <span class="o">=</span> <span class="n">key</span><span class="o">.</span><span class="n">stop</span>
        <span class="n">length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
        <span class="c1"># trim the sequence</span>
        <span class="k">if</span> <span class="n">stop</span> <span class="o">&lt;</span> <span class="n">start</span><span class="p">:</span>
            <span class="n">out</span><span class="o">.</span><span class="n">set_origin</span><span class="p">(</span><span class="n">stop</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">out</span><span class="o">.</span><span class="n">delete_sequence</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">start</span><span class="o">-</span><span class="n">stop</span><span class="p">),</span> <span class="n">disrupt</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">out</span><span class="o">.</span><span class="n">delete_sequence</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">stop</span><span class="p">,</span> <span class="n">length</span><span class="p">),</span> <span class="n">disrupt</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">out</span><span class="o">.</span><span class="n">delete_sequence</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">start</span><span class="p">),</span> <span class="n">disrupt</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">out</span></div>

<div class="viewcode-block" id="Plasmid.splice"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.splice">[docs]</a>    <span class="k">def</span> <span class="nf">splice</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Get a slice sequence from start of all features to end of all features</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="p">[</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">strand</span><span class="p">]</span> <span class="o">=</span> <span class="n">unwrap_FeatureLocation</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>
        <span class="c1"># find starting index</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">)):</span>
            <span class="n">gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="p">[</span><span class="n">s1</span><span class="p">,</span> <span class="n">s2</span><span class="p">,</span> <span class="n">strand</span><span class="p">]</span> <span class="o">=</span> <span class="n">unwrap_FeatureLocation</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>
            <span class="n">c1</span> <span class="o">=</span> <span class="n">s2</span> <span class="o">&gt;</span> <span class="n">s1</span>
            <span class="n">c2</span> <span class="o">=</span> <span class="n">end</span> <span class="o">&gt;</span> <span class="n">start</span>
            <span class="n">c3</span> <span class="o">=</span> <span class="n">s2</span> <span class="o">&gt;</span> <span class="n">end</span>
            <span class="n">c4</span> <span class="o">=</span> <span class="n">s2</span> <span class="o">&gt;</span> <span class="n">start</span>
            <span class="c1"># exit if full Plasmid is covered</span>
            <span class="k">if</span> <span class="o">~</span><span class="n">c1</span> <span class="o">&amp;</span> <span class="o">~</span><span class="n">c2</span> <span class="o">&amp;</span> <span class="n">c4</span><span class="p">:</span>
                <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">inplace</span><span class="p">(</span><span class="n">inplace</span><span class="p">)</span>
            <span class="k">elif</span> <span class="p">(</span><span class="n">c1</span> <span class="o">&amp;</span> <span class="n">c2</span> <span class="o">&amp;</span> <span class="n">c3</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="o">~</span><span class="n">c1</span> <span class="o">&amp;</span> <span class="n">c2</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="o">~</span><span class="n">c1</span> <span class="o">&amp;</span> <span class="o">~</span><span class="n">c2</span> <span class="o">&amp;</span> <span class="n">c3</span><span class="p">):</span>
                <span class="n">end</span> <span class="o">=</span> <span class="n">s2</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">splice_sequence</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">),</span> <span class="n">inplace</span><span class="p">)</span></div>
    
    <span class="k">def</span> <span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">key</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Splice Plasmid sequence with range or filter through features on the Plasmid</span>
<span class="sd">        key = if key is a range or slice, then slice using base pair index from start to end</span>
<span class="sd">              if key is a str, then return the column values of pandas dataframe</span>
<span class="sd">              if key is an int, then return feature at the feature index</span>
<span class="sd">              if key is a list of integers, then return features only at the indexes in the list </span>
<span class="sd">              if key is a boolean array, then return features only at indexes which evaluate True </span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="c1"># splice based on slice or range</span>
        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">key</span><span class="p">)</span><span class="o">==</span><span class="nb">range</span><span class="p">:</span>
            <span class="n">out</span><span class="o">.</span><span class="n">splice_sequence</span><span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">elif</span> <span class="nb">type</span><span class="p">(</span><span class="n">key</span><span class="p">)</span><span class="o">==</span><span class="nb">slice</span><span class="p">:</span>
            <span class="n">start</span> <span class="o">=</span> <span class="n">key</span><span class="o">.</span><span class="n">start</span>
            <span class="n">stop</span> <span class="o">=</span> <span class="n">key</span><span class="o">.</span><span class="n">stop</span>
            <span class="k">if</span> <span class="n">start</span><span class="o">==</span><span class="kc">None</span><span class="p">:</span>
                <span class="n">start</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">if</span> <span class="n">stop</span><span class="o">==</span><span class="kc">None</span><span class="p">:</span>
                <span class="n">stop</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
            <span class="n">out</span><span class="o">.</span><span class="n">splice_sequence</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">),</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="c1"># splice based on feature index</span>
        <span class="k">elif</span> <span class="nb">type</span><span class="p">(</span><span class="n">key</span><span class="p">)</span><span class="o">==</span><span class="nb">int</span><span class="p">:</span>
            <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="p">[</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">key</span><span class="p">]]</span>
        <span class="c1"># returns column entries from pandas dataframe</span>
        <span class="k">elif</span> <span class="nb">type</span><span class="p">(</span><span class="n">key</span><span class="p">)</span> <span class="o">==</span> <span class="nb">str</span> <span class="ow">or</span> <span class="n">isStringArray</span><span class="p">(</span><span class="n">key</span><span class="p">):</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">to_dataframe</span><span class="p">()[</span><span class="n">key</span><span class="p">]</span>
        <span class="k">elif</span> <span class="n">isIntArray</span><span class="p">(</span><span class="n">key</span><span class="p">):</span>
            <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="p">[</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">key</span><span class="p">]</span>
        <span class="c1"># slice based on boolean array</span>
        <span class="k">elif</span> <span class="n">isBool</span><span class="p">(</span><span class="n">key</span><span class="p">):</span>
            <span class="n">keep</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">key</span><span class="p">))[</span><span class="n">key</span><span class="p">]</span>
            <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="p">[</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">keep</span><span class="p">]</span>
        <span class="c1"># work on output from pandas dataframe</span>
        <span class="k">elif</span> <span class="nb">type</span><span class="p">(</span><span class="n">key</span><span class="p">)</span> <span class="o">==</span> <span class="n">pd</span><span class="o">.</span><span class="n">core</span><span class="o">.</span><span class="n">series</span><span class="o">.</span><span class="n">Series</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">out</span><span class="p">[</span><span class="n">key</span><span class="o">.</span><span class="n">values</span><span class="p">]</span>
        <span class="c1"># return key and column</span>
        <span class="k">elif</span> <span class="nb">type</span><span class="p">(</span><span class="n">key</span><span class="p">)</span><span class="o">==</span><span class="nb">tuple</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">key</span><span class="p">:</span>
                <span class="n">out</span> <span class="o">=</span> <span class="n">out</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>
            <span class="k">return</span> <span class="n">out</span>
        <span class="n">out</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> 
        <span class="k">return</span> <span class="n">out</span>

    <span class="k">def</span> <span class="fm">__setitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Sets a feature to something</span>
<span class="sd">        key = 2D index of [&lt;index&gt;, &lt;column&gt;] or [&lt;index&gt;, &lt;column&gt;] or 1D [&lt;index&gt;] of features or sequences to replace</span>
<span class="sd">        value = slice of a Plasmid dataframe, which contains one sequence and features associated with this position</span>
<span class="sd">                the sequence at the location of the feature will be replaced with the provided sequence</span>
<span class="sd">                the SeqFeature at the index will be replaced with the new feature</span>
<span class="sd">                all modifications are done in place</span>
<span class="sd">        &#39;&#39;&#39;</span>       
        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">key</span><span class="p">)</span><span class="o">==</span><span class="nb">int</span><span class="p">:</span> <span class="c1"># debug if features may get overriden</span>
            <span class="p">[</span><span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">,</span> <span class="n">strand</span><span class="p">]</span> <span class="o">=</span> <span class="n">unwrap_FeatureLocation</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">replace_sequence</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">),</span> <span class="n">value</span><span class="p">,</span> <span class="n">disrupt</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">elif</span> <span class="nb">type</span><span class="p">(</span><span class="n">key</span><span class="p">)</span><span class="o">==</span><span class="nb">range</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">replace_sequence</span><span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="n">value</span><span class="p">,</span> <span class="n">disrupt</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">elif</span> <span class="nb">type</span><span class="p">(</span><span class="n">key</span><span class="p">)</span><span class="o">==</span><span class="nb">slice</span><span class="p">:</span>
            <span class="n">start</span> <span class="o">=</span> <span class="n">key</span><span class="o">.</span><span class="n">start</span>
            <span class="n">stop</span> <span class="o">=</span> <span class="n">key</span><span class="o">.</span><span class="n">stop</span>
            <span class="k">if</span> <span class="n">start</span><span class="o">==</span><span class="kc">None</span><span class="p">:</span>
                <span class="n">start</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">if</span> <span class="n">stop</span><span class="o">==</span><span class="kc">None</span><span class="p">:</span>
                <span class="n">stop</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">replace_sequence</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">),</span> <span class="n">value</span><span class="p">,</span> <span class="n">disrupt</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">elif</span> <span class="nb">type</span><span class="p">(</span><span class="n">key</span><span class="p">)</span><span class="o">==</span><span class="nb">tuple</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">key</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span><span class="o">==</span><span class="nb">str</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">set_column</span><span class="p">(</span><span class="n">key</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">key</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">value</span><span class="p">)</span>
            <span class="k">elif</span> <span class="nb">type</span><span class="p">(</span><span class="n">key</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">==</span><span class="nb">str</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">set_column</span><span class="p">(</span><span class="n">key</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">key</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">value</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">error</span><span class="p">()</span>

    <span class="k">def</span> <span class="fm">__delitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">key</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Delete feature, slice of features, or range of sequences</span>
<span class="sd">        key = if key is a slice, then features from start,end,stride are delete</span>
<span class="sd">              if key is an int, then the feature at the index is deleted</span>
<span class="sd">              if key is a range, then sequences from bp index start to end is deleted</span>
<span class="sd">              if key is a boolean array, then features at index with True are deleted</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">key</span><span class="p">)</span><span class="o">==</span><span class="nb">int</span><span class="p">:</span>
            <span class="k">del</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">key</span><span class="p">])</span>
        <span class="k">elif</span> <span class="nb">type</span><span class="p">(</span><span class="n">key</span><span class="p">)</span><span class="o">==</span><span class="nb">range</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">delete_sequence</span><span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">elif</span> <span class="nb">type</span><span class="p">(</span><span class="n">key</span><span class="p">)</span><span class="o">==</span><span class="nb">slice</span><span class="p">:</span>
            <span class="n">start</span> <span class="o">=</span> <span class="n">key</span><span class="o">.</span><span class="n">start</span>
            <span class="n">stop</span> <span class="o">=</span> <span class="n">key</span><span class="o">.</span><span class="n">stop</span>
            <span class="k">if</span> <span class="n">start</span><span class="o">==</span><span class="kc">None</span><span class="p">:</span>
                <span class="n">start</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">if</span> <span class="n">stop</span><span class="o">==</span><span class="kc">None</span><span class="p">:</span>
                <span class="n">stop</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">delete_sequence</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">),</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">isBool</span><span class="p">(</span><span class="n">key</span><span class="p">):</span>
            <span class="n">keep</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">key</span><span class="p">))[</span><span class="n">key</span><span class="o">==</span><span class="kc">False</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">keep</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">error</span><span class="p">()</span>

<div class="viewcode-block" id="Plasmid.translate"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.translate">[docs]</a>    <span class="k">def</span> <span class="nf">translate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">frame</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">table</span><span class="o">=</span><span class="s1">&#39;Standard&#39;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Translate the DNA sequence to amino acids</span>
<span class="sd">        frame = codon reading frame such as 0, 1, or 2</span>
<span class="sd">        table = codon table to use. Standard codon table is used by default</span>
<span class="sd">        return a string</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">strand</span> <span class="o">=</span> <span class="mi">1</span> 
        <span class="k">elif</span> <span class="n">strand</span><span class="o">==</span><span class="kc">None</span><span class="p">:</span>
            <span class="n">strand</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">strand</span>
        
        <span class="k">if</span> <span class="n">strand</span><span class="o">==</span><span class="mi">1</span><span class="p">:</span>
            <span class="n">AA_seq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="n">frame</span><span class="p">:]</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">table</span><span class="o">=</span><span class="n">table</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">strand</span><span class="o">==-</span><span class="mi">1</span><span class="p">:</span>
            <span class="n">AA_seq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()[</span><span class="n">frame</span><span class="p">:]</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">table</span><span class="o">=</span><span class="n">table</span><span class="p">)</span>
        <span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="n">AA_seq</span><span class="p">)</span></div>

<div class="viewcode-block" id="Plasmid.to_dataframe"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.to_dataframe">[docs]</a>    <span class="k">def</span> <span class="nf">to_dataframe</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">str_only</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Obtain a dataframe of genetic features on the Plasmid</span>
<span class="sd">        str_only = export objects formatted into strings </span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="n">data</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
            <span class="n">label</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">feature</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">type</span>
            <span class="c1"># figure out strand location</span>
            <span class="p">[</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">strand</span><span class="p">]</span> <span class="o">=</span> <span class="n">unwrap_FeatureLocation</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">strand</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">color</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s1">&#39;ApEinfo_fwdcolor&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">color</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s1">&#39;ApEinfo_revcolor&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
<span class="w">            </span><span class="sd">&#39;&#39;&#39; debug: will break auto annotations on Plasmid merge</span>
<span class="sd">            # handle sequences which wrap around the origin</span>
<span class="sd">            if start &gt; end:</span>
<span class="sd">                seq = str(self.SeqRecord.seq[start:]) + str(self.SeqRecord.seq[:end])</span>
<span class="sd">            else: </span>
<span class="sd">                seq = str(self.SeqRecord.seq[start:end])</span>
<span class="sd">            # reorient the sequence if needed</span>
<span class="sd">            if fwd_only and strand==-1:</span>
<span class="sd">                seq = reverse_complement(seq)</span>
<span class="sd">                strand = 1</span>
<span class="sd">            &#39;&#39;&#39;</span>
            <span class="n">seq</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
            <span class="n">length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
            <span class="n">loc</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">location</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">str_only</span><span class="p">:</span>
                <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">label</span><span class="p">,</span> <span class="n">feature</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">length</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">location</span><span class="p">),</span> <span class="n">color</span><span class="p">])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">label</span><span class="p">,</span> <span class="n">feature</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">length</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="n">color</span><span class="p">])</span>
        <span class="n">col</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">,</span><span class="s1">&#39;type&#39;</span><span class="p">,</span><span class="s1">&#39;start&#39;</span><span class="p">,</span><span class="s1">&#39;end&#39;</span><span class="p">,</span><span class="s1">&#39;length&#39;</span><span class="p">,</span><span class="s1">&#39;strand&#39;</span><span class="p">,</span><span class="s1">&#39;location&#39;</span><span class="p">,</span><span class="s1">&#39;color&#39;</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">col</span><span class="p">)</span></div>

    <span class="k">def</span> <span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Return string showing memory address and table contents</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="c1"># get memory address</span>
        <span class="n">loc</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="p">))</span><span class="o">+</span><span class="s1">&#39; at &#39;</span><span class="o">+</span><span class="nb">hex</span><span class="p">(</span><span class="nb">id</span><span class="p">(</span><span class="bp">self</span><span class="p">))</span>
        <span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="n">loc</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_table</span><span class="p">()</span>

    <span class="k">def</span> <span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Returns nucleotide sequence of Plasmid dataframe</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>

    <span class="k">def</span> <span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Return the number of genetic features</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">)</span>
    
<div class="viewcode-block" id="Plasmid.get_colored"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.get_colored">[docs]</a>    <span class="k">def</span> <span class="nf">get_colored</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Print out colored representation of Plasmid</span>
<span class="sd">        width = number of characters per line</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="n">rec</span> <span class="o">=</span> <span class="n">Graphic</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="bp">self</span><span class="p">)</span>
        <span class="n">out</span> <span class="o">=</span> <span class="n">rec</span><span class="o">.</span><span class="n">colorize_plasmid</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="n">width</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">out</span></div>
 
<div class="viewcode-block" id="Plasmid.get_table"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.get_table">[docs]</a>    <span class="k">def</span> <span class="nf">get_table</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="c1"># add genbank header info</span>
        <span class="n">table</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">table</span><span class="o">+=</span><span class="nb">str</span><span class="p">(</span><span class="n">k</span><span class="o">+</span><span class="s1">&#39;:&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="n">k</span><span class="p">]))</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span> 
        <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">to_dataframe</span><span class="p">()</span>
        <span class="n">table</span><span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="n">df</span><span class="p">[[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">,</span><span class="s1">&#39;type&#39;</span><span class="p">,</span><span class="s1">&#39;location&#39;</span><span class="p">,</span><span class="s1">&#39;length&#39;</span><span class="p">,</span><span class="s1">&#39;color&#39;</span><span class="p">]])</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span>
        <span class="c1"># colorize the colors</span>
        <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;color&#39;</span><span class="p">]):</span>
            <span class="n">val</span> <span class="o">=</span> <span class="n">Graphic</span><span class="p">()</span><span class="o">.</span><span class="n">get_colored_text</span><span class="p">(</span><span class="n">c</span><span class="p">,</span><span class="n">fg</span><span class="o">=</span><span class="n">c</span><span class="p">)</span>
            <span class="n">table</span> <span class="o">=</span> <span class="n">table</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">c</span><span class="p">,</span><span class="n">val</span><span class="p">)</span>
        <span class="n">table</span><span class="o">+=</span> <span class="s1">&#39;total length:&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="fm">__str__</span><span class="p">()))</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span>
        <span class="k">return</span> <span class="n">table</span></div>

<div class="viewcode-block" id="Plasmid.annotate"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.annotate">[docs]</a>    <span class="k">def</span> <span class="nf">annotate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">sequence</span><span class="p">,</span> <span class="n">feature</span><span class="o">=</span><span class="s1">&#39;unknown&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Adds annotations to a plasmid using sequence information</span>
<span class="sd">        name = name of the gene</span>
<span class="sd">        sequence = DNA or amino acid sequence of the feature</span>
<span class="sd">        feature = type of genetic feature such as cds, mRNA, primer_bind</span>
<span class="sd">        color = [fwd_color, rev_color] to use</span>
<span class="sd">        circular = search as though sequence is a circular construct</span>
<span class="sd">        inplace = performs modifications inplace</span>
<span class="sd">        returns a modified plasmid dataframe</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="c1"># remove gaps</span>
        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span> <span class="o">==</span> <span class="nb">str</span><span class="p">:</span>
            <span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">sequence</span><span class="o">==</span><span class="s1">&#39;&#39;</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span>

        <span class="c1"># apply default colors if it is not provided</span>
        <span class="n">color</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_default_annotation_color</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">feature</span><span class="p">,</span> <span class="n">color</span><span class="p">)</span>
        
        <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">inplace</span><span class="p">(</span><span class="n">inplace</span><span class="p">)</span>
        <span class="c1"># search for DNA sequence</span>
        <span class="k">if</span> <span class="n">isDNA</span><span class="p">(</span><span class="n">sequence</span><span class="p">):</span>
            <span class="c1"># convert U to T for DNA sequence only</span>
            <span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;U&#39;</span><span class="p">,</span><span class="s1">&#39;T&#39;</span><span class="p">)</span>
            <span class="n">pos</span> <span class="o">=</span> <span class="n">Aligner</span><span class="o">.</span><span class="n">search_DNA</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">),</span> <span class="n">circular</span><span class="o">=</span><span class="n">circular</span><span class="p">)</span>
        <span class="c1"># search for protein sequence</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">pos</span> <span class="o">=</span> <span class="n">Aligner</span><span class="o">.</span><span class="n">search_protein</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">),</span> <span class="n">circular</span><span class="o">=</span><span class="n">circular</span><span class="p">)</span>
        
        <span class="n">new_features</span><span class="o">=</span><span class="p">[]</span>
        <span class="c1"># add the annotation</span>
        <span class="k">for</span> <span class="p">[</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">strand</span><span class="p">]</span> <span class="ow">in</span> <span class="n">pos</span><span class="p">[[</span><span class="s1">&#39;start&#39;</span><span class="p">,</span><span class="s1">&#39;end&#39;</span><span class="p">,</span><span class="s1">&#39;strand&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
            <span class="c1"># create compound location which can wrap around the origin</span>
            <span class="n">loc</span> <span class="o">=</span> <span class="n">wrap_FeatureLocation</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>
            <span class="n">qualifiers</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">:[</span><span class="n">name</span><span class="p">],</span> <span class="s1">&#39;ApEinfo_fwdcolor&#39;</span><span class="p">:[</span><span class="n">color</span><span class="p">[</span><span class="mi">0</span><span class="p">]],</span> <span class="s1">&#39;ApEinfo_revcolor&#39;</span><span class="p">:[</span><span class="n">color</span><span class="p">[</span><span class="mi">1</span><span class="p">]]}</span>
            <span class="n">gene</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqFeature</span><span class="o">.</span><span class="n">SeqFeature</span><span class="p">(</span><span class="nb">type</span><span class="o">=</span><span class="n">feature</span><span class="p">,</span> <span class="n">location</span><span class="o">=</span><span class="n">loc</span><span class="p">,</span> <span class="n">qualifiers</span><span class="o">=</span><span class="n">qualifiers</span><span class="p">)</span>
            <span class="n">new_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

        <span class="c1"># append to Plasmid and update</span>
        <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="o">+=</span> <span class="n">new_features</span>
        <span class="n">out</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">out</span></div>

<div class="viewcode-block" id="Plasmid.get_default_annotation_color"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.get_default_annotation_color">[docs]</a>    <span class="k">def</span> <span class="nf">get_default_annotation_color</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">feature</span><span class="p">,</span> <span class="n">color</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Look up default annotation colors for a given name and feature</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="c1"># apply default colors if it is not provided</span>
        <span class="k">if</span> <span class="n">color</span><span class="o">==</span><span class="kc">None</span><span class="p">:</span>
            <span class="n">color</span> <span class="o">=</span> <span class="n">search_key</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;colors&#39;</span><span class="p">][</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">],</span> <span class="n">name</span><span class="p">,</span> <span class="n">rev</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">color</span><span class="o">==</span><span class="kc">None</span><span class="p">:</span>
            <span class="n">color</span> <span class="o">=</span> <span class="n">search_key</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;colors&#39;</span><span class="p">][</span><span class="s1">&#39;features&#39;</span><span class="p">],</span> <span class="n">feature</span><span class="p">,</span> <span class="n">rev</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="c1"># apply default colors</span>
        <span class="k">if</span> <span class="n">color</span><span class="o">==</span><span class="kc">None</span><span class="p">:</span>
            <span class="n">color</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;colors&#39;</span><span class="p">][</span><span class="s1">&#39;features&#39;</span><span class="p">][</span><span class="s1">&#39;unknown&#39;</span><span class="p">]</span>
        
        <span class="c1"># format to list</span>
        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">color</span><span class="p">)</span> <span class="o">==</span> <span class="nb">str</span><span class="p">:</span>
            <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="n">color</span><span class="p">,</span> <span class="n">color</span><span class="p">]</span>
        <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="n">color</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">color</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]]</span>
        <span class="k">return</span> <span class="n">color</span></div>
    
<div class="viewcode-block" id="Plasmid.annotate_by_location"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.annotate_by_location">[docs]</a>    <span class="k">def</span> <span class="nf">annotate_by_location</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">feature</span><span class="o">=</span><span class="s1">&#39;unknown&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Adds annotations to a plasmid using location information</span>
<span class="sd">        name = name of the gene</span>
<span class="sd">        start = start location of feature</span>
<span class="sd">        stop = stop location of feature</span>
<span class="sd">        strand = orientation of feature</span>
<span class="sd">        feature = type of genetic feature such as cds, mRNA, primer_bind</span>
<span class="sd">        color = [fwd_color, rev_color] to use</span>
<span class="sd">        inplace = performs modifications inplace</span>
<span class="sd">        returns a modified plasmid dataframe</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">inplace</span><span class="p">(</span><span class="n">inplace</span><span class="p">)</span>
        
        <span class="n">color</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_default_annotation_color</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">feature</span><span class="p">,</span> <span class="n">color</span><span class="p">)</span>
        <span class="c1"># create compound location which can wrap around the origin</span>
        <span class="n">loc</span> <span class="o">=</span> <span class="n">wrap_FeatureLocation</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>
        <span class="n">qualifiers</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">:[</span><span class="n">name</span><span class="p">],</span> <span class="s1">&#39;ApEinfo_fwdcolor&#39;</span><span class="p">:[</span><span class="n">color</span><span class="p">[</span><span class="mi">0</span><span class="p">]],</span> <span class="s1">&#39;ApEinfo_revcolor&#39;</span><span class="p">:[</span><span class="n">color</span><span class="p">[</span><span class="mi">1</span><span class="p">]]}</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqFeature</span><span class="o">.</span><span class="n">SeqFeature</span><span class="p">(</span><span class="nb">type</span><span class="o">=</span><span class="n">feature</span><span class="p">,</span> <span class="n">location</span><span class="o">=</span><span class="n">loc</span><span class="p">,</span> <span class="n">qualifiers</span><span class="o">=</span><span class="n">qualifiers</span><span class="p">)</span>
        <span class="c1"># append to Plasmid and update</span>
        <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="o">+=</span> <span class="p">[</span><span class="n">gene</span><span class="p">]</span>
        <span class="n">out</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">out</span></div>
    
<div class="viewcode-block" id="Plasmid.annotate_by_reference"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.annotate_by_reference">[docs]</a>    <span class="k">def</span> <span class="nf">annotate_by_reference</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Annotate using a reference library of sequences</span>
<span class="sd">        ref = dataframe with columns</span>
<span class="sd">            [name, sequence, feature_type, color]</span>
<span class="sd">        return annotated plasmid</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">inplace</span><span class="p">(</span><span class="n">inplace</span><span class="p">)</span>
        <span class="c1"># check if color in the data column</span>
        <span class="k">if</span> <span class="p">(</span><span class="s1">&#39;color&#39;</span> <span class="ow">in</span> <span class="n">ref</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span><span class="o">==</span><span class="kc">False</span><span class="p">:</span>
            <span class="n">ref</span><span class="p">[</span><span class="s1">&#39;color&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
            
        <span class="n">cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">,</span><span class="s1">&#39;sequence&#39;</span><span class="p">,</span><span class="s1">&#39;feature_type&#39;</span><span class="p">,</span><span class="s1">&#39;color&#39;</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">label</span><span class="p">,</span> <span class="n">seq</span><span class="p">,</span> <span class="n">ftype</span><span class="p">,</span> <span class="n">color</span> <span class="ow">in</span> <span class="n">ref</span><span class="p">[</span><span class="n">cols</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
            <span class="n">out</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">label</span><span class="p">,</span> <span class="n">sequence</span><span class="o">=</span><span class="n">seq</span><span class="p">,</span> <span class="n">feature</span><span class="o">=</span><span class="n">ftype</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">out</span></div>

<div class="viewcode-block" id="Plasmid.merge"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.merge">[docs]</a>    <span class="k">def</span> <span class="nf">merge</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Merge Plasmid dataframes together</span>
<span class="sd">        value = Plasmid or dataframe of features to merge into self</span>
<span class="sd">        how = if &#39;left&#39; then features from value are appended to self</span>
<span class="sd">              if &#39;right&#39; then features from self merged into value</span>
<span class="sd">              if &#39;and&#39; then only intersection of features are returned</span>
<span class="sd">              if &#39;nand&#39; then only non-intersecting features of self are returned</span>
<span class="sd">        returns Plasmid dataframe</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">inplace</span><span class="p">(</span><span class="n">inplace</span><span class="p">)</span>   
        <span class="k">if</span> <span class="n">how</span><span class="o">==</span><span class="s1">&#39;left&#39;</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">value</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span><span class="o">==</span><span class="s1">&#39;Plasmid&#39;</span><span class="p">:</span>
                <span class="n">value</span> <span class="o">=</span> <span class="n">value</span><span class="o">.</span><span class="n">to_dataframe</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">value</span><span class="p">[[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">,</span><span class="s1">&#39;type&#39;</span><span class="p">,</span><span class="s1">&#39;sequence&#39;</span><span class="p">,</span><span class="s1">&#39;color&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
                <span class="n">out</span> <span class="o">=</span> <span class="n">out</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">v</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">feature</span><span class="o">=</span><span class="n">v</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">sequence</span><span class="o">=</span><span class="n">v</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">v</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
            <span class="k">return</span> <span class="n">out</span>
        <span class="k">elif</span> <span class="n">how</span><span class="o">==</span><span class="s1">&#39;right&#39;</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">value</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="n">inplace</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">how</span><span class="o">==</span><span class="s1">&#39;and&#39;</span><span class="p">:</span> <span class="c1"># debug todo refinement</span>
            <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">value</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span><span class="o">==</span><span class="s1">&#39;Plasmid&#39;</span><span class="p">:</span>
                <span class="n">value</span> <span class="o">=</span> <span class="n">value</span><span class="o">.</span><span class="n">to_dataframe</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">value</span><span class="p">[[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">,</span><span class="s1">&#39;type&#39;</span><span class="p">,</span><span class="s1">&#39;sequence&#39;</span><span class="p">,</span><span class="s1">&#39;color&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
                <span class="n">out</span> <span class="o">=</span> <span class="n">out</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">v</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">feature</span><span class="o">=</span><span class="n">v</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">sequence</span><span class="o">=</span><span class="n">v</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">v</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
            <span class="k">return</span> <span class="n">out</span> 
        <span class="k">elif</span> <span class="n">how</span><span class="o">==</span><span class="s1">&#39;nand&#39;</span><span class="p">:</span> <span class="c1"># needs refinement debug</span>
            <span class="n">keep</span> <span class="o">=</span> <span class="p">[</span><span class="kc">True</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">)</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">out</span><span class="o">.</span><span class="n">to_dataframe</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">q</span> <span class="ow">in</span> <span class="n">value</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">)):</span>
                    <span class="k">if</span> <span class="n">keep</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="ow">and</span> <span class="n">q</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">==</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">lower</span><span class="p">():</span>
                        <span class="n">keep</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">keep</span><span class="p">))</span>
            <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="p">[</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">x</span><span class="p">[</span><span class="n">keep</span><span class="p">]]</span>
            <span class="k">return</span> <span class="n">out</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">error</span><span class="p">()</span></div>

<div class="viewcode-block" id="Plasmid.drop_duplicates"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.drop_duplicates">[docs]</a>    <span class="k">def</span> <span class="nf">drop_duplicates</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span> 
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Shift through feature list and remove duplicated annotations</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">inplace</span><span class="p">(</span><span class="n">inplace</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">out</span><span class="p">[</span><span class="n">out</span><span class="o">.</span><span class="n">duplicates</span><span class="p">()</span><span class="o">==</span><span class="kc">False</span><span class="p">]</span></div>

<div class="viewcode-block" id="Plasmid.duplicates"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.duplicates">[docs]</a>    <span class="k">def</span> <span class="nf">duplicates</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span> 
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Sort through feature list and return True/False index of entries which maybe duplicates</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">update</span><span class="p">()</span><span class="o">.</span><span class="n">to_dataframe</span><span class="p">()</span>
        <span class="n">dupe</span> <span class="o">=</span> <span class="p">[</span><span class="kc">False</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">dupe</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">==</span><span class="kc">False</span><span class="p">:</span>
                <span class="n">c1</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">][[</span><span class="s1">&#39;start&#39;</span><span class="p">,</span><span class="s1">&#39;end&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span>
                <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">)):</span>
                    <span class="n">c2</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">j</span><span class="p">][[</span><span class="s1">&#39;start&#39;</span><span class="p">,</span><span class="s1">&#39;end&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span>
                    <span class="k">if</span> <span class="n">c1</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">==</span><span class="n">c2</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">and</span> <span class="n">c1</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">==</span><span class="n">c2</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
                        <span class="n">dupe</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="k">if</span> <span class="n">c2</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">c1</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
                        <span class="k">break</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">dupe</span><span class="p">)</span></div>
 
    <span class="k">def</span> <span class="fm">__radd__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Performs Plasmid operation other + self</span>
<span class="sd">        Other can be a str</span>
<span class="sd">        Self is a Plasmid dataframe</span>
<span class="sd">        returns a modified Plasmid</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="c1"># make Plasmids linear by splitting at the origin</span>
        <span class="n">seq1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">linearize</span><span class="p">()</span>
        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">other</span><span class="p">)</span> <span class="o">==</span> <span class="nb">str</span><span class="p">:</span>
            <span class="n">seq2</span> <span class="o">=</span> <span class="n">Plasmid</span><span class="p">(</span><span class="n">other</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">seq2</span> <span class="o">=</span> <span class="n">other</span><span class="o">.</span><span class="n">linearize</span><span class="p">()</span>
        
        <span class="c1"># concatenate seq records</span>
        <span class="n">sr</span> <span class="o">=</span> <span class="n">seq2</span><span class="o">.</span><span class="n">SeqRecord</span> <span class="o">+</span> <span class="n">seq1</span><span class="o">.</span><span class="n">SeqRecord</span>
        <span class="n">seq1</span><span class="o">.</span><span class="n">SeqRecord</span> <span class="o">=</span> <span class="n">seq2</span><span class="o">.</span><span class="n">SeqRecord</span> <span class="o">+</span> <span class="n">seq1</span><span class="o">.</span><span class="n">SeqRecord</span>
        <span class="n">seq1</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">seq1</span>
 
    <span class="k">def</span> <span class="fm">__add__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Performs Plasmid operation self + other</span>
<span class="sd">        Other can be a str</span>
<span class="sd">        Self is a Plasmid dataframe</span>
<span class="sd">        return a modified Plasmid</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="c1"># make Plasmids linear by splitting at the origin</span>
        <span class="n">seq1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">linearize</span><span class="p">()</span>
        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">other</span><span class="p">)</span> <span class="o">==</span> <span class="nb">str</span><span class="p">:</span>
            <span class="n">seq2</span> <span class="o">=</span> <span class="n">Plasmid</span><span class="p">(</span><span class="n">other</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">seq2</span> <span class="o">=</span> <span class="n">other</span><span class="o">.</span><span class="n">linearize</span><span class="p">()</span>
        
        <span class="c1"># concatenate seq records</span>
        <span class="n">seq1</span><span class="o">.</span><span class="n">SeqRecord</span> <span class="o">=</span> <span class="n">seq1</span><span class="o">.</span><span class="n">SeqRecord</span> <span class="o">+</span> <span class="n">seq2</span><span class="o">.</span><span class="n">SeqRecord</span>
        <span class="n">seq1</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">seq1</span>
    
<div class="viewcode-block" id="Plasmid.linearize"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.linearize">[docs]</a>    <span class="k">def</span> <span class="nf">linearize</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Edit SeqRecord so Plasmid is now linear</span>
<span class="sd">        returns a linearized Plasmid</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="k">if</span> <span class="n">inplace</span><span class="p">:</span>
            <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">N</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
        <span class="n">new</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">keep</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
            <span class="p">[</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">strand</span><span class="p">]</span> <span class="o">=</span> <span class="n">unwrap_FeatureLocation</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>
            <span class="c1"># check if gene overlaps origin</span>
            <span class="k">if</span> <span class="n">start</span> <span class="o">&gt;</span> <span class="n">end</span><span class="p">:</span>
                <span class="c1"># partition the component locations</span>
                <span class="n">loc1</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="n">loc2</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">for</span> <span class="n">L</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="vm">__dict__</span><span class="p">[</span><span class="s1">&#39;parts&#39;</span><span class="p">]:</span>
                    <span class="k">if</span> <span class="n">L</span><span class="o">.</span><span class="n">end</span> <span class="o">&lt;=</span> <span class="n">end</span><span class="p">:</span>
                        <span class="n">loc1</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">L</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">loc2</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">L</span><span class="p">)</span>
                <span class="n">gene1</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                <span class="n">gene2</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                <span class="n">gene1</span><span class="o">.</span><span class="n">location</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqFeature</span><span class="o">.</span><span class="n">CompoundLocation</span><span class="p">(</span><span class="n">loc1</span><span class="p">)</span>
                <span class="n">gene2</span><span class="o">.</span><span class="n">location</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqFeature</span><span class="o">.</span><span class="n">CompoundLocation</span><span class="p">(</span><span class="n">loc2</span><span class="p">)</span>
                <span class="c1"># add some information to label tag to show it was split</span>
                <span class="n">gene1</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">+=</span><span class="s1">&#39; subsequence&#39;</span>
                <span class="n">gene2</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">+=</span><span class="s1">&#39; subsequence&#39;</span>
                <span class="n">new</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene1</span><span class="p">)</span>
                <span class="n">new</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene2</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">keep</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
        <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="n">new</span> <span class="o">+</span> <span class="n">keep</span>
        <span class="k">return</span> <span class="n">out</span></div>

<div class="viewcode-block" id="Plasmid.shift_features"><a class="viewcode-back" href="../../plasmid.html#plasmid.plasmid.Plasmid.shift_features">[docs]</a>    <span class="k">def</span> <span class="nf">shift_features</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">N</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Shifts base index of sequences by N base pairs</span>
<span class="sd">        record = SeqRecord to modify</span>
<span class="sd">        N = number of base pairs to shift</span>
<span class="sd">        circular = If SeqRecord is a Plasmid, which is circular, shift the index circularly around origin</span>
<span class="sd">        inplace = If false, a copy of modified self is returned and original data is not modified</span>
<span class="sd">        returns a Plasmid dataframe</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">inplace</span><span class="p">(</span><span class="n">inplace</span><span class="p">)</span>
        <span class="n">seq_length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">out</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">location</span> <span class="o">=</span> <span class="n">shift_CompoundLocation</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="n">N</span><span class="p">,</span> <span class="n">seq_length</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="n">circular</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">out</span></div></div>



</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Zhewei Chen.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>