{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e0c986a8-9997-42ed-9b87-aae8b0243866",
   "metadata": {},
   "source": [
    "# Demo for python gene editor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6615ef09-9e39-4b29-9886-fe81c402b715",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gene_editor as pge"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebfb4d40-230d-465f-986e-6884c9597492",
   "metadata": {},
   "source": [
    "## Manipulating and displaying genbank files\n",
    "### Reading genbank or fasta files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c6e0c6a-05c1-4fcf-bd48-5076f2e69de7",
   "metadata": {},
   "source": [
    "Genbank, fasta, and text files can be read using the `plasmid` dataframe class. This function is analogous to `read_csv` in pandas. This data structure uses Biopython in the backend to store genbank information such as sequence and genomic features or tags. The underlying data in Biopython is redisplayed into a more user friendly for splicing, sorting, and concatenating like a pandas dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "282f9174-ce6a-48d2-a258-a0899edeeeaf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reading xRFP.gb as genbank file\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<class 'gene_editor.plasmid'> at 0x7f35cc64a3b0\n",
       "molecule_type:DNA\n",
       "topology:circular\n",
       "data_file_division:   \n",
       "date:05-DEC-2022\n",
       "accessions:['<unknown', 'id>']\n",
       "keywords:['']\n",
       "source:\n",
       "organism:. .\n",
       "taxonomy:[]\n",
       "comment:\n",
       "ApEinfo:methylated:1\n",
       "                              locus_tag          type        location  length  \\\n",
       "0                         AmpR promoter      promoter       [0:29](+)      29   \n",
       "1                              AmpR RBS           RBS      [29:70](+)      41   \n",
       "2                          SpectomycinR           CDS     [70:868](+)     798   \n",
       "3                             BBa_B0053    terminator    [878:948](+)      70   \n",
       "4   J23108, 0.51, Constitutive Promoter      promoter  [1032:1067](+)      35   \n",
       "5          RBS 1.00 strength, BBa_B0034           RBS  [1086:1098](+)      12   \n",
       "6                       LacI, BBa_C0012           CDS  [1104:2196](+)    1092   \n",
       "7                             Esp3I_fix  misc_feature  [2072:2078](+)       6   \n",
       "8         Forward Terminator, BBa_B1002    terminator  [2196:2230](+)      34   \n",
       "9   pLacO2, single operon pLac ZC082818      promoter  [2230:2268](+)      38   \n",
       "10                                LacO2  protein_bind  [2238:2255](+)      17   \n",
       "11                 RFP targeting region  misc_binding  [2268:2288](+)      20   \n",
       "12         spec/1_xRFPg1_0_xRFPg1_0_.gb       unknown  [2268:2371](+)     103   \n",
       "13      dCas9 handle conserved region A  protein_bind  [2288:2300](+)      12   \n",
       "14     dCas9 handle conserved region A'  protein_bind  [2304:2318](+)      14   \n",
       "15      dCas9 handle conserved region B  protein_bind  [2318:2330](+)      12   \n",
       "16                   S. pyog terminator    terminator  [2330:2371](+)      41   \n",
       "17        Forward Terminator, BBa_B1010    terminator  [2371:2411](+)      40   \n",
       "18                               pSC101    rep_origin  [2411:4448](+)    2037   \n",
       "19                        T0 terminator    terminator  [2435:2541](+)     106   \n",
       "20                      repA101 protein           CDS  [3296:4247](+)     951   \n",
       "\n",
       "      color  \n",
       "0   \u001b[38;2;15;127;254m#0f7ffe\u001b[39m  \n",
       "1      \u001b[38;2;0;255;255mcyan\u001b[39m  \n",
       "2   \u001b[38;2;212;0;55m#d40037\u001b[39m  \n",
       "3   \u001b[38;2;157;27;28m#9d1b1c\u001b[39m  \n",
       "4   \u001b[38;2;64;128;0m#408000\u001b[39m  \n",
       "5      \u001b[38;2;0;255;255mcyan\u001b[39m  \n",
       "6   \u001b[38;2;252;102;101m#fc6665\u001b[39m  \n",
       "7      \u001b[38;2;0;255;255mcyan\u001b[39m  \n",
       "8   \u001b[38;2;128;64;0m#804000\u001b[39m  \n",
       "9   \u001b[38;2;0;128;128m#008080\u001b[39m  \n",
       "10  \u001b[38;2;7;63;128m#073f80\u001b[39m  \n",
       "11  \u001b[38;2;255;0;0m#ff0000\u001b[39m  \n",
       "12     \u001b[38;2;0;255;255mcyan\u001b[39m  \n",
       "13  \u001b[38;2;0;128;255m#0080ff\u001b[39m  \n",
       "14  \u001b[38;2;0;128;255m#0080ff\u001b[39m  \n",
       "15  \u001b[38;2;0;64;128m#004080\u001b[39m  \n",
       "16  \u001b[38;2;128;64;0m#804000\u001b[39m  \n",
       "17  \u001b[38;2;128;64;0m#804000\u001b[39m  \n",
       "18  \u001b[38;2;180;171;172m#b4abac\u001b[39m  \n",
       "19  \u001b[38;2;128;64;0m#804000\u001b[39m  \n",
       "20  \u001b[38;2;212;0;55m#d40037\u001b[39m  \n",
       "total length:4465"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pge.plasmid.read('xRFP.gb')\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dda430d1-7388-43db-976b-cea77afd4e77",
   "metadata": {},
   "source": [
    "The raw sequence with color annotations can be displayed with `.print()`. The dataframe can also be directly cast as a string to obtain the raw sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "344fded9-4aea-4b40-9601-ff16a842bd5b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[38;2;15;127;254mTTCAAATATCTATCCGCTCATGAGACAAT\u001b[39m\u001b[38;2;0;255;255mAACCCTGATAAATGCTTCAATAATATTGAAAAAGGAAGAAT\u001b[39m\u001b[38;2;212;0;55mATGAGTGAAAAAGTGCCCGCCGAGATTTCGGTGCAACTATCACAAGCACTCAACGTCATCGGGCGCCACTTGGAGTCGACGTTGCTGGCCGTGCATTTGTACGGCTCCGCACTGGATGGCGGATTGAAACCGTACAGTGATATTGATTTGCTGGTGACTGTAGCTGCACCGCTCAATGATGCCGTGCGGCAAGCCCTGCTCGTCGATCTCTTGGAGGTTTCAGCTTCCCCTGGCCAAAACAAGGCACTCCGCGCCTTGGAAGTGACCATCGTCGTGCACAGTGACATCGTACCTTGGCGTTATCCGGCCAGGCGGGAACTGCAGTTCGGAGAGTGGCAGCGCAAAGACATCCTTGCGGGCATCTTCGAGCCCGCCACAACCGATTCTGACTTGGCGATTCTGCTAACAAAGGCAAAGCAACATAGCGTCGTCTTGGCAGGTTCAGCAGCGAAGGATCTCTTCAGCTCAGTCCCAGAAAGCGATCTATTCAAGGCACTGGCCGATACTCTGAAGCTATGGAACTCGCCGCCAGATTGGGCGGGCGATGAGCGGAATGTAGTGCTTACTTTGTCTCGTATCTGGTACACCGCAGCAACCGGCAAGATCGCGCCAAAGGATGTTGCTGCCACTTGGGCAATGGCACGCTTGCCAGCTCAACATCAGCCCATCCTGTTGAATGCCAAGCGGGCTTATCTTGGGCAAGAAGAAGATTATTTGCCCGCTCGTGCGGATCAGGTGGCGGCGCTCATTAAATTCGTGAAGTATGAAGCAGTTAAACTGCTTGGTGCCAGCCAATGA\u001b[39mTAATACTAGC\u001b[38;2;157;27;28mTCCGGCAAAAAAACGGGCAAGGTGTCACCACCCTGCCCTTTTTCTTTAAAACCGAAAAGATTACTTCGCG\u001b[39mAAATAGGCGTATCACGAGGCAGAATTTCAGATAAAAAAAATCCTTAGCTTTCGCTAAGGATGATTTCTGGAATTCGCGGCCGCT\u001b[38;2;64;128;0mCTGACAGCTAGCTCAGTCCTAGGTATAATGCTAGC\u001b[39mAAAAGAATTCAAAAGATCT\u001b[38;2;0;255;255mAAAGAGGAGAAA\u001b[39mGGATCT\u001b[38;2;252;102;101mATGGTGAATGTGAAACCAGTAACGTTATACGATGTCGCAGAGTATGCCGGTGTCTCTTATCAGACCGTTTCCCGCGTGGTGAACCAGGCCAGCCACGTTTCTGCGAAAACGCGGGAAAAAGTGGAAGCGGCGATGGCGGAGCTGAATTACATTCCCAACCGCGTGGCACAACAACTGGCGGGCAAACAGTCGTTGCTGATTGGCGTTGCCACCTCCAGTCTGGCCCTGCACGCGCCGTCGCAAATTGTCGCGGCGATTAAATCTCGCGCCGATCAACTGGGTGCCAGCGTGGTGGTGTCGATGGTAGAACGAAGCGGCGTCGAAGCCTGTAAAGCGGCGGTGCACAATCTTCTCGCGCAACGCGTCAGTGGGCTGATCATTAACTATCCGCTGGATGACCAGGATGCCATTGCTGTGGAAGCTGCCTGCACTAATGTTCCGGCGTTATTTCTTGATGTCTCTGACCAGACACCCATCAACAGTATTATTTTCTCCCATGAAGACGGTACGCGACTGGGCGTGGAGCATCTGGTCGCATTGGGTCACCAGCAAATCGCGCTGTTAGCGGGCCCATTAAGTTCTGTCTCGGCGCGTCTGCGTCTGGCTGGCTGGCATAAATATCTCACTCGCAATCAAATTCAGCCGATAGCGGAACGGGAAGGCGACTGGAGTGCCATGTCCGGTTTTCAACAAACCATGCAAATGCTGAATGAGGGCATCGTTCCCACTGCGATGCTGGTTGCCAACGATCAGATGGCGCTGGGCGCAATGCGCGCCATTACCGAGTCCGGGCTGCGCGTTGGTGCGGATATCTCGGTAGTGGGATACGACGATACCGAAGACAGCTCATGTTATATCCCGCCGTTAACCACCATCAAACAGGATTTTCGCCTGCTGGGGCAAACCAGCGTGGACCGCTTGCTGCAACTCTCTCAGGGCCAGGCGGTGAAGGGCAATCAGCTGTTGCC\u001b[39m\u001b[38;2;0;255;255mCGTTTC\u001b[39m\u001b[38;2;252;102;101mACTGGTGAAAAGAAAAACCACCCTGGCGCCCAATACGCGAACCGCCTCTCCCCGCGCGTTGGCCGATTCATTAATGCAGCTGGCACGACAGGTTTCCCGACTGGAAAGCGGGCAGTAA\u001b[39m\u001b[38;2;128;64;0mCGCAAAAAACCCCGCTTCGGCGGGGTTTTTTCGC\u001b[39m\u001b[38;2;0;128;128mAATTGACA\u001b[39m\u001b[38;2;7;63;128mATGTGAGCGAGTAACAA\u001b[39m\u001b[38;2;0;128;128mGATACTGAGCACA\u001b[39m\u001b[38;2;255;0;0mAACTTTCAGTTTAGCGGTCT\u001b[39m\u001b[38;2;0;128;255mGTTTTAGAGCTA\u001b[39m\u001b[38;2;0;255;255mGAAA\u001b[39m\u001b[38;2;0;128;255mTAGCAAGTTAAAAT\u001b[39m\u001b[38;2;0;64;128mAAGGCTAGTCCG\u001b[39m\u001b[38;2;128;64;0mTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTTCGCCGCAAACCCCGCCCCTGACAGGGCGGGGTTTCGCCGC\u001b[39m\u001b[38;2;180;171;172mCTGTCAGACCAAGTTTACGAGCTC\u001b[39m\u001b[38;2;128;64;0mGCTTGGACTCCTGTTGATAGATCCAGTAATGACCTCAGAACTCCATCTGGATTTGTTCAGAACGCTCGGTTGCCGCCGGGCGTTTTTTATTGGTGAGAATCCAAGC\u001b[39m\u001b[38;2;180;171;172mACTAGGGACAGTAAGACGGGTAAGCCTGTTGATGATACCGCTGCCTTACTGGGTGCATTAGCCAGTCTGAATGACCTGTCACGGGATAATCCGAAGTGGTCAGACTGGAAAATCAGAGGGCAGGAACTGCTGAACAGCAAAAAGTCAGATAGCACCACATAGCAGACCCGCCATAAAACGCCCTGAGAAGCCCGTGACGGGCTTTTCTTGTATTATGGGTAGTTTCCTTGCATGAATCCATAAAAGGCGCCTGTAGTGCCATTTACCCCCATTCACTGCCAGAGCCGTGAGCGCAGCGAACTGAATGTCACGAAAAAGACAGCGACTCAGGTGCCTGATGGTCGGAGACAAAAGGAATATTCAGCGATTTGCCCGAGCTTGCGAGGGTGCTACTTAAGCCTTTAGGGTTTTAAGGTCTGTTTTGTAGAGGAGCAAACAGCGTTTGCGACATCCTTTTGTAATACTGCGGAACTGACTAAAGTAGTGAGTTATACACAGGGCTGGGATCTATTCTTTTTATCTTTTTTTATTCTTTCTTTATTCTATAAATTATAACCACTTGAATATAAACAAAAAAAACACACAAAGGTCTAGCGGAATTTACAGAGGGTCTAGCAGAATTTACAAGTTTTCCAGCAAAGGTCTAGCAGAATTTACAGATACCCACAACTCAAAGGAAAAGGACATGTAATTATCATTGACTAGCCCATCTCAATTGGTATAGTGATTAAAATCACCTAGACCAATTGAGATGT\u001b[39m\u001b[38;2;212;0;55mATGTCTGAATTAGTTGTTTTCAAAGCAAATGAACTAGCGATTAGTCGCTATGACTTAACGGAGCATGAAACCAAGCTAATTTTATGCTGTGTGGCACTACTCAACCCCACGATTGAAAACCCTACAAGGAAAGAACGGACGGTATCGTTCACTTATAACCAATACGCTCAGATGATGAACATCAGTAGGGAAAATGCTTATGGTGTATTAGCTAAAGCAACCAGAGAGCTGATGACGAGAACTGTGGAAATCAGGAATCCTTTGGTTAAAGGCTTTGAGATTTTCCAGTGGACAAACTATGCCAAGTTCTCAAGCGAAAAATTAGAATTAGTTTTTAGTGAAGAGATATTGCCTTATCTTTTCCAGTTAAAAAAATTCATAAAATATAATCTGGAACATGTTAAGTCTTTTGAAAACAAATACTCTATGAGGATTTATGAGTGGTTATTAAAAGAACTAACACAAAAGAAAACTCACAAGGCAAATATAGAGATTAGCCTTGATGAATTTAAGTTCATGTTAATGCTTGAAAATAACTACCATGAGTTTAAAAGGCTTAACCAATGGGTTTTGAAACCAATAAGTAAAGATTTAAACACTTACAGCAATATGAAATTGGTGGTTGATAAGCGAGGCCGCCCGACTGATACGTTGATTTTCCAAGTTGAACTAGATAGACAAATGGATCTCGTAACCGAACTTGAGAACAACCAGATAAAAATGAATGGTGACAAAATACCAACAACCATTACATCAGATTCCTACCTACGTAACGGACTAAGAAAAACACTACACGATGCTTTAACTGCAAAAATTCAGCTCACCAGTTTTGAGGCAAAATTTTTGAGTGACATGCAAAGTAAGCATGATCTCAATGGTTCGTTCTCATGGCTCACGCAAAAACAACGAACCACACTAGAGAACATACTGGCTAAATACGGAAGGATCTGA\u001b[39m\u001b[38;2;180;171;172mGGTTCTTATGGCTCTTGTATCTATCAGTGAAGCATCAAGACTAACAAACAAAAGTAGAACAACTGTTCACCGTTAGATATCAAAGGGAAAACTGTCCATATGCACAGATGAAAACGGTGTAAAAAAGATAGATACATCAGAGCTTTTACGAGTTTTTGGTGCATTTAAAGCTGTTCACCATGAACAGATCGACAATGTAAC\u001b[39mTACTAGAGCTAAATACA\n"
     ]
    }
   ],
   "source": [
    "df.print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be6a9262-49ac-4820-8851-002b2634ab96",
   "metadata": {},
   "source": [
    "The underlying sequence can be sliced like a string."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "85ada3aa-548b-4cbe-9f55-1dc1bfc99da9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<class 'gene_editor.plasmid'> at 0x7f35cc4d74f0\n",
       "molecule_type:DNA\n",
       "topology:circular\n",
       "data_file_division:   \n",
       "date:05-DEC-2022\n",
       "accessions:['<unknown', 'id>']\n",
       "keywords:['']\n",
       "source:\n",
       "organism:. .\n",
       "taxonomy:[]\n",
       "comment:\n",
       "ApEinfo:methylated:1\n",
       "                              locus_tag          type        location  length  \\\n",
       "0                         AmpR promoter      promoter       [0:29](+)      29   \n",
       "1                              AmpR RBS           RBS      [29:70](+)      41   \n",
       "2                          SpectomycinR           CDS     [70:868](+)     798   \n",
       "3                             BBa_B0053    terminator    [878:948](+)      70   \n",
       "4   J23108, 0.51, Constitutive Promoter      promoter  [1032:1067](+)      35   \n",
       "5          RBS 1.00 strength, BBa_B0034           RBS  [1086:1098](+)      12   \n",
       "6                       LacI, BBa_C0012           CDS  [1104:2196](+)    1092   \n",
       "7                             Esp3I_fix  misc_feature  [2072:2078](+)       6   \n",
       "8         Forward Terminator, BBa_B1002    terminator  [2196:2230](+)      34   \n",
       "9   pLacO2, single operon pLac ZC082818      promoter  [2230:2268](+)      38   \n",
       "10                                LacO2  protein_bind  [2238:2255](+)      17   \n",
       "11                 RFP targeting region  misc_binding  [2268:2288](+)      20   \n",
       "12         spec/1_xRFPg1_0_xRFPg1_0_.gb       unknown  [2268:2371](+)     103   \n",
       "13      dCas9 handle conserved region A  protein_bind  [2288:2300](+)      12   \n",
       "14     dCas9 handle conserved region A'  protein_bind  [2304:2318](+)      14   \n",
       "15      dCas9 handle conserved region B  protein_bind  [2318:2330](+)      12   \n",
       "16                   S. pyog terminator    terminator  [2330:2371](+)      41   \n",
       "17        Forward Terminator, BBa_B1010    terminator  [2371:2411](+)      40   \n",
       "\n",
       "      color  \n",
       "0   \u001b[38;2;15;127;254m#0f7ffe\u001b[39m  \n",
       "1      \u001b[38;2;0;255;255mcyan\u001b[39m  \n",
       "2   \u001b[38;2;212;0;55m#d40037\u001b[39m  \n",
       "3   \u001b[38;2;157;27;28m#9d1b1c\u001b[39m  \n",
       "4   \u001b[38;2;64;128;0m#408000\u001b[39m  \n",
       "5      \u001b[38;2;0;255;255mcyan\u001b[39m  \n",
       "6   \u001b[38;2;252;102;101m#fc6665\u001b[39m  \n",
       "7      \u001b[38;2;0;255;255mcyan\u001b[39m  \n",
       "8   \u001b[38;2;128;64;0m#804000\u001b[39m  \n",
       "9   \u001b[38;2;0;128;128m#008080\u001b[39m  \n",
       "10  \u001b[38;2;7;63;128m#073f80\u001b[39m  \n",
       "11  \u001b[38;2;255;0;0m#ff0000\u001b[39m  \n",
       "12     \u001b[38;2;0;255;255mcyan\u001b[39m  \n",
       "13  \u001b[38;2;0;128;255m#0080ff\u001b[39m  \n",
       "14  \u001b[38;2;0;128;255m#0080ff\u001b[39m  \n",
       "15  \u001b[38;2;0;64;128m#004080\u001b[39m  \n",
       "16  \u001b[38;2;128;64;0m#804000\u001b[39m  \n",
       "17  \u001b[38;2;128;64;0m#804000\u001b[39m  \n",
       "total length:2411"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[0:2411]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "id": "906e59b1-6e5c-45d6-aa70-0875e5a68feb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[38;2;15;127;254mTTCAAATATCTATCCGCTCATGAGACAAT\u001b[39m\u001b[38;2;0;255;255mAACCCTGATAAATGCTTCAATAATATTGAAAAAGGAAGAAT\u001b[39m\u001b[38;2;212;0;55mATGAGTGAAAAAGTGCCCGCCGAGATTTCGGTGCAACTATCACAAGCACTCAACGTCATCGGGCGCCACTTGGAGTCGACGTTGCTGGCCGTGCATTTGTACGGCTCCGCACTGGATGGCGGATTGAAACCGTACAGTGATATTGATTTGCTGGTGACTGTAGCTGCACCGCTCAATGATGCCGTGCGGCAAGCCCTGCTCGTCGATCTCTTGGAGGTTTCAGCTTCCCCTGGCCAAAACAAGGCACTCCGCGCCTTGGAAGTGACCATCGTCGTGCACAGTGACATCGTACCTTGGCGTTATCCGGCCAGGCGGGAACTGCAGTTCGGAGAGTGGCAGCGCAAAGACATCCTTGCGGGCATCTTCGAGCCCGCCACAACCGATTCTGACTTGGCGATTCTGCTAACAAAGGCAAAGCAACATAGCGTCGTCTTGGCAGGTTCAGCAGCGAAGGATCTCTTCAGCTCAGTCCCAGAAAGCGATCTATTCAAGGCACTGGCCGATACTCTGAAGCTATGGAACTCGCCGCCAGATTGGGCGGGCGATGAGCGGAATGTAGTGCTTACTTTGTCTCGTATCTGGTACACCGCAGCAACCGGCAAGATCGCGCCAAAGGATGTTGCTGCCACTTGGGCAATGGCACGCTTGCCAGCTCAACATCAGCCCATCCTGTTGAATGCCAAGCGGGCTTATCTTGGGCAAGAAGAAGATTATTTGCCCGCTCGTGCGGATCAGGTGGCGGCGCTCATTAAATTCGTGAAGTATGAAGCAGTTAAACTGCTTGGTGCCAGCCAATGA\u001b[39mTAATACTAGC\u001b[38;2;157;27;28mTCCGGCAAAAAAACGGGCAAGGTGTCACCACCCTGCCCTTTTTCTTTAAAACCGAAAAGATTACTTCGCG\u001b[39mAAATAGGCGTATCACGAGGCAGAATTTCAGATAAAAAAAATCCTTAGCTTTCGCTAAGGATGATTTCTGGAATTCGCGGCCGCT\u001b[38;2;64;128;0mCTGACAGCTAGCTCAGTCCTAGGTATAATGCTAGC\u001b[39mAAAAGAATTCAAAAGATCT\u001b[38;2;0;255;255mAAAGAGGAGAAA\u001b[39mGGATCT\u001b[38;2;252;102;101mATGGTGAATGTGAAACCAGTAACGTTATACGATGTCGCAGAGTATGCCGGTGTCTCTTATCAGACCGTTTCCCGCGTGGTGAACCAGGCCAGCCACGTTTCTGCGAAAACGCGGGAAAAAGTGGAAGCGGCGATGGCGGAGCTGAATTACATTCCCAACCGCGTGGCACAACAACTGGCGGGCAAACAGTCGTTGCTGATTGGCGTTGCCACCTCCAGTCTGGCCCTGCACGCGCCGTCGCAAATTGTCGCGGCGATTAAATCTCGCGCCGATCAACTGGGTGCCAGCGTGGTGGTGTCGATGGTAGAACGAAGCGGCGTCGAAGCCTGTAAAGCGGCGGTGCACAATCTTCTCGCGCAACGCGTCAGTGGGCTGATCATTAACTATCCGCTGGATGACCAGGATGCCATTGCTGTGGAAGCTGCCTGCACTAATGTTCCGGCGTTATTTCTTGATGTCTCTGACCAGACACCCATCAACAGTATTATTTTCTCCCATGAAGACGGTACGCGACTGGGCGTGGAGCATCTGGTCGCATTGGGTCACCAGCAAATCGCGCTGTTAGCGGGCCCATTAAGTTCTGTCTCGGCGCGTCTGCGTCTGGCTGGCTGGCATAAATATCTCACTCGCAATCAAATTCAGCCGATAGCGGAACGGGAAGGCGACTGGAGTGCCATGTCCGGTTTTCAACAAACCATGCAAATGCTGAATGAGGGCATCGTTCCCACTGCGATGCTGGTTGCCAACGATCAGATGGCGCTGGGCGCAATGCGCGCCATTACCGAGTCCGGGCTGCGCGTTGGTGCGGATATCTCGGTAGTGGGATACGACGATACCGAAGACAGCTCATGTTATATCCCGCCGTTAACCACCATCAAACAGGATTTTCGCCTGCTGGGGCAAACCAGCGTGGACCGCTTGCTGCAACTCTCTCAGGGCCAGGCGGTGAAGGGCAATCAGCTGTTGCC\u001b[39m\u001b[38;2;0;255;255mCGTTTC\u001b[39m\u001b[38;2;252;102;101mACTGGTGAAAAGAAAAACCACCCTGGCGCCCAATACGCGAACCGCCTCTCCCCGCGCGTTGGCCGATTCATTAATGCAGCTGGCACGACAGGTTTCCCGACTGGAAAGCGGGCAGTAA\u001b[39m\u001b[38;2;128;64;0mCGCAAAAAACCCCGCTTCGGCGGGGTTTTTTCGC\u001b[39m\u001b[38;2;0;128;128mAATTGACA\u001b[39m\u001b[38;2;7;63;128mATGTGAGCGAGTAACAA\u001b[39m\u001b[38;2;0;128;128mGATACTGAGCACA\u001b[39m\u001b[38;2;255;0;0mAACTTTCAGTTTAGCGGTCT\u001b[39m\u001b[38;2;0;128;255mGTTTTAGAGCTA\u001b[39m\u001b[38;2;0;255;255mGAAA\u001b[39m\u001b[38;2;0;128;255mTAGCAAGTTAAAAT\u001b[39m\u001b[38;2;0;64;128mAAGGCTAGTCCG\u001b[39m\u001b[38;2;128;64;0mTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTTCGCCGCAAACCCCGCCCCTGACAGGGCGGGGTTTCGCCGC\u001b[39m\n"
     ]
    }
   ],
   "source": [
    "df[0:2411].print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c6613ad-a767-47aa-8777-a69d7170b7aa",
   "metadata": {},
   "source": [
    "Plasmids are circular, not linear. Therefore, when we view a plasmids, it is from a certain origin or reference position. We can choose a new reference position to view the plasmid via `.set_origin`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "df48bc41-6f6b-439c-ba2d-444392377108",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'gene_editor.plasmid'> at 0x7f35cc36caf0\n",
      "molecule_type:DNA\n",
      "topology:circular\n",
      "data_file_division:   \n",
      "date:05-DEC-2022\n",
      "accessions:['<unknown', 'id>']\n",
      "keywords:['']\n",
      "source:\n",
      "organism:. .\n",
      "taxonomy:[]\n",
      "comment:\n",
      "ApEinfo:methylated:1\n",
      "                              locus_tag          type        location  length  \\\n",
      "0                         AmpR promoter      promoter       [0:29](+)      29   \n",
      "1                              AmpR RBS           RBS      [29:70](+)      41   \n",
      "2                          SpectomycinR           CDS     [70:868](+)     798   \n",
      "3                             BBa_B0053    terminator    [878:948](+)      70   \n",
      "4   J23108, 0.51, Constitutive Promoter      promoter  [1032:1067](+)      35   \n",
      "5          RBS 1.00 strength, BBa_B0034           RBS  [1086:1098](+)      12   \n",
      "6                       LacI, BBa_C0012           CDS  [1104:2196](+)    1092   \n",
      "7                             Esp3I_fix  misc_feature  [2072:2078](+)       6   \n",
      "8         Forward Terminator, BBa_B1002    terminator  [2196:2230](+)      34   \n",
      "9   pLacO2, single operon pLac ZC082818      promoter  [2230:2268](+)      38   \n",
      "10                                LacO2  protein_bind  [2238:2255](+)      17   \n",
      "11                 RFP targeting region  misc_binding  [2268:2288](+)      20   \n",
      "12         spec/1_xRFPg1_0_xRFPg1_0_.gb       unknown  [2268:2371](+)     103   \n",
      "13      dCas9 handle conserved region A  protein_bind  [2288:2300](+)      12   \n",
      "14     dCas9 handle conserved region A'  protein_bind  [2304:2318](+)      14   \n",
      "15      dCas9 handle conserved region B  protein_bind  [2318:2330](+)      12   \n",
      "16                   S. pyog terminator    terminator  [2330:2371](+)      41   \n",
      "17        Forward Terminator, BBa_B1010    terminator  [2371:2411](+)      40   \n",
      "\n",
      "      color  \n",
      "0   \u001b[38;2;15;127;254m#0f7ffe\u001b[39m  \n",
      "1      \u001b[38;2;0;255;255mcyan\u001b[39m  \n",
      "2   \u001b[38;2;212;0;55m#d40037\u001b[39m  \n",
      "3   \u001b[38;2;157;27;28m#9d1b1c\u001b[39m  \n",
      "4   \u001b[38;2;64;128;0m#408000\u001b[39m  \n",
      "5      \u001b[38;2;0;255;255mcyan\u001b[39m  \n",
      "6   \u001b[38;2;252;102;101m#fc6665\u001b[39m  \n",
      "7      \u001b[38;2;0;255;255mcyan\u001b[39m  \n",
      "8   \u001b[38;2;128;64;0m#804000\u001b[39m  \n",
      "9   \u001b[38;2;0;128;128m#008080\u001b[39m  \n",
      "10  \u001b[38;2;7;63;128m#073f80\u001b[39m  \n",
      "11  \u001b[38;2;255;0;0m#ff0000\u001b[39m  \n",
      "12     \u001b[38;2;0;255;255mcyan\u001b[39m  \n",
      "13  \u001b[38;2;0;128;255m#0080ff\u001b[39m  \n",
      "14  \u001b[38;2;0;128;255m#0080ff\u001b[39m  \n",
      "15  \u001b[38;2;0;64;128m#004080\u001b[39m  \n",
      "16  \u001b[38;2;128;64;0m#804000\u001b[39m  \n",
      "17  \u001b[38;2;128;64;0m#804000\u001b[39m  \n",
      "total length:2411\n",
      "\n",
      "\u001b[38;2;15;127;254mTTCAAATATCTATCCGCTCATGAGACAAT\u001b[39m\u001b[38;2;0;255;255mAACCCTGATAAATGCTTCAATAATATTGAAAAAGGAAGAAT\u001b[39m\u001b[38;2;212;0;55mATGAGTGAAAAAGTGCCCGCCGAGATTTCGGTGCAACTATCACAAGCACTCAACGTCATCGGGCGCCACTTGGAGTCGACGTTGCTGGCCGTGCATTTGTACGGCTCCGCACTGGATGGCGGATTGAAACCGTACAGTGATATTGATTTGCTGGTGACTGTAGCTGCACCGCTCAATGATGCCGTGCGGCAAGCCCTGCTCGTCGATCTCTTGGAGGTTTCAGCTTCCCCTGGCCAAAACAAGGCACTCCGCGCCTTGGAAGTGACCATCGTCGTGCACAGTGACATCGTACCTTGGCGTTATCCGGCCAGGCGGGAACTGCAGTTCGGAGAGTGGCAGCGCAAAGACATCCTTGCGGGCATCTTCGAGCCCGCCACAACCGATTCTGACTTGGCGATTCTGCTAACAAAGGCAAAGCAACATAGCGTCGTCTTGGCAGGTTCAGCAGCGAAGGATCTCTTCAGCTCAGTCCCAGAAAGCGATCTATTCAAGGCACTGGCCGATACTCTGAAGCTATGGAACTCGCCGCCAGATTGGGCGGGCGATGAGCGGAATGTAGTGCTTACTTTGTCTCGTATCTGGTACACCGCAGCAACCGGCAAGATCGCGCCAAAGGATGTTGCTGCCACTTGGGCAATGGCACGCTTGCCAGCTCAACATCAGCCCATCCTGTTGAATGCCAAGCGGGCTTATCTTGGGCAAGAAGAAGATTATTTGCCCGCTCGTGCGGATCAGGTGGCGGCGCTCATTAAATTCGTGAAGTATGAAGCAGTTAAACTGCTTGGTGCCAGCCAATGA\u001b[39mTAATACTAGC\u001b[38;2;157;27;28mTCCGGCAAAAAAACGGGCAAGGTGTCACCACCCTGCCCTTTTTCTTTAAAACCGAAAAGATTACTTCGCG\u001b[39mAAATAGGCGTATCACGAGGCAGAATTTCAGATAAAAAAAATCCTTAGCTTTCGCTAAGGATGATTTCTGGAATTCGCGGCCGCT\u001b[38;2;64;128;0mCTGACAGCTAGCTCAGTCCTAGGTATAATGCTAGC\u001b[39mAAAAGAATTCAAAAGATCT\u001b[38;2;0;255;255mAAAGAGGAGAAA\u001b[39mGGATCT\u001b[38;2;252;102;101mATGGTGAATGTGAAACCAGTAACGTTATACGATGTCGCAGAGTATGCCGGTGTCTCTTATCAGACCGTTTCCCGCGTGGTGAACCAGGCCAGCCACGTTTCTGCGAAAACGCGGGAAAAAGTGGAAGCGGCGATGGCGGAGCTGAATTACATTCCCAACCGCGTGGCACAACAACTGGCGGGCAAACAGTCGTTGCTGATTGGCGTTGCCACCTCCAGTCTGGCCCTGCACGCGCCGTCGCAAATTGTCGCGGCGATTAAATCTCGCGCCGATCAACTGGGTGCCAGCGTGGTGGTGTCGATGGTAGAACGAAGCGGCGTCGAAGCCTGTAAAGCGGCGGTGCACAATCTTCTCGCGCAACGCGTCAGTGGGCTGATCATTAACTATCCGCTGGATGACCAGGATGCCATTGCTGTGGAAGCTGCCTGCACTAATGTTCCGGCGTTATTTCTTGATGTCTCTGACCAGACACCCATCAACAGTATTATTTTCTCCCATGAAGACGGTACGCGACTGGGCGTGGAGCATCTGGTCGCATTGGGTCACCAGCAAATCGCGCTGTTAGCGGGCCCATTAAGTTCTGTCTCGGCGCGTCTGCGTCTGGCTGGCTGGCATAAATATCTCACTCGCAATCAAATTCAGCCGATAGCGGAACGGGAAGGCGACTGGAGTGCCATGTCCGGTTTTCAACAAACCATGCAAATGCTGAATGAGGGCATCGTTCCCACTGCGATGCTGGTTGCCAACGATCAGATGGCGCTGGGCGCAATGCGCGCCATTACCGAGTCCGGGCTGCGCGTTGGTGCGGATATCTCGGTAGTGGGATACGACGATACCGAAGACAGCTCATGTTATATCCCGCCGTTAACCACCATCAAACAGGATTTTCGCCTGCTGGGGCAAACCAGCGTGGACCGCTTGCTGCAACTCTCTCAGGGCCAGGCGGTGAAGGGCAATCAGCTGTTGCC\u001b[39m\u001b[38;2;0;255;255mCGTTTC\u001b[39m\u001b[38;2;252;102;101mACTGGTGAAAAGAAAAACCACCCTGGCGCCCAATACGCGAACCGCCTCTCCCCGCGCGTTGGCCGATTCATTAATGCAGCTGGCACGACAGGTTTCCCGACTGGAAAGCGGGCAGTAA\u001b[39m\u001b[38;2;128;64;0mCGCAAAAAACCCCGCTTCGGCGGGGTTTTTTCGC\u001b[39m\u001b[38;2;0;128;128mAATTGACA\u001b[39m\u001b[38;2;7;63;128mATGTGAGCGAGTAACAA\u001b[39m\u001b[38;2;0;128;128mGATACTGAGCACA\u001b[39m\u001b[38;2;255;0;0mAACTTTCAGTTTAGCGGTCT\u001b[39m\u001b[38;2;0;128;255mGTTTTAGAGCTA\u001b[39m\u001b[38;2;0;255;255mGAAA\u001b[39m\u001b[38;2;0;128;255mTAGCAAGTTAAAAT\u001b[39m\u001b[38;2;0;64;128mAAGGCTAGTCCG\u001b[39m\u001b[38;2;128;64;0mTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTTCGCCGCAAACCCCGCCCCTGACAGGGCGGGGTTTCGCCGC\u001b[39m\n"
     ]
    }
   ],
   "source": [
    "x = df[0:2411]\n",
    "print(x.__repr__())\n",
    "x.print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "429e75a2-5413-4db3-b898-33fc497569d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'gene_editor.plasmid'> at 0x7f35cc36cc10\n",
      "molecule_type:DNA\n",
      "topology:circular\n",
      "data_file_division:   \n",
      "date:05-DEC-2022\n",
      "accessions:['<unknown', 'id>']\n",
      "keywords:['']\n",
      "source:\n",
      "organism:. .\n",
      "taxonomy:[]\n",
      "comment:\n",
      "ApEinfo:methylated:1\n",
      "                              locus_tag          type        location  length  \\\n",
      "0   J23108, 0.51, Constitutive Promoter      promoter       [0:35](+)      35   \n",
      "1          RBS 1.00 strength, BBa_B0034           RBS      [54:66](+)      12   \n",
      "2                       LacI, BBa_C0012           CDS    [72:1164](+)    1092   \n",
      "3                             Esp3I_fix  misc_feature  [1040:1046](+)       6   \n",
      "4         Forward Terminator, BBa_B1002    terminator  [1164:1198](+)      34   \n",
      "5   pLacO2, single operon pLac ZC082818      promoter  [1198:1236](+)      38   \n",
      "6                                 LacO2  protein_bind  [1206:1223](+)      17   \n",
      "7                  RFP targeting region  misc_binding  [1236:1256](+)      20   \n",
      "8          spec/1_xRFPg1_0_xRFPg1_0_.gb       unknown  [1236:1339](+)     103   \n",
      "9       dCas9 handle conserved region A  protein_bind  [1256:1268](+)      12   \n",
      "10     dCas9 handle conserved region A'  protein_bind  [1272:1286](+)      14   \n",
      "11      dCas9 handle conserved region B  protein_bind  [1286:1298](+)      12   \n",
      "12                   S. pyog terminator    terminator  [1298:1339](+)      41   \n",
      "13        Forward Terminator, BBa_B1010    terminator  [1339:1379](+)      40   \n",
      "14                        AmpR promoter      promoter  [1379:1408](+)      29   \n",
      "15                             AmpR RBS           RBS  [1408:1449](+)      41   \n",
      "16                         SpectomycinR           CDS  [1449:2247](+)     798   \n",
      "17                            BBa_B0053    terminator  [2257:2327](+)      70   \n",
      "\n",
      "      color  \n",
      "0   \u001b[38;2;64;128;0m#408000\u001b[39m  \n",
      "1      \u001b[38;2;0;255;255mcyan\u001b[39m  \n",
      "2   \u001b[38;2;252;102;101m#fc6665\u001b[39m  \n",
      "3      \u001b[38;2;0;255;255mcyan\u001b[39m  \n",
      "4   \u001b[38;2;128;64;0m#804000\u001b[39m  \n",
      "5   \u001b[38;2;0;128;128m#008080\u001b[39m  \n",
      "6   \u001b[38;2;7;63;128m#073f80\u001b[39m  \n",
      "7   \u001b[38;2;255;0;0m#ff0000\u001b[39m  \n",
      "8      \u001b[38;2;0;255;255mcyan\u001b[39m  \n",
      "9   \u001b[38;2;0;128;255m#0080ff\u001b[39m  \n",
      "10  \u001b[38;2;0;128;255m#0080ff\u001b[39m  \n",
      "11  \u001b[38;2;0;64;128m#004080\u001b[39m  \n",
      "12  \u001b[38;2;128;64;0m#804000\u001b[39m  \n",
      "13  \u001b[38;2;128;64;0m#804000\u001b[39m  \n",
      "14  \u001b[38;2;15;127;254m#0f7ffe\u001b[39m  \n",
      "15     \u001b[38;2;0;255;255mcyan\u001b[39m  \n",
      "16  \u001b[38;2;212;0;55m#d40037\u001b[39m  \n",
      "17  \u001b[38;2;157;27;28m#9d1b1c\u001b[39m  \n",
      "total length:2411\n",
      "\n",
      "\u001b[38;2;64;128;0mCTGACAGCTAGCTCAGTCCTAGGTATAATGCTAGC\u001b[39mAAAAGAATTCAAAAGATCT\u001b[38;2;0;255;255mAAAGAGGAGAAA\u001b[39mGGATCT\u001b[38;2;252;102;101mATGGTGAATGTGAAACCAGTAACGTTATACGATGTCGCAGAGTATGCCGGTGTCTCTTATCAGACCGTTTCCCGCGTGGTGAACCAGGCCAGCCACGTTTCTGCGAAAACGCGGGAAAAAGTGGAAGCGGCGATGGCGGAGCTGAATTACATTCCCAACCGCGTGGCACAACAACTGGCGGGCAAACAGTCGTTGCTGATTGGCGTTGCCACCTCCAGTCTGGCCCTGCACGCGCCGTCGCAAATTGTCGCGGCGATTAAATCTCGCGCCGATCAACTGGGTGCCAGCGTGGTGGTGTCGATGGTAGAACGAAGCGGCGTCGAAGCCTGTAAAGCGGCGGTGCACAATCTTCTCGCGCAACGCGTCAGTGGGCTGATCATTAACTATCCGCTGGATGACCAGGATGCCATTGCTGTGGAAGCTGCCTGCACTAATGTTCCGGCGTTATTTCTTGATGTCTCTGACCAGACACCCATCAACAGTATTATTTTCTCCCATGAAGACGGTACGCGACTGGGCGTGGAGCATCTGGTCGCATTGGGTCACCAGCAAATCGCGCTGTTAGCGGGCCCATTAAGTTCTGTCTCGGCGCGTCTGCGTCTGGCTGGCTGGCATAAATATCTCACTCGCAATCAAATTCAGCCGATAGCGGAACGGGAAGGCGACTGGAGTGCCATGTCCGGTTTTCAACAAACCATGCAAATGCTGAATGAGGGCATCGTTCCCACTGCGATGCTGGTTGCCAACGATCAGATGGCGCTGGGCGCAATGCGCGCCATTACCGAGTCCGGGCTGCGCGTTGGTGCGGATATCTCGGTAGTGGGATACGACGATACCGAAGACAGCTCATGTTATATCCCGCCGTTAACCACCATCAAACAGGATTTTCGCCTGCTGGGGCAAACCAGCGTGGACCGCTTGCTGCAACTCTCTCAGGGCCAGGCGGTGAAGGGCAATCAGCTGTTGCC\u001b[39m\u001b[38;2;0;255;255mCGTTTC\u001b[39m\u001b[38;2;252;102;101mACTGGTGAAAAGAAAAACCACCCTGGCGCCCAATACGCGAACCGCCTCTCCCCGCGCGTTGGCCGATTCATTAATGCAGCTGGCACGACAGGTTTCCCGACTGGAAAGCGGGCAGTAA\u001b[39m\u001b[38;2;128;64;0mCGCAAAAAACCCCGCTTCGGCGGGGTTTTTTCGC\u001b[39m\u001b[38;2;0;128;128mAATTGACA\u001b[39m\u001b[38;2;7;63;128mATGTGAGCGAGTAACAA\u001b[39m\u001b[38;2;0;128;128mGATACTGAGCACA\u001b[39m\u001b[38;2;255;0;0mAACTTTCAGTTTAGCGGTCT\u001b[39m\u001b[38;2;0;128;255mGTTTTAGAGCTA\u001b[39m\u001b[38;2;0;255;255mGAAA\u001b[39m\u001b[38;2;0;128;255mTAGCAAGTTAAAAT\u001b[39m\u001b[38;2;0;64;128mAAGGCTAGTCCG\u001b[39m\u001b[38;2;128;64;0mTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTTCGCCGCAAACCCCGCCCCTGACAGGGCGGGGTTTCGCCGC\u001b[39m\u001b[38;2;15;127;254mTTCAAATATCTATCCGCTCATGAGACAAT\u001b[39m\u001b[38;2;0;255;255mAACCCTGATAAATGCTTCAATAATATTGAAAAAGGAAGAAT\u001b[39m\u001b[38;2;212;0;55mATGAGTGAAAAAGTGCCCGCCGAGATTTCGGTGCAACTATCACAAGCACTCAACGTCATCGGGCGCCACTTGGAGTCGACGTTGCTGGCCGTGCATTTGTACGGCTCCGCACTGGATGGCGGATTGAAACCGTACAGTGATATTGATTTGCTGGTGACTGTAGCTGCACCGCTCAATGATGCCGTGCGGCAAGCCCTGCTCGTCGATCTCTTGGAGGTTTCAGCTTCCCCTGGCCAAAACAAGGCACTCCGCGCCTTGGAAGTGACCATCGTCGTGCACAGTGACATCGTACCTTGGCGTTATCCGGCCAGGCGGGAACTGCAGTTCGGAGAGTGGCAGCGCAAAGACATCCTTGCGGGCATCTTCGAGCCCGCCACAACCGATTCTGACTTGGCGATTCTGCTAACAAAGGCAAAGCAACATAGCGTCGTCTTGGCAGGTTCAGCAGCGAAGGATCTCTTCAGCTCAGTCCCAGAAAGCGATCTATTCAAGGCACTGGCCGATACTCTGAAGCTATGGAACTCGCCGCCAGATTGGGCGGGCGATGAGCGGAATGTAGTGCTTACTTTGTCTCGTATCTGGTACACCGCAGCAACCGGCAAGATCGCGCCAAAGGATGTTGCTGCCACTTGGGCAATGGCACGCTTGCCAGCTCAACATCAGCCCATCCTGTTGAATGCCAAGCGGGCTTATCTTGGGCAAGAAGAAGATTATTTGCCCGCTCGTGCGGATCAGGTGGCGGCGCTCATTAAATTCGTGAAGTATGAAGCAGTTAAACTGCTTGGTGCCAGCCAATGA\u001b[39mTAATACTAGC\u001b[38;2;157;27;28mTCCGGCAAAAAAACGGGCAAGGTGTCACCACCCTGCCCTTTTTCTTTAAAACCGAAAAGATTACTTCGCG\u001b[39mAAATAGGCGTATCACGAGGCAGAATTTCAGATAAAAAAAATCCTTAGCTTTCGCTAAGGATGATTTCTGGAATTCGCGGCCGCT\n"
     ]
    }
   ],
   "source": [
    "x = x.set_origin(1032)\n",
    "print(x.__repr__())\n",
    "x.print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "362a43e5-70eb-4d95-bf8d-6b5458931e49",
   "metadata": {},
   "source": [
    "Genes can also be selected and sliced out of the dataframe via bool selection like in pandas "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f83b2d8f-0b89-4215-a0dc-f6cf2a755b3d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'gene_editor.plasmid'> at 0x7f35cc385f00\n",
      "molecule_type:DNA\n",
      "topology:circular\n",
      "data_file_division:   \n",
      "date:05-DEC-2022\n",
      "accessions:['<unknown', 'id>']\n",
      "keywords:['']\n",
      "source:\n",
      "organism:. .\n",
      "taxonomy:[]\n",
      "comment:\n",
      "ApEinfo:methylated:1\n",
      "                       locus_tag        type        location  length    color\n",
      "0  Forward Terminator, BBa_B1002  terminator  [1164:1198](+)      34  \u001b[38;2;128;64;0m#804000\u001b[39m\n",
      "1             S. pyog terminator  terminator  [1298:1339](+)      41  \u001b[38;2;128;64;0m#804000\u001b[39m\n",
      "2  Forward Terminator, BBa_B1010  terminator  [1339:1379](+)      40  \u001b[38;2;128;64;0m#804000\u001b[39m\n",
      "3                      BBa_B0053  terminator  [2257:2327](+)      70  \u001b[38;2;157;27;28m#9d1b1c\u001b[39m\n",
      "total length:2411\n",
      "\n",
      "CTGACAGCTAGCTCAGTCCTAGGTATAATGCTAGCAAAAGAATTCAAAAGATCTAAAGAGGAGAAAGGATCTATGGTGAATGTGAAACCAGTAACGTTATACGATGTCGCAGAGTATGCCGGTGTCTCTTATCAGACCGTTTCCCGCGTGGTGAACCAGGCCAGCCACGTTTCTGCGAAAACGCGGGAAAAAGTGGAAGCGGCGATGGCGGAGCTGAATTACATTCCCAACCGCGTGGCACAACAACTGGCGGGCAAACAGTCGTTGCTGATTGGCGTTGCCACCTCCAGTCTGGCCCTGCACGCGCCGTCGCAAATTGTCGCGGCGATTAAATCTCGCGCCGATCAACTGGGTGCCAGCGTGGTGGTGTCGATGGTAGAACGAAGCGGCGTCGAAGCCTGTAAAGCGGCGGTGCACAATCTTCTCGCGCAACGCGTCAGTGGGCTGATCATTAACTATCCGCTGGATGACCAGGATGCCATTGCTGTGGAAGCTGCCTGCACTAATGTTCCGGCGTTATTTCTTGATGTCTCTGACCAGACACCCATCAACAGTATTATTTTCTCCCATGAAGACGGTACGCGACTGGGCGTGGAGCATCTGGTCGCATTGGGTCACCAGCAAATCGCGCTGTTAGCGGGCCCATTAAGTTCTGTCTCGGCGCGTCTGCGTCTGGCTGGCTGGCATAAATATCTCACTCGCAATCAAATTCAGCCGATAGCGGAACGGGAAGGCGACTGGAGTGCCATGTCCGGTTTTCAACAAACCATGCAAATGCTGAATGAGGGCATCGTTCCCACTGCGATGCTGGTTGCCAACGATCAGATGGCGCTGGGCGCAATGCGCGCCATTACCGAGTCCGGGCTGCGCGTTGGTGCGGATATCTCGGTAGTGGGATACGACGATACCGAAGACAGCTCATGTTATATCCCGCCGTTAACCACCATCAAACAGGATTTTCGCCTGCTGGGGCAAACCAGCGTGGACCGCTTGCTGCAACTCTCTCAGGGCCAGGCGGTGAAGGGCAATCAGCTGTTGCCCGTTTCACTGGTGAAAAGAAAAACCACCCTGGCGCCCAATACGCGAACCGCCTCTCCCCGCGCGTTGGCCGATTCATTAATGCAGCTGGCACGACAGGTTTCCCGACTGGAAAGCGGGCAGTAA\u001b[38;2;128;64;0mCGCAAAAAACCCCGCTTCGGCGGGGTTTTTTCGC\u001b[39mAATTGACAATGTGAGCGAGTAACAAGATACTGAGCACAAACTTTCAGTTTAGCGGTCTGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCG\u001b[38;2;128;64;0mTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTTCGCCGCAAACCCCGCCCCTGACAGGGCGGGGTTTCGCCGC\u001b[39mTTCAAATATCTATCCGCTCATGAGACAATAACCCTGATAAATGCTTCAATAATATTGAAAAAGGAAGAATATGAGTGAAAAAGTGCCCGCCGAGATTTCGGTGCAACTATCACAAGCACTCAACGTCATCGGGCGCCACTTGGAGTCGACGTTGCTGGCCGTGCATTTGTACGGCTCCGCACTGGATGGCGGATTGAAACCGTACAGTGATATTGATTTGCTGGTGACTGTAGCTGCACCGCTCAATGATGCCGTGCGGCAAGCCCTGCTCGTCGATCTCTTGGAGGTTTCAGCTTCCCCTGGCCAAAACAAGGCACTCCGCGCCTTGGAAGTGACCATCGTCGTGCACAGTGACATCGTACCTTGGCGTTATCCGGCCAGGCGGGAACTGCAGTTCGGAGAGTGGCAGCGCAAAGACATCCTTGCGGGCATCTTCGAGCCCGCCACAACCGATTCTGACTTGGCGATTCTGCTAACAAAGGCAAAGCAACATAGCGTCGTCTTGGCAGGTTCAGCAGCGAAGGATCTCTTCAGCTCAGTCCCAGAAAGCGATCTATTCAAGGCACTGGCCGATACTCTGAAGCTATGGAACTCGCCGCCAGATTGGGCGGGCGATGAGCGGAATGTAGTGCTTACTTTGTCTCGTATCTGGTACACCGCAGCAACCGGCAAGATCGCGCCAAAGGATGTTGCTGCCACTTGGGCAATGGCACGCTTGCCAGCTCAACATCAGCCCATCCTGTTGAATGCCAAGCGGGCTTATCTTGGGCAAGAAGAAGATTATTTGCCCGCTCGTGCGGATCAGGTGGCGGCGCTCATTAAATTCGTGAAGTATGAAGCAGTTAAACTGCTTGGTGCCAGCCAATGATAATACTAGC\u001b[38;2;157;27;28mTCCGGCAAAAAAACGGGCAAGGTGTCACCACCCTGCCCTTTTTCTTTAAAACCGAAAAGATTACTTCGCG\u001b[39mAAATAGGCGTATCACGAGGCAGAATTTCAGATAAAAAAAATCCTTAGCTTTCGCTAAGGATGATTTCTGGAATTCGCGGCCGCT\n"
     ]
    }
   ],
   "source": [
    "y = x[x['type']=='terminator']\n",
    "print(y.__repr__())\n",
    "y.print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "42bb6ff7-9b22-4585-9ea9-2184c727bc66",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'gene_editor.plasmid'> at 0x7f35cc386ec0\n",
      "molecule_type:DNA\n",
      "topology:circular\n",
      "data_file_division:   \n",
      "date:05-DEC-2022\n",
      "accessions:['<unknown', 'id>']\n",
      "keywords:['']\n",
      "source:\n",
      "organism:. .\n",
      "taxonomy:[]\n",
      "comment:\n",
      "ApEinfo:methylated:1\n",
      "         locus_tag type     location  length    color\n",
      "0  LacI, BBa_C0012  CDS  [0:1092](+)    1092  \u001b[38;2;252;102;101m#fc6665\u001b[39m\n",
      "total length:1092\n",
      "\n",
      "\u001b[38;2;252;102;101mATGGTGAATGTGAAACCAGTAACGTTATACGATGTCGCAGAGTATGCCGGTGTCTCTTATCAGACCGTTTCCCGCGTGGTGAACCAGGCCAGCCACGTTTCTGCGAAAACGCGGGAAAAAGTGGAAGCGGCGATGGCGGAGCTGAATTACATTCCCAACCGCGTGGCACAACAACTGGCGGGCAAACAGTCGTTGCTGATTGGCGTTGCCACCTCCAGTCTGGCCCTGCACGCGCCGTCGCAAATTGTCGCGGCGATTAAATCTCGCGCCGATCAACTGGGTGCCAGCGTGGTGGTGTCGATGGTAGAACGAAGCGGCGTCGAAGCCTGTAAAGCGGCGGTGCACAATCTTCTCGCGCAACGCGTCAGTGGGCTGATCATTAACTATCCGCTGGATGACCAGGATGCCATTGCTGTGGAAGCTGCCTGCACTAATGTTCCGGCGTTATTTCTTGATGTCTCTGACCAGACACCCATCAACAGTATTATTTTCTCCCATGAAGACGGTACGCGACTGGGCGTGGAGCATCTGGTCGCATTGGGTCACCAGCAAATCGCGCTGTTAGCGGGCCCATTAAGTTCTGTCTCGGCGCGTCTGCGTCTGGCTGGCTGGCATAAATATCTCACTCGCAATCAAATTCAGCCGATAGCGGAACGGGAAGGCGACTGGAGTGCCATGTCCGGTTTTCAACAAACCATGCAAATGCTGAATGAGGGCATCGTTCCCACTGCGATGCTGGTTGCCAACGATCAGATGGCGCTGGGCGCAATGCGCGCCATTACCGAGTCCGGGCTGCGCGTTGGTGCGGATATCTCGGTAGTGGGATACGACGATACCGAAGACAGCTCATGTTATATCCCGCCGTTAACCACCATCAAACAGGATTTTCGCCTGCTGGGGCAAACCAGCGTGGACCGCTTGCTGCAACTCTCTCAGGGCCAGGCGGTGAAGGGCAATCAGCTGTTGCCCGTTTCACTGGTGAAAAGAAAAACCACCCTGGCGCCCAATACGCGAACCGCCTCTCCCCGCGCGTTGGCCGATTCATTAATGCAGCTGGCACGACAGGTTTCCCGACTGGAAAGCGGGCAGTAA\u001b[39m\n"
     ]
    }
   ],
   "source": [
    "y = x[x['locus_tag'].str.contains('LacI')]\n",
    "y = y.slice()\n",
    "print(y.__repr__())\n",
    "y.print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34ae073c-12ae-457e-99b7-e528b3137561",
   "metadata": {},
   "source": [
    "DNA sequences can be translated to amino acid sequences via `.translate()`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0f7f77d2-b284-4af9-a633-d6b0da8b8ff6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[38;2;252;102;101mATGGTGAATGTGAAACCAGTAACGTTATACGATGTCGCAGAGTATGCCGGTGTCTCTTATCAGACCGTTTCCCGCGTGGTGAACCAGGCCAGCCACGTTTCTGCGAAAACGCGGGAAAAAGTGGAAGCGGCGATGGCGGAGCTGAATTACATTCCCAACCGCGTGGCACAACAACTGGCGGGCAAACAGTCGTTGCTGATTGGCGTTGCCACCTCCAGTCTGGCCCTGCACGCGCCGTCGCAAATTGTCGCGGCGATTAAATCTCGCGCCGATCAACTGGGTGCCAGCGTGGTGGTGTCGATGGTAGAACGAAGCGGCGTCGAAGCCTGTAAAGCGGCGGTGCACAATCTTCTCGCGCAACGCGTCAGTGGGCTGATCATTAACTATCCGCTGGATGACCAGGATGCCATTGCTGTGGAAGCTGCCTGCACTAATGTTCCGGCGTTATTTCTTGATGTCTCTGACCAGACACCCATCAACAGTATTATTTTCTCCCATGAAGACGGTACGCGACTGGGCGTGGAGCATCTGGTCGCATTGGGTCACCAGCAAATCGCGCTGTTAGCGGGCCCATTAAGTTCTGTCTCGGCGCGTCTGCGTCTGGCTGGCTGGCATAAATATCTCACTCGCAATCAAATTCAGCCGATAGCGGAACGGGAAGGCGACTGGAGTGCCATGTCCGGTTTTCAACAAACCATGCAAATGCTGAATGAGGGCATCGTTCCCACTGCGATGCTGGTTGCCAACGATCAGATGGCGCTGGGCGCAATGCGCGCCATTACCGAGTCCGGGCTGCGCGTTGGTGCGGATATCTCGGTAGTGGGATACGACGATACCGAAGACAGCTCATGTTATATCCCGCCGTTAACCACCATCAAACAGGATTTTCGCCTGCTGGGGCAAACCAGCGTGGACCGCTTGCTGCAACTCTCTCAGGGCCAGGCGGTGAAGGGCAATCAGCTGTTGCCCGTTTCACTGGTGAAAAGAAAAACCACCCTGGCGCCCAATACGCGAACCGCCTCTCCCCGCGCGTTGGCCGATTCATTAATGCAGCTGGCACGACAGGTTTCCCGACTGGAAAGCGGGCAGTAA\u001b[39m\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'MVNVKPVTLYDVAEYAGVSYQTVSRVVNQASHVSAKTREKVEAAMAELNYIPNRVAQQLAGKQSLLIGVATSSLALHAPSQIVAAIKSRADQLGASVVVSMVERSGVEACKAAVHNLLAQRVSGLIINYPLDDQDAIAVEAACTNVPALFLDVSDQTPINSIIFSHEDGTRLGVEHLVALGHQQIALLAGPLSSVSARLRLAGWHKYLTRNQIQPIAEREGDWSAMSGFQQTMQMLNEGIVPTAMLVANDQMALGAMRAITESGLRVGADISVVGYDDTEDSSCYIPPLTTIKQDFRLLGQTSVDRLLQLSQGQAVKGNQLLPVSLVKRKTTLAPNTRTASPRALADSLMQLARQVSRLESGQ*'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y = x[x['locus_tag'].str.contains('LacI')].slice()\n",
    "y.print()\n",
    "y.translate()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "882e0b26-3da2-4769-ad46-08953c052db0",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Editing genes and generating new constructs\n",
    "New genes and constructs can be generated by combining genetic parts via concatenation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "85e0fca6-c42d-49cf-82c3-4b44ea99bcc3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reading dcas9_RFP.gb as genbank file\n",
      "<class 'gene_editor.plasmid'> at 0x7f35cc4d7640\n",
      "molecule_type:DNA\n",
      "topology:circular\n",
      "data_file_division:   \n",
      "date:05-DEC-2022\n",
      "accessions:['<unknown', 'id>']\n",
      "keywords:['']\n",
      "source:\n",
      "organism:. .\n",
      "taxonomy:[]\n",
      "comment:\n",
      "ApEinfo:methylated:1\n",
      "                         locus_tag type    location  length    color\n",
      "0  mRFP, uniprot drFP583, pdb 2H5O  CDS  [0:678](+)     678  \u001b[38;2;128;0;64m#800040\u001b[39m\n",
      "total length:678\n",
      "\n",
      "reading xRFP.gb as genbank file\n",
      "<class 'gene_editor.plasmid'> at 0x7f35cc64a0b0\n",
      "molecule_type:DNA\n",
      "topology:circular\n",
      "data_file_division:   \n",
      "date:05-DEC-2022\n",
      "accessions:['<unknown', 'id>']\n",
      "keywords:['']\n",
      "source:\n",
      "organism:. .\n",
      "taxonomy:[]\n",
      "comment:\n",
      "ApEinfo:methylated:1\n",
      "                      locus_tag type   location  length color\n",
      "0  RBS 1.00 strength, BBa_B0034  RBS  [0:12](+)      12  \u001b[38;2;0;255;255mcyan\u001b[39m\n",
      "total length:12\n",
      "\n",
      "reading xRFP.gb as genbank file\n",
      "<class 'gene_editor.plasmid'> at 0x7f35ccdeca90\n",
      "molecule_type:DNA\n",
      "topology:circular\n",
      "data_file_division:   \n",
      "date:05-DEC-2022\n",
      "accessions:['<unknown', 'id>']\n",
      "keywords:['']\n",
      "source:\n",
      "organism:. .\n",
      "taxonomy:[]\n",
      "comment:\n",
      "ApEinfo:methylated:1\n",
      "                             locus_tag      type   location  length    color\n",
      "0  pLacO2, single operon pLac ZC082818  promoter  [0:38](+)      38  \u001b[38;2;0;128;128m#008080\u001b[39m\n",
      "total length:38\n",
      "\n",
      "<class 'gene_editor.plasmid'> at 0x7f35cc384970\n",
      "molecule_type:DNA\n",
      "topology:circular\n",
      "                             locus_tag      type     location  length    color\n",
      "0  pLacO2, single operon pLac ZC082818  promoter    [0:38](+)      38  \u001b[38;2;0;128;128m#008080\u001b[39m\n",
      "1         RBS 1.00 strength, BBa_B0034       RBS   [44:56](+)      12     \u001b[38;2;0;255;255mcyan\u001b[39m\n",
      "2      mRFP, uniprot drFP583, pdb 2H5O       CDS  [62:740](+)     678  \u001b[38;2;128;0;64m#800040\u001b[39m\n",
      "total length:740\n",
      "\n",
      "\u001b[38;2;0;128;128mAATTGACAATGTGAGCGAGTAACAAGATACTGAGCACA\u001b[39mgagacc\u001b[38;2;0;255;255mAAAGAGGAGAAA\u001b[39mggtctc\u001b[38;2;128;0;64mATGGCGAGTAGCGAAGACGTTATCAAAGAGTTCATGCGTTTCAAAGTTCGTATGGAAGGTTCCGTTAACGGTCACGAGTTCGAAATCGAAGGTGAAGGTGAAGGTCGTCCGTACGAAGGTACCCAGACCGCTAAACTGAAAGTTACCAAAGGTGGTCCGCTGCCGTTCGCTTGGGACATCCTGTCCCCGCAGTTCCAGTACGGTTCCAAAGCTTACGTTAAACACCCGGCTGACATCCCGGACTACCTGAAACTGTCCTTCCCGGAAGGTTTCAAATGGGAACGTGTTATGAACTTCGAAGACGGTGGTGTTGTTACCGTTACCCAGGACTCCTCCCTGCAAGACGGTGAGTTCATCTACAAAGTTAAACTGCGTGGTACCAACTTCCCGTCCGACGGTCCGGTTATGCAGAAAAAAACCATGGGTTGGGAAGCTTCCACCGAACGTATGTACCCGGAAGACGGTGCTCTGAAAGGTGAAATCAAAATGCGTCTGAAACTGAAAGACGGTGGTCACTACGACGCTGAAGTTAAAACCACCTACATGGCTAAAAAACCGGTTCAGCTGCCGGGTGCTTACAAAACCGACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTTGAACAGTACGAACGTGCTGAAGGTCGTCACTCCACCGGTGCTTAA\u001b[39m\n"
     ]
    }
   ],
   "source": [
    "# slice out the RFP gene\n",
    "RFP = pge.plasmid.read('dcas9_RFP.gb')\n",
    "RFP = RFP[RFP['locus_tag'].str.contains('mRFP')].slice()\n",
    "print(RFP.__repr__())\n",
    "# slice out the ribosome binding site\n",
    "RBS = pge.plasmid.read('xRFP.gb')\n",
    "RBS = RBS[RBS['locus_tag'].str.contains('BBa_B0034')].slice()\n",
    "print(RBS.__repr__())\n",
    "# slice out the promoter\n",
    "pLac = pge.plasmid.read('xRFP.gb')\n",
    "pLac = pLac[pLac['locus_tag'].str.contains('pLac')].slice()\n",
    "print(pLac.__repr__())\n",
    "\n",
    "# assemble the promoter, rbs, and mRFP\n",
    "df = pLac + 'gagacc' + RBS + 'ggtctc' + RFP\n",
    "print(df.__repr__())\n",
    "df.print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b53991d-6174-47de-8f9d-214b90101203",
   "metadata": {},
   "source": [
    "## Annotating features and writing genbank files\n",
    "Annotation of new genes or features can be done with the `.annotate` function. The data can be written to genbank files via `.to_genbank`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "id": "b446e236-3fe0-41bf-af4b-249d39b1945a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on method annotate in module gene_editor:\n",
      "\n",
      "annotate(file='', label='', sequence='', pos=[], feature='unknown', color=['cyan', 'dodgerblue'], circular=True, inplace=False) method of gene_editor.plasmid instance\n",
      "    Adds annotations to a plasmid using a parts library\n",
      "    label = name of the gene\n",
      "    sequence = DNA sequence of the feature, which will be used for matching on the plasmid\n",
      "    pos = position of the genetic feature [start, end, strand]\n",
      "    feature = type of genetic feature such as cds, mRNA, primer_bind\n",
      "    color = [fwd_color, rev_color] to use\n",
      "    If sequence is not provided, start and end are used. Otherwise, sequence overrides start, end, and strand options.\n",
      "    inplace = performs modifications inplace\n",
      "    returns a modified plasmid dataframe\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(df.annotate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "cca9a685-6e53-4950-971e-fcaaefddb726",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'gene_editor.plasmid'> at 0x7f35cc384970\n",
      "molecule_type:DNA\n",
      "topology:circular\n",
      "                             locus_tag      type     location  length    color\n",
      "0  pLacO2, single operon pLac ZC082818  promoter    [0:38](+)      38  \u001b[38;2;0;128;128m#008080\u001b[39m\n",
      "1         RBS 1.00 strength, BBa_B0034       RBS   [44:56](+)      12     \u001b[38;2;0;255;255mcyan\u001b[39m\n",
      "2      mRFP, uniprot drFP583, pdb 2H5O       CDS  [62:740](+)     678  \u001b[38;2;128;0;64m#800040\u001b[39m\n",
      "total length:740\n",
      "\n",
      "\u001b[38;2;0;128;128mAATTGACAATGTGAGCGAGTAACAAGATACTGAGCACA\u001b[39mgagacc\u001b[38;2;0;255;255mAAAGAGGAGAAA\u001b[39mggtctc\u001b[38;2;128;0;64mATGGCGAGTAGCGAAGACGTTATCAAAGAGTTCATGCGTTTCAAAGTTCGTATGGAAGGTTCCGTTAACGGTCACGAGTTCGAAATCGAAGGTGAAGGTGAAGGTCGTCCGTACGAAGGTACCCAGACCGCTAAACTGAAAGTTACCAAAGGTGGTCCGCTGCCGTTCGCTTGGGACATCCTGTCCCCGCAGTTCCAGTACGGTTCCAAAGCTTACGTTAAACACCCGGCTGACATCCCGGACTACCTGAAACTGTCCTTCCCGGAAGGTTTCAAATGGGAACGTGTTATGAACTTCGAAGACGGTGGTGTTGTTACCGTTACCCAGGACTCCTCCCTGCAAGACGGTGAGTTCATCTACAAAGTTAAACTGCGTGGTACCAACTTCCCGTCCGACGGTCCGGTTATGCAGAAAAAAACCATGGGTTGGGAAGCTTCCACCGAACGTATGTACCCGGAAGACGGTGCTCTGAAAGGTGAAATCAAAATGCGTCTGAAACTGAAAGACGGTGGTCACTACGACGCTGAAGTTAAAACCACCTACATGGCTAAAAAACCGGTTCAGCTGCCGGGTGCTTACAAAACCGACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTTGAACAGTACGAACGTGCTGAAGGTCGTCACTCCACCGGTGCTTAA\u001b[39m\n"
     ]
    }
   ],
   "source": [
    "print(df.__repr__())\n",
    "df.print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b04088d4-41ca-4777-afef-c46796bda880",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'gene_editor.plasmid'> at 0x7f35cc3d6d10\n",
      "molecule_type:DNA\n",
      "topology:circular\n",
      "                             locus_tag          type     location  length  \\\n",
      "0  pLacO2, single operon pLac ZC082818      promoter    [0:38](+)      38   \n",
      "1                                 BsaI  protein_bind   [38:44](-)       6   \n",
      "2         RBS 1.00 strength, BBa_B0034           RBS   [44:56](+)      12   \n",
      "3                                 BsaI  protein_bind   [56:62](+)       6   \n",
      "4      mRFP, uniprot drFP583, pdb 2H5O           CDS  [62:740](+)     678   \n",
      "\n",
      "     color  \n",
      "0  \u001b[38;2;0;128;128m#008080\u001b[39m  \n",
      "1   \u001b[38;2;255;165;0morange\u001b[39m  \n",
      "2     \u001b[38;2;0;255;255mcyan\u001b[39m  \n",
      "3      \u001b[38;2;255;0;0mred\u001b[39m  \n",
      "4  \u001b[38;2;128;0;64m#800040\u001b[39m  \n",
      "total length:740\n",
      "\n",
      "\u001b[38;2;0;128;128mAATTGACAATGTGAGCGAGTAACAAGATACTGAGCACA\u001b[39m\u001b[38;2;255;165;0mgagacc\u001b[39m\u001b[38;2;0;255;255mAAAGAGGAGAAA\u001b[39m\u001b[38;2;255;0;0mggtctc\u001b[39m\u001b[38;2;128;0;64mATGGCGAGTAGCGAAGACGTTATCAAAGAGTTCATGCGTTTCAAAGTTCGTATGGAAGGTTCCGTTAACGGTCACGAGTTCGAAATCGAAGGTGAAGGTGAAGGTCGTCCGTACGAAGGTACCCAGACCGCTAAACTGAAAGTTACCAAAGGTGGTCCGCTGCCGTTCGCTTGGGACATCCTGTCCCCGCAGTTCCAGTACGGTTCCAAAGCTTACGTTAAACACCCGGCTGACATCCCGGACTACCTGAAACTGTCCTTCCCGGAAGGTTTCAAATGGGAACGTGTTATGAACTTCGAAGACGGTGGTGTTGTTACCGTTACCCAGGACTCCTCCCTGCAAGACGGTGAGTTCATCTACAAAGTTAAACTGCGTGGTACCAACTTCCCGTCCGACGGTCCGGTTATGCAGAAAAAAACCATGGGTTGGGAAGCTTCCACCGAACGTATGTACCCGGAAGACGGTGCTCTGAAAGGTGAAATCAAAATGCGTCTGAAACTGAAAGACGGTGGTCACTACGACGCTGAAGTTAAAACCACCTACATGGCTAAAAAACCGGTTCAGCTGCCGGGTGCTTACAAAACCGACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTTGAACAGTACGAACGTGCTGAAGGTCGTCACTCCACCGGTGCTTAA\u001b[39m\n"
     ]
    }
   ],
   "source": [
    "df = df.annotate(label='BsaI', sequence='GGTCTC', color=['red','orange'], feature='protein_bind')\n",
    "df = df.drop_duplicates()\n",
    "print(df.__repr__())\n",
    "df.print()\n",
    "df.to_genbank('demo_RFP.gb')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "20485041-5836-4abe-8857-0a7414187dcf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOCUS       .                        740 bp    DNA     circular UNK 01-JAN-1980\n",
      "DEFINITION  .\n",
      "ACCESSION   <unknown id>\n",
      "VERSION     <unknown id>\n",
      "KEYWORDS    .\n",
      "SOURCE      .\n",
      "  ORGANISM  .\n",
      "            .\n",
      "FEATURES             Location/Qualifiers\n",
      "     promoter        1..38\n",
      "                     /locus_tag=\"pLacO2, single operon pLac ZC082818\"\n",
      "                     /ApEinfo_label=\"pLacO2, single operon pLac ZC082818\"\n",
      "                     /ApEinfo_fwdcolor=\"#008080\"\n",
      "                     /ApEinfo_revcolor=\"#008080\"\n",
      "     protein_bind    complement(39..44)\n",
      "                     /locus_tag=\"BsaI\"\n",
      "                     /ApEinfo_fwdcolor=\"red\"\n",
      "                     /ApEinfo_revcolor=\"orange\"\n",
      "                     /ApEinfo_label=\"BsaI\"\n",
      "     RBS             45..56\n",
      "                     /locus_tag=\"RBS 1.00 strength, BBa_B0034\"\n",
      "                     /ApEinfo_label=\"RBS 1.00 strength, BBa_B0034\"\n",
      "                     /ApEinfo_fwdcolor=\"cyan\"\n",
      "                     /ApEinfo_revcolor=\"cyan\"\n",
      "     protein_bind    57..62\n",
      "                     /locus_tag=\"BsaI\"\n",
      "                     /ApEinfo_fwdcolor=\"red\"\n",
      "                     /ApEinfo_revcolor=\"orange\"\n",
      "                     /ApEinfo_label=\"BsaI\"\n",
      "     CDS             63..740\n",
      "                     /locus_tag=\"mRFP, uniprot drFP583, pdb 2H5O\"\n",
      "                     /ApEinfo_label=\"mRFP, uniprot drFP583, pdb 2H5O\"\n",
      "                     /ApEinfo_fwdcolor=\"#800040\"\n",
      "                     /ApEinfo_revcolor=\"#800040\"\n",
      "ORIGIN\n",
      "        1 aattgacaat gtgagcgagt aacaagatac tgagcacaga gaccaaagag gagaaaggtc\n",
      "       61 tcatggcgag tagcgaagac gttatcaaag agttcatgcg tttcaaagtt cgtatggaag\n",
      "      121 gttccgttaa cggtcacgag ttcgaaatcg aaggtgaagg tgaaggtcgt ccgtacgaag\n",
      "      181 gtacccagac cgctaaactg aaagttacca aaggtggtcc gctgccgttc gcttgggaca\n",
      "      241 tcctgtcccc gcagttccag tacggttcca aagcttacgt taaacacccg gctgacatcc\n",
      "      301 cggactacct gaaactgtcc ttcccggaag gtttcaaatg ggaacgtgtt atgaacttcg\n",
      "      361 aagacggtgg tgttgttacc gttacccagg actcctccct gcaagacggt gagttcatct\n",
      "      421 acaaagttaa actgcgtggt accaacttcc cgtccgacgg tccggttatg cagaaaaaaa\n",
      "      481 ccatgggttg ggaagcttcc accgaacgta tgtacccgga agacggtgct ctgaaaggtg\n",
      "      541 aaatcaaaat gcgtctgaaa ctgaaagacg gtggtcacta cgacgctgaa gttaaaacca\n",
      "      601 cctacatggc taaaaaaccg gttcagctgc cgggtgctta caaaaccgac atcaaactgg\n",
      "      661 acatcacctc ccacaacgaa gactacacca tcgttgaaca gtacgaacgt gctgaaggtc\n",
      "      721 gtcactccac cggtgcttaa\n",
      "//\n",
      "\n"
     ]
    }
   ],
   "source": [
    "with open('demo_RFP.gb','r') as f:\n",
    "    text = f.read()\n",
    "print(text)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a045a08-923c-44e5-b8d1-aa6109ccdbae",
   "metadata": {},
   "source": [
    "Annotations can also search for amino acid sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "64da7ef4-d638-4351-bb2d-dbd8d8f72046",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MASSEDVIKEFMRFKVRMEGSVNGHEFEIEGEGEGRPYEGTQTAKLKVTKGGPLPFAWDILSPQFQYGSKAYVKHPADIPDYLKLSFPEGFKWERVMNFEDGGVVTVTQDSSLQDGEFIYKVKLRGTNFPSDGPVMQKKTMGWEASTERMYPEDGALKGEIKMRLKLKDGGHYDAEVKTTYMAKKPVQLPGAYKTDIKLDITSHNEDYTIVEQYERAEGRHSTGA*\n",
      "\u001b[38;2;0;128;128mAATTGACAATGTGAGCGAGTAACAAGATACTGAGCACA\u001b[39mgagacc\u001b[38;2;0;255;255mAAAGAGGAGAAA\u001b[39mggtctc\u001b[38;2;128;0;64mATGGCGAGTAGCGAAGACGTTATCAAAGAGTTCATGCGTTTCAAAGTTCGTATGGAAGGTTCCGTTAACGGTCACGAGTTCGAAATCGAAGGTGAAGGTGAAGGTCGTCCGTACGAAGGTACCCAGACCGCTAAACTGAAAGTTACCAAAGGTGGTCCGCTGCCGTTCGCTTGGGACATCCTGTCCCCGCAGTTCCAGTACGGTTCCAAAGCTTACGTTAAACACCCGGCTGACATCCCGGACTACCTGAAACTGTCCTTCCCGGAAGGTTTCAAATGGGAACGTGTTATGAACTTCGAAGACGGTGGTGTTGTTACCGTTACCCAGGACTCCTCCCTGCAAGACGGTGAGTTCATCTACAAAGTTAAACTGCGTGGTACCAACTTCCCGTCCGACGGTCCGGTTATGCAGAAAAAAACCATGGGTTGGGAAGCTTCCACCGAACGTATGTACCCGGAA\u001b[39m\u001b[38;2;255;165;0mGACGGTGCTCTGAAAGGTGAAATCAAAATGCGTCTGAAACTGAAAGACGGT\u001b[39m\u001b[38;2;128;0;64mGGTCACTACGACGCTGAAGTTAAAACCACCTACATGGCTAAAAAACCGGTTCAGCTGCCGGGTGCTTACAAAACCGACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTTGAACAGTACGAACGTGCTGAAGGTCGTCACTCCACCGGTGCTTAA\u001b[39m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/zchen/.local/lib/python3.10/site-packages/Bio/Seq.py:3482: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\n",
      "/home/zchen/.local/lib/python3.10/site-packages/Bio/Seq.py:3482: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\n",
      "/home/zchen/.local/lib/python3.10/site-packages/Bio/Seq.py:3482: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\n",
      "/home/zchen/.local/lib/python3.10/site-packages/Bio/Seq.py:3482: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "print(RFP.translate())\n",
    "\n",
    "df = pLac + 'gagacc' + RBS + 'ggtctc' + RFP\n",
    "df = df.annotate(label='peptide', sequence='DGALKGEIKMRLKLKDG', color='orange')\n",
    "df = df.drop_duplicates()\n",
    "df.print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "691d329a-2ad6-4928-91c6-4bea2d330668",
   "metadata": {},
   "source": [
    "## Extension PCR\n",
    "This library provides functionality for designing primers for cloning the DNA constructs using the `Design` class."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "0e5833e2-bc46-4b3c-8dfe-1ffa38637d82",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on method xtPCR in module gene_editor:\n",
      "\n",
      "xtPCR(fL, seq, fR=None, padding=[2, 2], niter=3, w=[10, 100, 1, 1, 2], verbose=False, get_cost=False) method of gene_editor.Design instance\n",
      "    Find primers which can seed and extend a PCR fragment\n",
      "    fL = flanking sequence on 5' end\n",
      "    seq = sequence on 3' end which gets amplified\n",
      "    fR = flanking sequence on 3' end\n",
      "    padding = number of extra primers to try\n",
      "    w = weights for cost function\n",
      "    method = optimization method\n",
      "    returns list of primers\n",
      "\n"
     ]
    }
   ],
   "source": [
    "pcr = pge.Design()\n",
    "help(pcr.xtPCR)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c296be10-7f4b-482e-99c2-a796bdaa66e9",
   "metadata": {},
   "source": [
    "The following shows to obtain extension PCR primers that will add promoter and rbs sequences to the RFP gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "6de9a6bf-f33b-4d3b-98f2-f058815713d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reading dcas9_RFP.gb as genbank file\n",
      "reading xRFP.gb as genbank file\n",
      "reading xRFP.gb as genbank file\n",
      "<class 'gene_editor.plasmid'> at 0x7f35cc3f25f0\n",
      "molecule_type:DNA\n",
      "topology:circular\n",
      "                             locus_tag      type     location  length    color\n",
      "0  pLacO2, single operon pLac ZC082818  promoter    [0:38](+)      38  \u001b[38;2;0;128;128m#008080\u001b[39m\n",
      "1         RBS 1.00 strength, BBa_B0034       RBS   [44:56](+)      12     \u001b[38;2;0;255;255mcyan\u001b[39m\n",
      "2      mRFP, uniprot drFP583, pdb 2H5O       CDS  [62:740](+)     678  \u001b[38;2;128;0;64m#800040\u001b[39m\n",
      "total length:740\n",
      "\n",
      "\u001b[38;2;0;128;128mAATTGACAATGTGAGCGAGTAACAAGATACTGAGCACA\u001b[39mgagacc\u001b[38;2;0;255;255mAAAGAGGAGAAA\u001b[39mggtctc\u001b[38;2;128;0;64mATGGCGAGTAGCGAAGACGTTATCAAAGAGTTCATGCGTTTCAAAGTTCGTATGGAAGGTTCCGTTAACGGTCACGAGTTCGAAATCGAAGGTGAAGGTGAAGGTCGTCCGTACGAAGGTACCCAGACCGCTAAACTGAAAGTTACCAAAGGTGGTCCGCTGCCGTTCGCTTGGGACATCCTGTCCCCGCAGTTCCAGTACGGTTCCAAAGCTTACGTTAAACACCCGGCTGACATCCCGGACTACCTGAAACTGTCCTTCCCGGAAGGTTTCAAATGGGAACGTGTTATGAACTTCGAAGACGGTGGTGTTGTTACCGTTACCCAGGACTCCTCCCTGCAAGACGGTGAGTTCATCTACAAAGTTAAACTGCGTGGTACCAACTTCCCGTCCGACGGTCCGGTTATGCAGAAAAAAACCATGGGTTGGGAAGCTTCCACCGAACGTATGTACCCGGAAGACGGTGCTCTGAAAGGTGAAATCAAAATGCGTCTGAAACTGAAAGACGGTGGTCACTACGACGCTGAAGTTAAAACCACCTACATGGCTAAAAAACCGGTTCAGCTGCCGGGTGCTTACAAAACCGACATCAAACTGGACATCACCTCCCACAACGAAGACTACACCATCGTTGAACAGTACGAACGTGCTGAAGGTCGTCACTCCACCGGTGCTTAA\u001b[39m\n"
     ]
    }
   ],
   "source": [
    "# slice out the RFP gene\n",
    "RFP = pge.plasmid.read('dcas9_RFP.gb')\n",
    "RFP = RFP[RFP['locus_tag'].str.contains('mRFP')].slice()\n",
    "# slice out the ribosome binding site\n",
    "RBS = pge.plasmid.read('xRFP.gb')\n",
    "RBS = RBS[RBS['locus_tag'].str.contains('BBa_B0034')].slice()\n",
    "# slice out the promoter\n",
    "pLac = pge.plasmid.read('xRFP.gb')\n",
    "pLac = pLac[pLac['locus_tag'].str.contains('pLac')].slice()\n",
    "\n",
    "# assemble the promoter, rbs, and mRFP\n",
    "df = pLac + 'gagacc' + RBS + 'ggtctc' + RFP\n",
    "print(df.__repr__())\n",
    "df.print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "bfe6916f-8a87-458f-a3c6-b57cf5d87b5d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running fwd\n",
      "running rev\n",
      "  locus_tag         Tm                                           sequence  \\\n",
      "0       0_F  56.437456  ACAgagaccAAAGAGGAGAAAggtctc ATGGCGAGTAGCGAAGAC...   \n",
      "1     fin_F  55.207985  AATTGACAATGTGAGCGAGTAACAAGATACTGAGC ACAgagaccA...   \n",
      "0     fin_R  56.777386                                 TTAAGCACCGGTGGAGTG   \n",
      "\n",
      "                   annealed  strand  \n",
      "0  ATGGCGAGTAGCGAAGACGTTATC       1  \n",
      "1     ACAgagaccAAAGAGGAGAAA       1  \n",
      "0        TTAAGCACCGGTGGAGTG      -1  \n",
      "[['0_F' 56.43745579420619\n",
      "  'ACAgagaccAAAGAGGAGAAAggtctc ATGGCGAGTAGCGAAGACGTTATC'\n",
      "  'ATGGCGAGTAGCGAAGACGTTATC' 1]\n",
      " ['fin_F' 55.207984641427004\n",
      "  'AATTGACAATGTGAGCGAGTAACAAGATACTGAGC ACAgagaccAAAGAGGAGAAA'\n",
      "  'ACAgagaccAAAGAGGAGAAA' 1]\n",
      " ['fin_R' 56.777386483231 '  TTAAGCACCGGTGGAGTG' 'TTAAGCACCGGTGGAGTG' -1]]\n"
     ]
    }
   ],
   "source": [
    "pcr = pge.Design()\n",
    "pcr.params['xtPCR']['Tm'] = 55         # target annealing temperature for xtPCR\n",
    "pcr.params['xtPCR']['len'] = [15, 60]  # defines the [min, max] primer lengths\n",
    "\n",
    "insert = pLac + 'gagacc' + RBS + 'ggtctc'\n",
    "res = pcr.xtPCR(insert, RFP, ' ')\n",
    "print(res)\n",
    "print(res.values)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fec38134-d3fe-4754-92ac-36990f128427",
   "metadata": {},
   "source": [
    "## Gibson assembly\n",
    "The following shows how to design primers for gibson assembly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "id": "94e8695f-c592-4665-8688-1daf8c9c00fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on method Gibson in module gene_editor:\n",
      "\n",
      "Gibson(seqlist, w=[10, 1], method='differential_evolution', circular=True) method of gene_editor.Design instance\n",
      "    Design primers for gibson assembly\n",
      "    seqlist = list of sequences to assemble via gibson in order \n",
      "    circular = assemble fragments into a circular construct\n",
      "    returns list of primers\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(pcr.Gibson)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "id": "7a971588-f759-4fe2-8390-a890ad689bc1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reading xRFP.gb as genbank file\n",
      "reading dcas9_RFP.gb as genbank file\n",
      "reading xRFP.gb as genbank file\n"
     ]
    }
   ],
   "source": [
    "# slice out the LacI gene\n",
    "LacI = pge.plasmid.read('xRFP.gb')\n",
    "LacI = LacI[LacI['locus_tag'].str.contains('LacI')].slice()\n",
    "\n",
    "# slice out the RFP gene\n",
    "RFP = pge.plasmid.read('dcas9_RFP.gb')\n",
    "RFP = RFP[RFP['locus_tag'].str.contains('mRFP')].slice()\n",
    "\n",
    "# slice out the origin of replication\n",
    "df = pge.plasmid.read('xRFP.gb')\n",
    "vec = df[df['locus_tag'].str.contains('pSC101')]\n",
    "start = vec['start'][0]\n",
    "stop = vec['end'][0]\n",
    "vec = df[start:stop]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "id": "80638f89-eabe-4a88-9bb2-1eb27868ed4e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Invalid input provided for ggsite\n",
      "res.x [10.17460988 14.62532845 27.85966033]\n",
      "res.fun -57.0\n",
      "exclude: []\n",
      "overlaps: ['GCGGGCAGTAAAAAActttt', 'TGCTTAACGCCctttt CTG', 'ttt ATGGTGAATGTGAAAC']\n",
      "Tm overlap: [49.290031925644485, 49.73623253267522, 42.289058891654236]\n",
      "processing primers for frag 0\n",
      "running fwd\n",
      "running rev\n",
      "processing primers for frag 1\n",
      "running fwd\n",
      "running rev\n",
      "processing primers for frag 2\n",
      "running fwd\n",
      "running rev\n",
      "     locus_tag         Tm                                           sequence  \\\n",
      "0  frag0_fin_F  55.851352                       ttt  ATGGTGAATGTGAAACCAGTAAC   \n",
      "1  frag0_fin_R  56.106442                        aaaagTTTT TTACTGCCCGCTTTCCA   \n",
      "2  frag1_fin_F  55.335316            GCGGGCAGTAAAAAActttt  ATGGCGAGTAGCGAAGA   \n",
      "3  frag1_fin_R  56.777386                   CAG aaaagGGCG TTAAGCACCGGTGGAGTG   \n",
      "4  frag2_fin_F  55.363272            TGCTTAACGCCctttt  CTGTCAGACCAAGTTTACGAG   \n",
      "5  frag2_fin_R  54.627626  GTTTCACATTCACCAT aaaagCCCC GTTACATTGTCGATCTGTT...   \n",
      "6         seq0        NaN  ttt ATGGTGAATGTGAAACCAGTAACGTTATACGATGTCGCAGAG...   \n",
      "7         seq1        NaN  GCGGGCAGTAAAAAActttt ATGGCGAGTAGCGAAGACGTTATCA...   \n",
      "8         seq2        NaN  TGCTTAACGCCctttt CTGTCAGACCAAGTTTACGAGCTCGCTTG...   \n",
      "\n",
      "                  annealed  strand  \n",
      "0  ATGGTGAATGTGAAACCAGTAAC     1.0  \n",
      "1        TTACTGCCCGCTTTCCA    -1.0  \n",
      "2        ATGGCGAGTAGCGAAGA     1.0  \n",
      "3       TTAAGCACCGGTGGAGTG    -1.0  \n",
      "4    CTGTCAGACCAAGTTTACGAG     1.0  \n",
      "5  GTTACATTGTCGATCTGTTCATG    -1.0  \n",
      "6                      NaN     NaN  \n",
      "7                      NaN     NaN  \n",
      "8                      NaN     NaN  \n"
     ]
    }
   ],
   "source": [
    "seq = []\n",
    "seq+= [[' ',LacI,'AAAActttt']]\n",
    "seq+= [[' ',RFP,'CGCCctttt']]\n",
    "seq+= [[' ',vec,'GGGGctttt']]\n",
    "\n",
    "pcr = pge.Design()\n",
    "pcr.params['gibson']['Tm'] = 50     # target annealing temperature of gibson fragments    \n",
    "pcr.params['gibson']['window'] = 30 # +/i window in bp around frag edges to look for gibson overlap\n",
    "pcr.params['gibson']['len'] = 20    # length of gibson overlap\n",
    "\n",
    "pcr.params['xtPCR']['Tm'] = 55         # target annealing temperature for xtPCR\n",
    "pcr.params['xtPCR']['len'] = [15, 60]  # defines the [min, max] primer lengths\n",
    "pcr.params['xtPCR']['nM'] = [20, 500] # defines the [seed, finisher] primer conc in nM\n",
    "\n",
    "res = pcr.Gibson(seq)\n",
    "print(res)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ff41969-4672-448a-98fd-903f8e132d37",
   "metadata": {},
   "source": [
    "## Golden gate assembly\n",
    "The following shows how to design primers for golden gate assembly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "id": "2a2911c4-c54f-4d1b-a4b2-b4a42e6ca375",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on method GoldenGate in module gene_editor:\n",
      "\n",
      "GoldenGate(seqlist, exclude=[], w=[0, 1], circular=True) method of gene_editor.Design instance\n",
      "    Design primers for goldengate assembly\n",
      "    seqlist = list of sequences to assemble\n",
      "    exclude = sites to exclude\n",
      "    circular = assemble fragments into a circular construct\n",
      "    verbose = print out assembled construct and highlight plasmid locations\n",
      "    returns list of primers\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(pcr.GoldenGate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 258,
   "id": "1a9656f9-8002-4e1e-9947-2f70cf245e65",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reading xRFP.gb as genbank file\n",
      "reading dcas9_RFP.gb as genbank file\n",
      "reading xRFP.gb as genbank file\n"
     ]
    }
   ],
   "source": [
    "# slice out the LacI gene\n",
    "LacI = pge.plasmid.read('xRFP.gb')\n",
    "LacI = LacI[LacI['locus_tag'].str.contains('LacI')].slice()\n",
    "\n",
    "# slice out the RFP gene\n",
    "RFP = pge.plasmid.read('dcas9_RFP.gb')\n",
    "RFP = RFP[RFP['locus_tag'].str.contains('mRFP')].slice()\n",
    "\n",
    "# slice out the origin of replication\n",
    "df = pge.plasmid.read('xRFP.gb')\n",
    "vec = df[df['locus_tag'].str.contains('pSC101')]\n",
    "start = vec['start'][0]\n",
    "stop = vec['end'][0]\n",
    "vec = df[start:stop]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 259,
   "id": "a3bfa742-a892-4745-b7fb-3fd3060f5197",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "res.x [16.93129484 22.60276769 10.20414393]\n",
      "res.fun -12.0\n",
      "exclude: []\n",
      "overlaps: ['tttt', 'GTCA', 'CGGG']\n",
      "Tm overlap: [-79.98814282127228, -59.96603763048182, -39.688627672698374]\n",
      "processing primers for frag 0\n",
      "running fwd\n",
      "running rev\n",
      "processing primers for frag 1\n",
      "running fwd\n",
      "running rev\n",
      "processing primers for frag 2\n",
      "running fwd\n",
      "running rev\n",
      "     locus_tag         Tm                                           sequence  \\\n",
      "0  frag0_fin_F  55.851352  atatatatggGGTCTCcCGGGGctttt ATGGTGAATGTGAAACCA...   \n",
      "1  frag0_fin_R  56.106442       atatatatggGGTCTCcaaaagTTTT TTACTGCCCGCTTTCCA   \n",
      "2  frag1_fin_F  55.335316            atatatatggGGTCTCctttt ATGGCGAGTAGCGAAGA   \n",
      "3  frag1_fin_R  56.777386  atatatatggGGTCTCcTGACAGaaaagGGCG TTAAGCACCGGTG...   \n",
      "4  frag2_fin_F  57.814294           atatatatggGGTCTCc GTCAGACCAAGTTTACGAGCTC   \n",
      "5  frag2_fin_R  54.627626       atatatatggGGTCTCcCCC GTTACATTGTCGATCTGTTCATG   \n",
      "6         seq0        NaN  atatatatggGGTCTCcCGGGGcttttATGGTGAATGTGAAACCAG...   \n",
      "7         seq1        NaN  atatatatggGGTCTCcttttATGGCGAGTAGCGAAGACGTTATCA...   \n",
      "8         seq2        NaN  atatatatggGGTCTCcGTCAGACCAAGTTTACGAGCTCGCTTGGA...   \n",
      "\n",
      "                  annealed  strand  \n",
      "0  ATGGTGAATGTGAAACCAGTAAC     1.0  \n",
      "1        TTACTGCCCGCTTTCCA    -1.0  \n",
      "2        ATGGCGAGTAGCGAAGA     1.0  \n",
      "3       TTAAGCACCGGTGGAGTG    -1.0  \n",
      "4   GTCAGACCAAGTTTACGAGCTC     1.0  \n",
      "5  GTTACATTGTCGATCTGTTCATG    -1.0  \n",
      "6                      NaN     NaN  \n",
      "7                      NaN     NaN  \n",
      "8                      NaN     NaN  \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/lib/python3/dist-packages/scipy/optimize/_numdiff.py:576: RuntimeWarning: invalid value encountered in subtract\n",
      "  df = fun(x) - f0\n"
     ]
    }
   ],
   "source": [
    "seq = []\n",
    "seq+= [['',LacI,'AAAActttt']]\n",
    "seq+= [['',RFP,'CGCCctttt']]\n",
    "seq+= [['',vec,'GGGGctttt']]\n",
    "\n",
    "pcr = pge.Design()\n",
    "pcr.params['goldengate']['window'] = 20 # +/i window in bp around frag edges to look for overlap\n",
    "pcr.params['goldengate']['ggN'] = 4     # length of golden gate overlap\n",
    "pcr.params['goldengate']['ggsite'] = 'GGTCTCc'     # golden gate enzyme site\n",
    "pcr.params['goldengate']['padding'] = 'atatatatgg' # padding around the golden gate site\n",
    "pcr.params['xtPCR']['len'] = [15, 60]  # defines the [min, max] primer lengths\n",
    "pcr.params['xtPCR']['nM'] = [20, 500] # defines the [seed, finisher] primer conc in nM\n",
    "pcr.params['xtPCR']['Tm'] = 55 # defines the [seed, finisher] primer conc in nM\n",
    "\n",
    "res = pcr.GoldenGate(seq)\n",
    "print(res)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
